---
title: "Gene prioritization at AD GWAS loci"
output: html_document
---

## Introduction
To prioritise genes at Alzheimer's disease risk loci, we integrate multiple sources of evidence:

1. distance from gene to genomic window defined by independent lead SNPs
2. colocalisation across multiple QTL datasets
3. fine-mapping probability of coding variants
4. expression in specific cell types / tissues
5. network-based prioritisation of genes

Each of these inputs (except for distance) has been processed in a separate analysis, and is integrated here.

For the network analysis, we started with an unbiased selection of genes based on the AD genome-wide significant loci, plus some highly likely causal genes from the literature. This gave us a score for every gene based on its enrichment in the AD network. Because gene scores come from a network where all genes at a given locus were left out of the input, this network is unbiased by locus-specific features.

We therefore use the gene scores from the network as unbiased scores that reflect the evidence of each gene being involved in AD. Across our AD loci, we then use the four locus-specific features (distance, coloc, fine-mapping, tissue expression) to predict the gene network scores at our AD loci. By using a hierarchical model, we first assess the enrichment of each predictor separately. We then combine the best predictors into a multi-predictor model, which incorporates both locus-specific evidence and network evidence.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(tidyverse)
library(ggExtra)
library(cowplot)
library(scales)
library(pROC)
library(egg)
knitr::opts_chunk$set(fig.width=7, fig.height=5)

theme_set(theme_bw())

savePlots = F

ad_dir = "/Users/jeremys/work/opentargets/AD_finemap"
locus_genes_file = paste0(ad_dir, "/genes/AD.loci.1Mb_window.genedist.all.tsv")
annotated_snps_file = paste0(ad_dir, "/annotated/AD.meta.annotated.selected.probable.paintor.tsv")
coloc_scores_file = paste0(ad_dir, "/coloc/coloc.AD.meta.colocScores.per_gene.tsv")
gene_expr_file = paste0(ad_dir, "/genes/AD.loci.1Mb_window.gene_overlaps.all.selected_tpms.tsv")
bulk_expr_tpm_file = paste0(ad_dir, "/expression/expr.bulk.gtex.selected.tpm.tsv.gz")
sc_expr_brain_tpm_file = paste0(ad_dir, "/expression/brain_cell_type_tpm.tsv.gz")
bulk_expr_relative_file = paste0(ad_dir, "/expression/expr.bulk.gtex.relative.tsv.gz")
sc_expr_brain_relative_file = paste0(ad_dir, "/expression/expr.sc.brain.relative.tsv.gz")
network_file = paste0(ad_dir, "/network/network.annotated.all.tsv")
network_locus_file = paste0(ad_dir, "/network/ad_network.loci.500kb.tsv")
gene_minp_file = paste0(ad_dir, "/genes/genes.pc.10kb_window.minp.tsv.gz")
output_root = "/Users/jeremys/work/opentargets/AD_finemap/genes/"

locus_genes.df = readr::read_tsv(locus_genes_file)

# Read in expression datasets, both TPM, and relative expression
bulk_expr.tpm.df = readr::read_tsv(bulk_expr_tpm_file)
bulk_expr.rel.df = readr::read_tsv(bulk_expr_relative_file)
sc_brain_expr.tpm.df = readr::read_tsv(sc_expr_brain_tpm_file)
sc_brain_expr.rel.df = readr::read_tsv(sc_expr_brain_relative_file)

# Get the set of all network-ranked genes
# We rescale the page.rank.pctile to be relative to all genes, i.e. a uniform distribution normalisation.
network_all.df = readr::read_tsv(network_file) %>%
  group_by(gene_id) %>%
  rename(page.rank.pctile.orig = page.rank.pctile) %>%
  mutate(networkScore = max(0, page.rank.pctile.orig - 50) / 100 * 2, dist_rank_grp = "all") %>%
  ungroup() %>%
  mutate(page.rank.pctile = rank(page.rank.pctile.orig) * 100 / n())

annotated.snps.df = readr::read_tsv(annotated_snps_file, col_names = T,
                                    col_types = cols(.default = col_character(), mean_prob="d", paintor_pp="d",
                                                     finemap_prob_nc="d", gcta_maxCondProb="d", spliceai_max_DS="d"))

colocScores.df = readr::read_tsv(coloc_scores_file) %>%
  filter(!is.na(ensembl_id))
colocScores.df$maxH4 = apply(colocScores.df %>% select(-locus_name, -ensembl_id, -geneSymbol, -colocScore, -scoreRank), MARGIN = 1,
                             function(x) max(x, na.rm=T))

if (!file.exists(gene_minp_file)) {
  gene_pvalues_file = paste0(ad_dir, "/genes/genes.pc.10kb_window.snp_overlaps.tsv.gz")
  genes.minp.df = readr::read_tsv(gene_pvalues_file) %>%
    na.omit() %>%
    group_by(gene_id) %>%
    summarise(chr = first(chr),
              gene_window_start = first(gene_window_start),
              gene_window_end = first(gene_window_end),
              snp_pos = first(snp_pos),
              snp_id = first(snp_id),
              minp = min(pvalue)) %>%
    filter(minp > -1)
  write_tsv(genes.minp.df, gene_minp_file)
} 
genes.minp.df = readr::read_tsv(gene_minp_file)
  
```


```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 3, fig.width = 5}
geneDistScore = function(x) {
  distBias = 100
  maxDist = 5e5
  scores = (log10(maxDist) - log10((abs(x)+distBias))) / (log10(maxDist) - log10(distBias))
  scores[scores < 0] = 0
  scores
}

geneDistScore(c(1, 10, 100, 1000, 10000, 100000))

p.dist = ggplot(data.frame(x = c(1, 500000)), aes(x = x)) +
  stat_function(fun = geneDistScore) +
  scale_x_log10() + xlab("Gene distance") + ylab("Score")
p.dist
pdf(file = paste0(ad_dir, "/genes/geneDistScore.pdf"), width=5, height=3)
p.dist
dev.off()
```


```{r, warning=FALSE, message=FALSE, echo=FALSE}
# Join together the table of AD Locus genes with the other datasets,
# gene expression, network score, etc.
bulk_expr.tpm.df$maxExpr = apply(bulk_expr.tpm.df[,3:ncol(bulk_expr.tpm.df)], MARGIN=1, FUN=function(x) max(x, na.rm=T))
locus_genes.expr.df = locus_genes.df %>%
  rename(gene_id = geneID) %>%
  left_join(bulk_expr.tpm.df %>% select(-symbol), by="gene_id")
# Keep only genes that are either protein coding, or that have max expression
# across datasets of at least 1 TPM.
locus_genes.expr.df = locus_genes.expr.df %>%
  filter(grepl("protein_coding", biotype) | maxExpr > 1) %>%
  select(-maxExpr, -lead_neglog10p)

write.table(locus_genes.expr.df, file=paste0(output_root, "AD.loci.1Mb_window.expressed_genes.all.tpms.tsv"),
            col.names = T, row.names = F, quote=F, na="", sep="\t")


# Coding variant fine-mapping evidence
codingScores.df = annotated.snps.df %>%
  filter(grepl("missense|frameshift_variant|splice_donor_variant|stop_gained|stop_lost", Consequence)) %>%
  group_by(Gene) %>%
  summarise(codingScore = sum(mean_prob))
#sum(codingScores.df$codingScore)

getGeneScores = function(locus_genes.expr.df, useMaxH4 = T) {
  merged.df = locus_genes.expr.df %>%
    left_join(colocScores.df %>% select(locus_name, ensembl_id, colocScore, colocMaxH4 = maxH4, -scoreRank),
              by = c("locus" = "locus_name", "gene_id" = "ensembl_id")) %>%
    group_by(locus, gene_id) %>%
    mutate(colocScore = ifelse(is.na(colocScore), 0, colocScore),
           colocMaxH4 = ifelse(is.na(colocMaxH4), 0, colocMaxH4)) %>%
    left_join(codingScores.df, by=c("gene_id"="Gene")) %>%
    mutate(codingScore = ifelse(is.na(codingScore), 0, codingScore),
           geneDistScore = geneDistScore(dist),
           geneLogDist = log10(abs(dist) + 1)) %>%
    ungroup() %>%
    mutate(colocScoreRank = rank(-colocScore),
           colocMaxH4Rank = rank(-colocMaxH4))
  
  if (useMaxH4) {
    merged.df = merged.df %>%
      rowwise() %>%
      mutate(colocScore = max(0, colocMaxH4 - 0.8) *5,
             colocScoreRank = colocMaxH4Rank)
  }
  
  # Old version that included brain expression in overall expression score
  # merged.df = merged.df %>%
  #   group_by(locus, gene_id) %>%
  #   mutate(brainExprSpecificityScore = max(0, brain_quantile - 0.5) * 2,
  #          microgliaExprSpecificityScore = max(0, microglia_quantile - 0.5) * 2,
  #          brainExprScore = if_else(is.na(brainExprSpecificityScore), 0, brainExprSpecificityScore),
  #          microgliaExprScore = if_else(is.na(microgliaExprSpecificityScore), 0, microgliaExprSpecificityScore),
  #          exprScore = max(microgliaExprScore, brainExprScore),
  #          meanBrainExp = mean(c(ROSMAP_brain, `BrainSeq brain`, GTEx_hippocampus, `Brain - Cerebellum`, `Brain - Cortex`), na.rm=T),
  #          brainIsExpressedScore = if_else(is.na(meanBrainExp) | meanBrainExp < 1, 0, 1),
  #          microgliaIsExpressedScore = if_else(is.na(primary_microglia) | primary_microglia < 1, 0, 1),
  #          isExpressedScore = max(microgliaIsExpressedScore, brainIsExpressedScore),
  #          microgliaExprLevelScore = if_else(is.na(primary_microglia) | primary_microglia < 1, 0, log10(primary_microglia) / 3),
  #          brainExprLevelScore = if_else(is.na(meanBrainExp) | meanBrainExp < 1, 0, log10(meanBrainExp) / 3),
  #          exprLevelScore = max(microgliaExprLevelScore, brainExprLevelScore)
  #          )
  # merged.df = merged.df %>%
  #   mutate(exprScore = max(microgliaExprScore, brainExprScore))
  
  merged.df = merged.df %>%
    left_join(network_all.df %>% select(gene_id, page.rank.pctile, seed_gene, degree), by="gene_id") %>%
    group_by(gene_id) %>%
    mutate(networkScore = max(0, page.rank.pctile - 50) / 100 * 2) %>%
    mutate(networkScore = if_else(is.na(networkScore), 0, networkScore))
  
  # Rename expression colnames to indicate the dataset
  sc_brain.df = sc_brain_expr.rel.df %>% select(-symbol)
  colnames(sc_brain.df) = c("gene_id", paste0("sc_", colnames(sc_brain.df)[2:ncol(sc_brain.df)]))
  
  bulk_expr.df = bulk_expr.rel.df %>% select(-symbol)
  colnames(bulk_expr.df) = c("gene_id", paste0("bulk_", colnames(bulk_expr.df)[2:ncol(bulk_expr.df)]))
  
  merged.df = merged.df %>%
    left_join(sc_brain.df, by="gene_id") %>%
    left_join(bulk_expr.df, by="gene_id")
  
  getExprScore = function(pctile, tpm) {
    ifelse(is.na(pctile), 0, max(0, pctile - 50) / 50)
  }
  
  merged.df = merged.df %>%
    rowwise() %>%
    mutate(primaryMicrogliaScore = getExprScore(bulk_primary_microglia, bulk_primary_microglia_tpm),
           scMicrogliaScore = getExprScore(sc_Microglia, sc_Microglia_tpm),
           scAstrocyteScore = getExprScore(sc_Astrocyte, sc_Astrocyte_tpm),
           exprScore = (primaryMicrogliaScore + scMicrogliaScore) / 2,
           bulkHippocampusScore = getExprScore(bulk_Brain_Hippocampus, bulk_Brain_Hippocampus_tpm),
           bulkBrainCortexScore = getExprScore(bulk_Brain_Cortex, bulk_Brain_Cortex_tpm),
           totalScore = sum(codingScore, colocScore, geneDistScore, exprScore, networkScore, na.rm = T),
           totalScore.noNetwork = sum(codingScore, colocScore, geneDistScore, exprScore, na.rm = T),
           totalScore.noColoc = sum(codingScore, geneDistScore, exprScore, networkScore, na.rm = T),
           totalScore.noCoding = sum(colocScore, geneDistScore, exprScore, networkScore, na.rm = T),
           totalScore.noDist = sum(codingScore, colocScore, exprScore, networkScore, na.rm = T),
           totalScore.noExpr = sum(codingScore, colocScore, geneDistScore, networkScore, na.rm = T)) %>%
    ungroup()
  merged.df
}

merged.df = locus_genes.expr.df %>%
  select(leadSNP:dist,
         bulk_primary_microglia_tpm = primary_microglia,
         bulk_Brain_Hippocampus_tpm = Brain_Hippocampus,
         bulk_Brain_Cortex_tpm = Brain_Cortex) %>%
  left_join(sc_brain_expr.tpm.df %>% select(gene_id, sc_Microglia_tpm = Microglia, sc_Astrocyte_tpm = Astrocyte), by = "gene_id")
         

merged.df = getGeneScores(merged.df, useMaxH4 = T)

```


```{r echo=FALSE, message=FALSE, warning=FALSE}
# First define the data we're working with, "mergenet", which is all genes at genome-wide
# significant AD regions which are present in the network - so excluding genes (mostly
# non-coding ones) which aren't in the network. Also exclude APOE since I haven't done
# coloc there and am not certain of the fine-mapping window.
apoe_locus_genes = merged.df %>% filter(locus == "APOE") %>% .$gene_id
mergenet.df = merged.df %>% filter(!is.na(page.rank.pctile), locus != "APOE")
mergenet.df = mergenet.df %>% arrange(desc(dist)) %>%
  group_by(locus) %>%
  mutate(dist_rank = rank(abs(dist), ties.method = "first")) %>%
  group_by(gene_id)

# Define categorical columns for distance rank
mergenet.df$dist_rank_grp = ">=5"
mergenet.df$dist_rank_grp[mergenet.df$dist_rank < 5] = "2-4"
mergenet.df$dist_rank_grp[mergenet.df$dist_rank <= 1] = "1"
mergenet.df$dist_rank_grp = factor(mergenet.df$dist_rank_grp, levels=c("1", "2-4", ">=5"))

mergenet.df$dist_rank_fine = ">=5"
mergenet.df$dist_rank_fine[mergenet.df$dist_rank == 4] = "4"
mergenet.df$dist_rank_fine[mergenet.df$dist_rank == 3] = "3"
mergenet.df$dist_rank_fine[mergenet.df$dist_rank == 2] = "2"
mergenet.df$dist_rank_fine[mergenet.df$dist_rank == 1] = "1"
mergenet.df$dist_rank_fine = factor(mergenet.df$dist_rank_fine, levels=c("1", "2", "3", "4", ">=5"))
  
# We will also consider how the predictors fare when considering only genes within 100 kb.
mergenet.100k.df = mergenet.df %>% filter(abs(dist) < 100000)

# A function to split a continuous annotation into categories
categorize = function(vals, thresholds) {
  threshold = sort(thresholds)
  if (length(thresholds) < 1) {
    return(NULL)
  }
  catLevels = c(paste0("<= ", threshold[1]))
  cats = rep(catLevels[1], length(vals))
  for (i in 1:length(thresholds)) {
    if (i == length(thresholds)) {
      catLevels[i+1] = paste0("> ", threshold[i])
      cats[vals > threshold[i]] = catLevels[i+1]
    } else {
      catLevels[i+1] = paste0(threshold[i], "-", threshold[i+1])
      cats[vals > threshold[i]] = catLevels[i+1]
    }
  }
  cats = factor(cats, levels = catLevels)
  return(cats)
}

```

### Locus score enrichment across gene network scores

First let's look at the enrichment of locus-specific scores across levels of the gene network score.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
myspan = 0.4
ggplot(mergenet.df, aes(x=page.rank.pctile, y=geneDistScore)) + geom_point(alpha=0.5) + geom_smooth(span=myspan) + theme_bw()
ggplot(mergenet.df, aes(x=page.rank.pctile, y=geneLogDist)) + geom_point(alpha=0.5) + geom_smooth(span=myspan) + theme_bw()
ggplot(mergenet.df, aes(x=page.rank.pctile, y=colocMaxH4)) + geom_point(alpha=0.5) + geom_smooth(span=myspan) + theme_bw()
ggplot(mergenet.df, aes(x=page.rank.pctile, y=codingScore)) + geom_point(alpha=0.5) + theme_bw()
ggplot(mergenet.df, aes(x=page.rank.pctile, y=sc_Microglia)) + geom_point(alpha=0.5) + geom_smooth(span=myspan) + theme_bw()
ggplot(mergenet.df, aes(x=page.rank.pctile, y=bulk_primary_microglia)) + geom_point(alpha=0.5) + geom_smooth(span=myspan) + theme_bw()
ggplot(mergenet.df, aes(x=page.rank.pctile, y=bulk_Brain_Hippocampus)) + geom_point(alpha=0.5) + geom_smooth(span=myspan) + theme_bw()
```

There is clearly *some* trend towards higher network score for e.g. high values of colocMaxH4, geneDistScore, etc., but the level of enrichment isn't very apparent when plotting this way. Let's stratify by discrete levels of the network score instead. We can look at both boxplots and violin plots to see the distribution of other predictors for high vs low network score genes. I do this considering either genes within 100 kb of our GWAS regions, or within 500 kb.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=9, fig.height=7}
mergenet.df$page.rank.pctile_cat = categorize(mergenet.df$page.rank.pctile, thresholds = c(50, 80))
mergenet.100k.df = mergenet.df %>% filter(abs(dist) < 100000)

merged.gather.df = mergenet.df %>%
  select(dist, page.rank.pctile, page.rank.pctile_cat, geneLogDist, colocMaxH4, codingScore, sc_Microglia, bulk_primary_microglia, bulk_Brain_Hippocampus) %>%
  gather(key = "scoreType", value = "value", geneLogDist:bulk_Brain_Hippocampus)

merged.100kb.gather.df = mergenet.100k.df %>%
  select(dist, page.rank.pctile, page.rank.pctile_cat, geneLogDist, colocMaxH4, codingScore, sc_Microglia, bulk_primary_microglia, bulk_Brain_Hippocampus) %>%
  gather(key = "scoreType", value = "value", geneLogDist:bulk_Brain_Hippocampus)

ggplot(merged.gather.df, aes(x=page.rank.pctile_cat, y=value, fill=page.rank.pctile_cat)) + geom_point(position=position_jitterdodge(), width = 0.2, alpha = 0.3) + geom_boxplot(alpha=0.6, width=0.3, outlier.shape=NA) + theme_bw() + facet_wrap(~ scoreType, nrow=2, scales="free") + ggtitle("Genes within 500kb")

ggplot(merged.100kb.gather.df, aes(x=page.rank.pctile_cat, y=value, fill=page.rank.pctile_cat)) + geom_point(position=position_jitterdodge(), width = 0.2, alpha = 0.3) + geom_boxplot(alpha=0.6, width=0.3, outlier.shape=NA) + theme_bw() + facet_wrap(~ scoreType, nrow=2, scales="free") + ggtitle("Genes within 100kb")

ggplot(merged.gather.df, aes(x=page.rank.pctile_cat, y=value, fill=page.rank.pctile_cat)) + geom_point(position=position_jitterdodge(), width = 0.25, alpha = 0.3) + geom_violin(alpha=0.6, outlier.shape=NA) + theme_bw() + facet_wrap(~ scoreType, nrow=2, scales="free") + ggtitle("Genes within 500kb")

ggplot(merged.100kb.gather.df, aes(x=page.rank.pctile_cat, y=value, fill=page.rank.pctile_cat)) + geom_point(position=position_jitterdodge(), width = 0.25, alpha = 0.3) + geom_violin(alpha=0.6, outlier.shape=NA) + theme_bw() + facet_wrap(~ scoreType, nrow=2, scales="free") + ggtitle("Genes within 100kb")

```

It looks as if each of the predictor scores is higher for high-network genes (>80th %ile), except for the brain expression score, which is no different. And log(gene distance), which is much lower for high-network genes, as expected.

Another way to look at it is with density plots.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=6, fig.height=3.5}
mergenet.df$colocMaxH4_cat = categorize(mergenet.df$colocMaxH4, thresholds = c(0.5, 0.8, 0.9))
mergenet.df$page.rank.pctile_cat = categorize(mergenet.df$page.rank.pctile, thresholds = c(50, 80))
mergenet.df$sc_microglia_cat = categorize(mergenet.df$sc_Microglia, thresholds = c(50, 80))
mergenet.df$bulk_microglia_cat = categorize(mergenet.df$bulk_primary_microglia, thresholds = c(50, 80))
mergenet.df$bulk_Brain_Hippocampus_cat = categorize(mergenet.df$bulk_Brain_Hippocampus, thresholds = c(50, 80))
mergenet.df$codingScore_cat = categorize(mergenet.df$codingScore, thresholds = c(0.05))
mergenet.df$gene_dist_cat = categorize(abs(mergenet.df$dist), thresholds = c(0, 100000))

ggplot(mergenet.df, aes(x=page.rank.pctile, col=gene_dist_cat)) + geom_density() + theme_bw()
ggplot(mergenet.df, aes(x=page.rank.pctile, col=colocMaxH4_cat)) + geom_density() + theme_bw()
ggplot(mergenet.df, aes(x=page.rank.pctile, col=codingScore_cat)) + geom_density() + theme_bw()
ggplot(mergenet.df, aes(x=page.rank.pctile, col=sc_microglia_cat)) + geom_density() + theme_bw()
ggplot(mergenet.df, aes(x=page.rank.pctile, col=bulk_microglia_cat)) + geom_density() + theme_bw()
ggplot(mergenet.df, aes(x=page.rank.pctile, col=bulk_Brain_Hippocampus_cat)) + geom_density() + theme_bw()

size = 14
p.coloc = ggplot(mergenet.df, aes(x=page.rank.pctile, col=colocMaxH4_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.8) + theme_bw(size) + guides(color = "none") + ggtitle("Colocalisation")
p.dist = ggplot(mergenet.df, aes(x=page.rank.pctile, col=gene_dist_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.5) + theme_bw(size) + guides(color = "none") + ggtitle("Gene Distance")
p.coding = ggplot(mergenet.df, aes(x=page.rank.pctile, col=codingScore_cat)) + geom_density(alpha = 0.5, size=1) + theme_bw(size) + guides(color = "none") + ggtitle("Coding variant")
p.expr = ggplot(mergenet.df, aes(x=page.rank.pctile, col=bulk_microglia_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.6) + theme_bw(size) + guides(color = "none") + ggtitle("Expr (microglia)")

pdf(file = paste0(output_root, "predictors_vs_network.density.pdf"), width = 9, height = 6)
cowplot::plot_grid(p.coloc, p.dist, p.coding, p.expr, nrow = 2)
dev.off()

```

Let's also do statistical tests to see if the predictor scores differ significantly for high-network genes (> 80th %ile of page rank scores) relative to low-network genes (<50th %ile).

First for genes within 500 kb.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
wilcox.test(mergenet.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$geneLogDist %>% abs(),
            mergenet.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$geneLogDist %>% abs(), alternative = "less")

wilcox.test(mergenet.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$colocMaxH4,
            mergenet.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$colocMaxH4, alternative = "greater")

wilcox.test(mergenet.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$codingScore,
            mergenet.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$codingScore, alternative = "greater")

wilcox.test(mergenet.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$sc_Microglia,
            mergenet.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$sc_Microglia, alternative = "greater")

wilcox.test(mergenet.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$bulk_primary_microglia,
            mergenet.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$bulk_primary_microglia, alternative = "greater")

wilcox.test(mergenet.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$bulk_Brain_Hippocampus,
            mergenet.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$bulk_Brain_Hippocampus, alternative = "greater")
```

And again for genes within 100 kb.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
wilcox.test(mergenet.100k.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$geneLogDist %>% abs(),
            mergenet.100k.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$geneLogDist %>% abs(), alternative = "less")

wilcox.test(mergenet.100k.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$colocMaxH4,
            mergenet.100k.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$colocMaxH4, alternative = "greater")

wilcox.test(mergenet.100k.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$codingScore,
            mergenet.100k.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$codingScore, alternative = "greater")

wilcox.test(mergenet.100k.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$sc_Microglia,
            mergenet.100k.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$sc_Microglia, alternative = "greater")

wilcox.test(mergenet.100k.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$bulk_primary_microglia,
            mergenet.100k.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$bulk_primary_microglia, alternative = "greater")

wilcox.test(mergenet.100k.df %>% filter(page.rank.pctile_cat == "> 80") %>% .$bulk_Brain_Hippocampus,
            mergenet.100k.df %>% filter(page.rank.pctile_cat == "<= 50") %>% .$bulk_Brain_Hippocampus, alternative = "greater")
```

In general, the predictor distributions for high-network genes are more clearly different from those within 500 kb than from those within 100 kb. This could partly be down to more precise characterisation of the actual distribution (more genes) for genes within 500 kb.

Let's now see what the level of enrichment is, in terms of log-odds for having a high network score based on different values of the locus predictors.

```{r echo=FALSE, message=FALSE, warning=FALSE}
# Set the distance category factor levels so that comparisons are with
# respect to the farthest distance category (100+ kb)
mergenet.df$gene_dist_cat = categorize(abs(mergenet.df$dist), thresholds = c(0, 100000))
mergenet.df$gene_dist_cat = factor(mergenet.df$gene_dist_cat, levels = c("> 1e+05", "0-1e+05", "<= 0"))

mergenet.df$sc_microglia_cat = categorize(abs(mergenet.df$sc_Microglia), thresholds = c(50, 80))
mergenet.df$sc_microglia_cat = factor(mergenet.df$sc_microglia_cat, levels = c("<= 50", "50-80", "> 80"))

mergenet.df$bulk_microglia_cat = categorize(abs(mergenet.df$bulk_primary_microglia), thresholds = c(50, 80))
mergenet.df$bulk_microglia_cat = factor(mergenet.df$bulk_microglia_cat, levels = c("<= 50", "50-80", "> 80"))

mergenet.df$bulk_Brain_Hippocampus_cat = categorize(abs(mergenet.df$bulk_Brain_Hippocampus), thresholds = c(50, 80))
mergenet.df$bulk_Brain_Hippocampus_cat = factor(mergenet.df$bulk_Brain_Hippocampus_cat, levels = c("<= 50", "50-80", "> 80"))

mergenet.df$coloc_cat = categorize(abs(mergenet.df$colocMaxH4), thresholds = c(0.2, 0.8, 0.9))
mergenet.df$coloc_cat = factor(mergenet.df$coloc_cat, levels = c("<= 0.2", "0.2-0.8", "0.8-0.9", "> 0.9"))

mergenet.df$coding_cat = categorize(abs(mergenet.df$codingScore), thresholds = c(0.01))
mergenet.df$coding_cat = factor(mergenet.df$coding_cat, levels = c("<= 0.01", "> 0.01"))

mergenet.pctile_80_vs_lo = mergenet.df %>% filter(page.rank.pctile_cat != "50-80") %>% ungroup()

###############################################################################
# Gene distance
logistic <- glm(page.rank.pctile_cat ~ gene_dist_cat, data=mergenet.pctile_80_vs_lo, family="binomial")
summary(logistic)
#logistic <- glm(page.rank.pctile_cat ~ geneLogDist, data=mergenet.pctile_80_vs_lo, family="binomial")
#summary(logistic)

###############################################################################
# Microglia expression
logistic <- glm(page.rank.pctile_cat ~ sc_microglia_cat, data=mergenet.pctile_80_vs_lo, family="binomial")
summary(logistic)

logistic <- glm(page.rank.pctile_cat ~ bulk_microglia_cat, data=mergenet.pctile_80_vs_lo, family="binomial")
summary(logistic)

###############################################################################
# Brain expression
logistic <- glm(page.rank.pctile_cat ~ bulk_Brain_Hippocampus_cat, data=mergenet.pctile_80_vs_lo, family="binomial")
summary(logistic)

###############################################################################
# Coloc
logistic <- glm(page.rank.pctile_cat ~ coloc_cat, data=mergenet.pctile_80_vs_lo, family="binomial")
summary(logistic)

###############################################################################
# Coding variant presence (doesn't really work - counts too low!)
logistic <- glm(page.rank.pctile_cat ~ coding_cat, data=mergenet.pctile_80_vs_lo, family="binomial")
summary(logistic)

```

```{r echo=FALSE, message=FALSE, warning=FALSE}
# This function shows the odds ratios for entries in cat1 to be in each
# of the levels of cat2, relative to the lowest level of cat2.
stratifiedOddsRatios = function(df, cat1, cat2, vsBaseCategory = T) {
  df = df[, c(cat1, cat2)]
  if (any(is.na(df))) {
    warning("stratifiedOddsRatios: Note - input table has NA values. These will be omitted.")
    df = na.omit(df)
  }
  cat1levels = levels(pull(df,cat1))
  cat2levels = levels(pull(df,cat2))
  cat1base = levels(pull(df,cat1))[1]
  cat2base = levels(pull(df,cat2))[1]
  oddsRatio.df = data.frame(comparisonStr = character(), cat1 = character(), cat2 = character(),
                            estimate = numeric(), confint_lo = numeric(), confint_hi = numeric())
  cat1higherLevels = cat1levels[cat1levels != cat1base]
  cat2higherLevels = cat2levels[cat2levels != cat2base]
  
  for (cat1level in cat1higherLevels) {
    for (cat2level in cat2higherLevels) {
      if (vsBaseCategory) {
        mat = matrix(c(sum(df[,cat1] == cat1level & df[,cat2] == cat2level),
                       sum(df[,cat1] == cat1base & df[,cat2] == cat2level),
                       sum(df[,cat1] == cat1level & df[,cat2] == cat2base),
                       sum(df[,cat1] == cat1base & df[,cat2] == cat2base)), nrow=2)
        comparisonStr = sprintf("%s vs. %s to %s vs. %s", cat1level, cat1base, cat2level, cat2base)
      } else {
        mat = matrix(c(sum(df[,cat1] == cat1level & df[,cat2] == cat2level),
                       sum(df[,cat1] != cat1level & df[,cat2] == cat2level),
                       sum(df[,cat1] == cat1level & df[,cat2] != cat2level),
                       sum(df[,cat1] != cat1level & df[,cat2] != cat2level)), nrow=2)
        comparisonStr = sprintf("%s to %s", cat1level, cat2level)
      }
      res = fisher.test(mat)
      oddsRatio.df = bind_rows(oddsRatio.df,
                               data.frame(comparisonStr, cat1 = cat1level, cat2 = cat2level,
                                          estimate = res$estimate, confint_lo = res$conf.int[1], confint_hi = res$conf.int[2]))
    }
  }
  oddsRatio.df
}

stratifiedOddsRatioPlot = function(or.df, cat1name, cat2name, cat1levels = NULL, cat2levels = NULL) {
  if (!is.null(cat1levels)) {
    or.df$cat1 = factor(or.df$cat1, levels = cat1levels)
  }
  if (!is.null(cat2levels)) {
    or.df$cat2 = factor(or.df$cat2, levels = cat2levels)
  }
  ggplot(or.df, aes(x=cat1, y=estimate, fill=cat2)) +
    geom_bar(stat = "identity", position = "dodge") +
    geom_errorbar(aes(ymin = confint_lo, ymax = confint_hi), position = position_dodge2(width = 0.4, padding = 0.6), color = "grey40") +
    scale_y_log10() + ylab("Odds ratio") +
    xlab(cat1name) +
    scale_fill_discrete(name = cat2name)
}

mergenet.df$page.rank.pctile_cat = categorize(mergenet.df$page.rank.pctile, thresholds = c(50, 80, 90))
pr_pctile_cat_levels = c("<= 50", "50-80", "80-90", "> 90")
mergenet.df$page.rank.pctile_cat = factor(mergenet.df$page.rank.pctile_cat, levels = pr_pctile_cat_levels)

mergenet.df$gene_dist_cat2 = categorize(abs(mergenet.df$dist), thresholds = c(0, 10000, 100000, 200000))
gene_dist_cat_levels = c("> 2e+05", "1e+05-2e+05", "10000-1e+05", "0-10000", "<= 0")
mergenet.df$gene_dist_cat2 = factor(mergenet.df$gene_dist_cat2, levels = gene_dist_cat_levels)
xtabs(~ gene_dist_cat2 + page.rank.pctile_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "gene_dist_cat2", "page.rank.pctile_cat")
stratifiedOddsRatioPlot(or.df, "gene_dist_cat2", "page.rank.pctile_cat", cat1levels = gene_dist_cat_levels, cat2levels = pr_pctile_cat_levels) +
  ggtitle("Gene distance enrichment")


mergenet.df$coloc_cat2 = categorize(mergenet.df$colocMaxH4, thresholds = c(0.5, 0.8, 0.9, 0.95))
coloc_cat_levels = c("<= 0.5", "0.5-0.8", "0.8-0.9", "0.9-0.95", "> 0.95")
mergenet.df$coloc_cat2 = factor(mergenet.df$coloc_cat2, levels = coloc_cat_levels)
xtabs(~ coloc_cat2 + page.rank.pctile_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "coloc_cat2", "page.rank.pctile_cat")
stratifiedOddsRatioPlot(or.df, "coloc_cat2", "page.rank.pctile_cat", cat1levels = coloc_cat_levels, cat2levels = pr_pctile_cat_levels) +
  ggtitle("Coloc enrichment") + xlab("Max coloc score across datasets")


mergenet.df$sc_microglia_cat = categorize(abs(mergenet.df$sc_Microglia), thresholds = c(50, 80, 90))
microglia_expr_cat_levels = c("<= 50", "50-80", "80-90", "> 90")
mergenet.df$sc_microglia_cat = factor(mergenet.df$sc_microglia_cat, levels = microglia_expr_cat_levels)
xtabs(~ sc_microglia_cat + page.rank.pctile_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "sc_microglia_cat", "page.rank.pctile_cat")
stratifiedOddsRatioPlot(or.df, "sc_microglia_cat", "page.rank.pctile_cat", cat1levels = microglia_expr_cat_levels, cat2levels = pr_pctile_cat_levels) +
  ggtitle("SC Microglia expression enrichment") + xlab("SC Microglia relative expression")


mergenet.df$bulk_microglia_cat = categorize(abs(mergenet.df$bulk_primary_microglia), thresholds = c(50, 80, 90))
mergenet.df$bulk_microglia_cat = factor(mergenet.df$bulk_microglia_cat, levels = microglia_expr_cat_levels)
xtabs(~ bulk_microglia_cat + page.rank.pctile_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "bulk_microglia_cat", "page.rank.pctile_cat")
stratifiedOddsRatioPlot(or.df, "bulk_microglia_cat", "page.rank.pctile_cat", cat1levels = microglia_expr_cat_levels, cat2levels = pr_pctile_cat_levels) +
  ggtitle("Primary Microglia expression enrichment") + xlab("Primary Microglia relative expression")


mergenet.df$bulk_Brain_Hippocampus_cat = categorize(abs(mergenet.df$bulk_Brain_Hippocampus), thresholds = c(50, 80, 90))
brain_expr_cat_levels = c("<= 50", "50-80", "80-90", "> 90")
mergenet.df$bulk_Brain_Hippocampus_cat = factor(mergenet.df$bulk_Brain_Hippocampus_cat, levels = brain_expr_cat_levels)
xtabs(~ bulk_Brain_Hippocampus_cat + page.rank.pctile_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "bulk_Brain_Hippocampus_cat", "page.rank.pctile_cat")
stratifiedOddsRatioPlot(or.df, "bulk_Brain_Hippocampus_cat", "page.rank.pctile_cat", cat1levels = brain_expr_cat_levels, cat2levels = pr_pctile_cat_levels) +
  ggtitle("Hippocampus expression enrichment") + xlab("Hippocampus relative expression")


mergenet.df$coding_cat = categorize(abs(mergenet.df$codingScore), thresholds = c(0.05))
coding_cat_levels = c("<= 0.05", "> 0.05")
mergenet.df$coding_cat = factor(mergenet.df$coding_cat, levels = coding_cat_levels)
xtabs(~ coding_cat + page.rank.pctile_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "coding_cat", "page.rank.pctile_cat")
stratifiedOddsRatioPlot(or.df, "coding_cat", "page.rank.pctile_cat", cat1levels = coding_cat_levels, cat2levels = pr_pctile_cat_levels) +
  ggtitle("Coding variant enrichment") + xlab("Coding variant fine-mapped probability")

```

We now make a combined logistic regression model, using all predictors, to predict high network genes.

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.width=9}
mergenet.df$coloc_cat = categorize(abs(mergenet.df$colocMaxH4), thresholds = c(0.9))
mergenet.df$coloc_cat = factor(mergenet.df$coloc_cat, levels = c("<= 0.9", "> 0.9"))
mergenet.pctile_80_vs_lo = mergenet.df %>% filter(page.rank.pctile_cat != "50-80") %>% ungroup()

xtabs(~ page.rank.pctile_cat + coloc_cat, data = mergenet.pctile_80_vs_lo)
xtabs(~ page.rank.pctile_cat + coding_cat, data = mergenet.pctile_80_vs_lo)
logistic <- glm(page.rank.pctile_cat ~ sc_microglia_cat + bulk_microglia_cat + coloc_cat + coding_cat + geneLogDist, data=mergenet.pctile_80_vs_lo, family="binomial")
summary(logistic)
 
predicted.data <- data.frame(
  probability.of.high_network = logistic$fitted.values,
  high_network = mergenet.pctile_80_vs_lo$page.rank.pctile_cat)

mergenet.df2 = mergenet.df
mergenet.df2$predicted = predict(logistic, newdata = mergenet.df, type="response")
summary(mergenet.df2$predicted)
mergenet.df2 = mergenet.df2 %>% arrange(predicted)

mergenet.df2$index = 1:nrow(mergenet.df2)
mergenet.df2$page.rank.pctile_gt_80 = mergenet.df2$page.rank.pctile > 80
ggplot(mergenet.df2, aes(x=index, y=predicted, col=page.rank.pctile_gt_80)) +
  geom_point(alpha = 0.4)

ggplot(mergenet.df2, aes(x=predicted, y=page.rank.pctile)) +
  geom_point(alpha=0.5) + geom_smooth() +
  geom_text(mapping = aes(label=symbol), size=2, col="blue", hjust=0, nudge_x = 0.01,
            data = mergenet.df2 %>% filter(predicted > 0.55))

mergenet.df2$bin = as.integer(mergenet.df2$index * 20 / (nrow(mergenet.df2)+1)+1)
overallRatioNetworkHi = sum(mergenet.df2$page.rank.pctile >= 80) / sum(mergenet.df2$page.rank.pctile < 80)
mergenet.odds.df = mergenet.df2 %>% group_by(bin) %>%
  summarise(odds = sum(page.rank.pctile >= 80) / sum(page.rank.pctile < 80) / overallRatioNetworkHi)
ggplot(mergenet.odds.df, aes(x=bin, y=odds)) + geom_bar(stat="identity") + geom_hline(yintercept = 1)

```

With this combined model, genes overlapping the GWAS peak are significantly predictive of having a high network score, consistent with the nearest gene often being causal, while being 1 - 100kb away shows little predictive value. Microglial expression > 80th %ile is predictive, but surprisingly, having coloc H4 > 0.8 is NOT predictive. Having a coding variant doesn't show a significant p value because ALL genes with coding variants have a high network score, and so enrichment can't be calculated. However, the AIC of a model with the coding variant predictor is significantly better than one without.

These results perhaps indicate that the network score is mainly capturing microglial gene pathways?

Let's examine the AU-ROC for predicting high-network genes. Here we use the full dataset for training and also for "testing", which will give inflated results. However, the "true positive" set of (high-network) genes is also very imperfect, so that will cause us to under-estimate the performance, probably to a much greater extent than the over-fitting.

```{r echo=FALSE, message=FALSE, warning=FALSE}
mergenet.df2$high_pr = mergenet.df2$page.rank.pctile >= 80
roc.pred <- roc(mergenet.df2$high_pr, mergenet.df2$predicted)

# AUC over full ROC curve
ggroc(list("Logistic pred"=roc.pred)) +
  annotate(geom="text", x=0.95, y=0.6, label=sprintf("Logistic AUC = %.2f", auc(roc.pred)), hjust=0) +
  ggtitle("ROC") +
  theme_classic(12) + scale_color_discrete(guide=F) + geom_abline(intercept=1, slope=1)

```



### Score enrichment across gene distance

I think it is worth doing similar analyses, this time taking *distance* as the thing to be predicted. That is, we want to predict those genes that overlap the peak. (Or alternatively, genes within 10 kb or 100 kb of the peak.) Distance to a GWAS peak is an unbiased source of evidence for gene involvement.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
myspan = 0.4
ggplot(mergenet.df, aes(x=dist, y=page.rank.pctile)) + geom_point(alpha=0.5) + geom_smooth(span=myspan) + theme_bw()
ggplot(mergenet.df, aes(x=dist, y=colocMaxH4)) + geom_point(alpha=0.5) + geom_smooth(span=myspan) + theme_bw()
ggplot(mergenet.df, aes(x=dist, y=codingScore)) + geom_point(alpha=0.5) + theme_bw()
ggplot(mergenet.df, aes(x=dist, y=sc_Microglia)) + geom_point(alpha=0.5) + geom_smooth(span=myspan) + theme_bw()
ggplot(mergenet.df, aes(x=dist, y=bulk_primary_microglia)) + geom_point(alpha=0.5) + geom_smooth(span=myspan) + theme_bw()
ggplot(mergenet.df, aes(x=dist, y=bulk_Brain_Hippocampus)) + geom_point(alpha=0.5) + geom_smooth(span=myspan) + theme_bw()
```

There is clearly an enrichment of high scores for each of the predictors (except brain expression) near a distance of zero. This is strongest for coloc, which isn't surprising because we know that most gene regulation (eQTL / sQTL) is local. Therefore, to a certain extent colocalisations are a proxy for distance (or vice versa).

An alternative way to visualize this is to look at the distance distribution for genes with a high predictor score (e.g. high coloc) vs. low score.

```{r, warning=FALSE, message=FALSE, echo=FALSE, eval=FALSE}
mergenet.df$colocMaxH4_cat = categorize(mergenet.df$colocMaxH4, thresholds = c(0.5, 0.8))
mergenet.df$page.rank.pctile_cat = categorize(mergenet.df$page.rank.pctile, thresholds = c(50, 80))
pr_pctile_cat_levels = c("<= 50", "50-80", "> 80")
mergenet.df$codingScore_cat = categorize(mergenet.df$codingScore, thresholds = c(0.05))
mergenet.df$sc_microglia_cat = categorize(mergenet.df$sc_Microglia, thresholds = c(50, 80))
microglia_expr_cat_levels = c("<= 50", "50-80", "> 80")
mergenet.df$bulk_microglia_cat = categorize(mergenet.df$bulk_primary_microglia, thresholds = c(50, 80))
mergenet.df$bulk_Brain_Hippocampus_cat = categorize(mergenet.df$bulk_Brain_Hippocampus, thresholds = c(50, 80))
mergenet.df$codingScore_cat = categorize(mergenet.df$codingScore, thresholds = c(0.05))

ggplot(mergenet.df, aes(x=dist, col=colocMaxH4_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.8) + theme_bw()
ggplot(mergenet.df %>% filter(abs(dist) < 100000), aes(x=dist, col=colocMaxH4_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.8) + theme_bw()
ggplot(mergenet.df, aes(x=dist, col=page.rank.pctile_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.5) + theme_bw()
ggplot(mergenet.df %>% filter(abs(dist) < 100000), aes(x=dist, col=page.rank.pctile_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.8) + theme_bw()
ggplot(mergenet.df, aes(x=dist, col=codingScore_cat)) + geom_density(alpha = 0.5, size=1) + theme_bw()
ggplot(mergenet.df, aes(x=dist, col=sc_microglia_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.6) + theme_bw()
ggplot(mergenet.df %>% filter(abs(dist) < 100000), aes(x=dist, col=sc_microglia_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.6) + theme_bw()
ggplot(mergenet.df, aes(x=dist, col=bulk_microglia_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.6) + theme_bw()
ggplot(mergenet.df %>% filter(abs(dist) < 100000), aes(x=dist, col=bulk_microglia_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.6) + theme_bw()
ggplot(mergenet.df, aes(x=dist, col=bulk_Brain_Hippocampus_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.6) + theme_bw()
ggplot(mergenet.df %>% filter(abs(dist) < 100000), aes(x=dist, col=bulk_Brain_Hippocampus_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.6) + theme_bw()

size = 14
p.coloc = ggplot(mergenet.df, aes(x=dist, col=colocMaxH4_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.8) + theme_bw(size) + guides(color = "none") + ggtitle("Colocalisation")
p.pagerank = ggplot(mergenet.df, aes(x=dist, col=page.rank.pctile_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.5) + theme_bw(size) + guides(color = "none") + ggtitle("Network")
p.coding = ggplot(mergenet.df, aes(x=dist, col=codingScore_cat)) + geom_density(alpha = 0.5, size=1) + theme_bw(size) + guides(color = "none") + ggtitle("Coding variant")
p.expr = ggplot(mergenet.df, aes(x=dist, col=bulk_microglia_cat)) + geom_density(alpha = 0.5, size=1, adjust = 0.6) + theme_bw(size) + guides(color = "none") + ggtitle("Expr (microglia)")
pdf(file = paste0(output_root, "predictors_vs_distance.density.pdf"), width = 9, height = 6)
cowplot::plot_grid(p.coloc, p.pagerank, p.coding, p.expr, nrow = 2)
dev.off()

```


Let's see what this looks like when broken into distance categories.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=9, fig.height=7}
mergenet.df$gene_dist_cat = categorize(abs(mergenet.df$dist), thresholds = c(0, 10000, 100000, 200000))
# Set the distance category factor levels in order
mergenet.df$gene_dist_cat = factor(as.character(mergenet.df$gene_dist_cat), levels = c("<= 0", "0-10000", "10000-1e+05", "1e+05-2e+05", "> 2e+05"))

merged.gather.df = mergenet.df %>%
  select(dist, gene_dist_cat, dist_rank_fine, dist_rank_grp, geneLogDist, page.rank.pctile_cat, page.rank.pctile, colocMaxH4, codingScore, sc_Microglia, bulk_primary_microglia, bulk_Brain_Hippocampus) %>%
  gather(key = "scoreType", value = "value", page.rank.pctile:bulk_Brain_Hippocampus)

ggplot(merged.gather.df, aes(x=gene_dist_cat, y=value, fill=gene_dist_cat)) + geom_point(position=position_jitterdodge(), width = 0.3, alpha = 0.3) + geom_boxplot(alpha=0.6, width=0.35, outlier.shape=NA) + theme_bw() + facet_wrap(~ scoreType, nrow=2, scales="free") + ggtitle("Genes within 500kb") + theme(axis.text.x = element_text(angle = 35, hjust = 1))

ggplot(merged.gather.df, aes(x=gene_dist_cat, y=value, fill=gene_dist_cat)) + geom_point(position=position_jitterdodge(), width = 0.3, alpha = 0.3) + geom_violin(alpha=0.6, outlier.shape=NA) + theme_bw() + facet_wrap(~ scoreType, nrow=2, scales="free") + ggtitle("Genes within 500kb") + theme(axis.text.x = element_text(angle = 35, hjust = 1))

ggplot(merged.gather.df, aes(x=dist_rank_fine, y=value, fill=dist_rank_fine)) + geom_point(position=position_jitterdodge(), width = 0.35, alpha = 0.3) + geom_boxplot(alpha=0.6, width=0.4, outlier.shape=NA) + theme_bw() + facet_wrap(~ scoreType, nrow=2, scales="free") + ggtitle("Genes within 500kb") + theme(axis.text.x = element_text(angle = 35, hjust = 1))


```

Clear enrichment in high predictor scores is seen for genes overlapping the peak, as well as for genes within 10 kb of the peak. When we look at the distance ranks of the genes, almost all of this enrichment is accounted for by the nearest gene. There is very little enrichment for genes ranked 2, 3, 4 by distance, with the possible exception of microglial expression, which is also higher for second-nearest genes.

We can also look at odds ratios for predicting nearby genes.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=7, fig.height=3.5}
mergenet.df$gene_dist_cat = categorize(abs(mergenet.df$dist), thresholds = c(10000, 100000))
# Set the distance category factor levels in order
gene_dist_cat_levels = c("> 1e+05", "10000-1e+05", "<= 10000")
mergenet.df$gene_dist_cat = factor(as.character(mergenet.df$gene_dist_cat), levels = gene_dist_cat_levels)

mergenet.df$page.rank.pctile_cat = categorize(mergenet.df$page.rank.pctile, thresholds = c(50, 80, 90, 95))
pr_pctile_cat_levels = c("<= 50", "50-80", "80-90", "90-95", "> 95")

xtabs(~ page.rank.pctile_cat + gene_dist_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "page.rank.pctile_cat", "gene_dist_cat")
stratifiedOddsRatioPlot(or.df, "page.rank.pctile_cat", "gene_dist_cat", cat1levels = pr_pctile_cat_levels, cat2levels = gene_dist_cat_levels) +
  ggtitle("Network enrichment") + xlab("Pagerank percentile")

xtabs(~ sc_microglia_cat + gene_dist_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "sc_microglia_cat", "gene_dist_cat")
stratifiedOddsRatioPlot(or.df, "sc_microglia_cat", "gene_dist_cat", cat1levels = microglia_expr_cat_levels, cat2levels = gene_dist_cat_levels) +
  ggtitle("SC Microglia expression enrichment") + xlab("SC Microglia relative expression")

xtabs(~ bulk_microglia_cat + gene_dist_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "bulk_microglia_cat", "gene_dist_cat")
stratifiedOddsRatioPlot(or.df, "bulk_microglia_cat", "gene_dist_cat", cat1levels = microglia_expr_cat_levels, cat2levels = gene_dist_cat_levels) +
  ggtitle("Primary Microglia expression enrichment") + xlab("Primary Microglia relative expression")

xtabs(~ bulk_Brain_Hippocampus_cat + gene_dist_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "bulk_Brain_Hippocampus_cat", "gene_dist_cat")
stratifiedOddsRatioPlot(or.df, "bulk_Brain_Hippocampus_cat", "gene_dist_cat", cat1levels = microglia_expr_cat_levels, cat2levels = gene_dist_cat_levels) +
  ggtitle("Brain expression enrichment") + xlab("Brain relative expression")

xtabs(~ coloc_cat2 + gene_dist_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "coloc_cat2", "gene_dist_cat")
stratifiedOddsRatioPlot(or.df, "coloc_cat2", "gene_dist_cat", cat1levels = coloc_cat_levels, cat2levels = gene_dist_cat_levels) +
  ggtitle("Coloc enrichment") + xlab("Max coloc score across datasets")

xtabs(~ coding_cat + gene_dist_cat, data = mergenet.df)
or.df = stratifiedOddsRatios(mergenet.df, "coding_cat", "gene_dist_cat")
stratifiedOddsRatioPlot(or.df, "coding_cat", "gene_dist_cat", cat1levels = coding_cat_levels, cat2levels = gene_dist_cat_levels) +
  ggtitle("Coding variant enrichment") + xlab("Coding variant fine-mapped probability")

```


Let's statistically test whether genes in different distance windows have higher predictor scores than distal genes. First page.rank.pctile.

```{r echo=FALSE, message=FALSE, warning=FALSE}
mergenet.df$gene_dist_cat = categorize(abs(mergenet.df$dist), thresholds = c(0, 10000, 100000, 200000))
mergenet.df$gene_dist_cat = factor(as.character(mergenet.df$gene_dist_cat), levels = c("<= 0", "0-10000", "10000-1e+05", "1e+05-2e+05", "> 2e+05"))

wilcox.test(mergenet.df %>% filter(gene_dist_cat == "<= 0") %>% .$page.rank.pctile,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$page.rank.pctile,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "0-10000") %>% .$page.rank.pctile,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$page.rank.pctile,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "10000-1e+05") %>% .$page.rank.pctile,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$page.rank.pctile,
            alternative = "greater")
```

Only overlapping genes have detectably higher page.rank.pctile.

Next, coloc.

```{r echo=FALSE, message=FALSE, warning=FALSE}
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "<= 0") %>% .$colocMaxH4,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$colocMaxH4,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "0-10000") %>% .$colocMaxH4,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$colocMaxH4,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "10000-1e+05") %>% .$colocMaxH4,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$colocMaxH4,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "1e+05-2e+05") %>% .$colocMaxH4,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$colocMaxH4,
            alternative = "greater")
```

All of the distances are enriched for high-coloc genes, relative to distal (>200k) genes, although 100 - 200 kb distant is barely enriched.

Next, check codingScore.

```{r echo=FALSE, message=FALSE, warning=FALSE}
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "<= 0") %>% .$codingScore,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$codingScore,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "0-10000") %>% .$codingScore,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$codingScore,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "10000-1e+05") %>% .$codingScore,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$codingScore,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "1e+05-2e+05") %>% .$codingScore,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$codingScore,
            alternative = "greater")
```

Mainly overlapping genes are enriched for high coding score, but there is enrichment in the next two distance bins as well.

Finally, look at expression. First single-cell microglia.

```{r echo=FALSE, message=FALSE, warning=FALSE}
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "<= 0") %>% .$sc_Microglia,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$sc_Microglia,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "0-10000") %>% .$sc_Microglia,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$sc_Microglia,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "10000-1e+05") %>% .$sc_Microglia,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$sc_Microglia,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "1e+05-2e+05") %>% .$sc_Microglia,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$sc_Microglia,
            alternative = "greater")
```

And also bulk microglia.

```{r echo=FALSE, message=FALSE, warning=FALSE}
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "<= 0") %>% .$bulk_primary_microglia,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$bulk_primary_microglia,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "0-10000") %>% .$bulk_primary_microglia,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$bulk_primary_microglia,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "10000-1e+05") %>% .$bulk_primary_microglia,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$bulk_primary_microglia,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "1e+05-2e+05") %>% .$bulk_primary_microglia,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$bulk_primary_microglia,
            alternative = "greater")
```

All nearby gene distance bins, except for 100 - 200 kb, are enriched for high microglial expression relative to distal genes (>200 kb).

Now look at brain expression, specifically GTEx hippocampus relative expression.

```{r echo=FALSE, message=FALSE, warning=FALSE}
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "<= 0") %>% .$bulk_Brain_Hippocampus,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$bulk_Brain_Hippocampus,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "0-10000") %>% .$bulk_Brain_Hippocampus,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$bulk_Brain_Hippocampus,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "10000-1e+05") %>% .$bulk_Brain_Hippocampus,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$bulk_Brain_Hippocampus,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(gene_dist_cat == "1e+05-2e+05") %>% .$bulk_Brain_Hippocampus,
            mergenet.df %>% filter(gene_dist_cat == "> 2e+05") %>% .$bulk_Brain_Hippocampus,
            alternative = "greater")
```

Use a logistic regression model to predict AD genes based on distance, i.e. those near peaks, using the other data as predictors.

```{r echo=FALSE, message=FALSE, warning=FALSE}
mergenet.df$gene_dist_cat = categorize(abs(mergenet.df$dist), thresholds = c(10000, 100000))
# Set the distance category factor levels in order
mergenet.df$gene_dist_cat = factor(as.character(mergenet.df$gene_dist_cat), levels = c("> 1e+05", "10000-1e+05", "<= 10000"))

mergenet.df$coloc_cat = categorize(abs(mergenet.df$colocMaxH4), thresholds = c(0.2, 0.8, 0.95))
mergenet.df$coloc_cat = factor(mergenet.df$coloc_cat, levels = c("<= 0.2", "0.2-0.8", "0.8-0.95", "> 0.95"))

mergenet.df$page.rank.pctile_cat = categorize(mergenet.df$page.rank.pctile, thresholds = c(50, 80, 90))

mergenet.near_vs_far = mergenet.df %>% filter(gene_dist_cat != "10000-1e+05")

xtabs(~ gene_dist_cat + coloc_cat, data = mergenet.near_vs_far)
xtabs(~ gene_dist_cat + coding_cat, data = mergenet.near_vs_far)
xtabs(~ gene_dist_cat + sc_microglia_cat, data = mergenet.near_vs_far)
xtabs(~ gene_dist_cat + bulk_microglia_cat, data = mergenet.near_vs_far)
xtabs(~ gene_dist_cat + page.rank.pctile_cat, data = mergenet.near_vs_far)

logistic <- glm(gene_dist_cat ~ sc_microglia_cat, data=mergenet.near_vs_far, family="binomial")
summary(logistic)
logistic <- glm(gene_dist_cat ~ bulk_microglia_cat, data=mergenet.near_vs_far, family="binomial")
summary(logistic)
logistic <- glm(gene_dist_cat ~ page.rank.pctile_cat, data=mergenet.near_vs_far, family="binomial")
summary(logistic)
logistic <- glm(gene_dist_cat ~ coloc_cat, data=mergenet.near_vs_far, family="binomial")
summary(logistic)
logistic <- glm(gene_dist_cat ~ coding_cat, data=mergenet.near_vs_far, family="binomial")
summary(logistic)

logistic <- glm(gene_dist_cat ~ sc_microglia_cat + bulk_microglia_cat + coloc_cat + page.rank.pctile_cat, data=mergenet.near_vs_far, family="binomial")
summary(logistic)

logistic <- glm(gene_dist_cat ~ sc_microglia_cat + bulk_microglia_cat + coloc_cat + page.rank.pctile_cat + coding_cat, data=mergenet.near_vs_far, family="binomial")
summary(logistic)

mergenet.df2 = mergenet.df
mergenet.df2$predicted = predict(logistic, newdata = mergenet.df, type="response")
summary(mergenet.df2$predicted)
mergenet.df2 = mergenet.df2 %>% arrange(predicted)
#View(mergenet.df %>% select(locus, symbol, dist, microglia_expr_cat, coloc_cat, page.rank.pctile_cat, coding_cat, predicted, predicted_odds) %>% arrange(locus, desc(predicted)))

mergenet.df2$index = 1:nrow(mergenet.df2)
mergenet.df2$dist_lt_10k = mergenet.df2$gene_dist_cat == "<= 10000"
ggplot(mergenet.df2, aes(x=index, y=predicted, col=dist_lt_10k)) +
  geom_point(alpha = 0.4)

ggplot(mergenet.df2, aes(x=predicted, y=abs(dist+1))) +
  geom_point(alpha=0.5) + geom_smooth() +
  geom_text(mapping = aes(label=symbol), size=2, col="blue", hjust=0, nudge_x = 0.01,
            data = mergenet.df2 %>% filter(predicted > 0.55)) +
  scale_y_log10()

overallRatioNear = sum(mergenet.df2$gene_dist_cat == "<= 10000") / sum(mergenet.df2$gene_dist_cat == "> 1e+05")
mergenet.df2$bin = as.integer(mergenet.df2$index * 20 / (nrow(mergenet.df2)+1)+1)
mergenet.odds.df = mergenet.df2 %>% group_by(bin) %>%
  summarise(odds = sum(gene_dist_cat == "<= 10000") / sum(gene_dist_cat == "> 1e+05") / overallRatioNear)

ggplot(mergenet.odds.df, aes(x=bin, y=odds)) + geom_bar(stat="identity") + geom_hline(yintercept = 1) + scale_y_log10()

```

```{r echo=FALSE, message=FALSE, warning=FALSE, fig.width=9}
mergenet.df2$index = 1:nrow(mergenet.df2)
mergenet.df2$page.rank.pctile_gt_80 = mergenet.df2$page.rank.pctile > 80
ggplot(mergenet.df2, aes(x=index, y=predicted, col=page.rank.pctile_gt_80)) +
  geom_point(alpha = 0.4)

ggplot(mergenet.df2, aes(x=predicted, y=page.rank.pctile)) +
  geom_point(alpha=0.5) + geom_smooth() +
  geom_text(mapping = aes(label=symbol), size=2, col="blue", hjust=0, nudge_x = 0.01,
            data = mergenet.df2 %>% filter(predicted > 0.55))

mergenet.df2$bin = as.integer(mergenet.df2$index * 20 / (nrow(mergenet.df2)+1)+1)
mergenet.odds.df = mergenet.df2 %>% group_by(bin) %>%
  summarise(odds = sum(page.rank.pctile >= 80) / sum(page.rank.pctile < 80))

overallOdds = sum(mergenet.df2$page.rank.pctile >= 80) / sum(mergenet.df2$page.rank.pctile < 80)
ggplot(mergenet.odds.df, aes(x=bin, y=odds)) + geom_bar(stat="identity") + geom_hline(yintercept = overallOdds)

```


### Coloc vs network scores

Examine the relationship between coloc scores and network scores. Here, we want to see if coloc in certain cell type groups associated with higher network scores, and whether the coloc threshold matters. So we look at genes with coloc > 0.9, or 0.8. First we look at broad cell type groups, and then at more refined groupings.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
coloc_dataset_groups.df = readr::read_tsv(paste0(ad_dir, "/coloc/coloc_dataset_groups.tsv"))
coloc_dataset_groups.df = bind_rows(coloc_dataset_groups.df,
                                    data.frame(dataset_short="maxH4", celltype_group="maxH4", celltype_group_broad="maxH4"))

colocNet.df = mergenet.df %>%
  select(locus, symbol, gene_id, page.rank.pctile, seed_gene, degree, networkScore, totalScore.noColoc) %>%
  group_by(gene_id) %>%
  filter(!is.na(networkScore), !gene_id %in% apoe_locus_genes) %>%
  left_join(colocScores.df, by=c("gene_id" = "ensembl_id")) %>%
  mutate(maxH4 = ifelse(is.na(maxH4), 0, maxH4)) %>%
  gather(key = "qtl_dataset", value = "H4", Alasoo_mac_IFNg:maxH4)
# I notice that after this step there are some loci with a gene that is in the network but that
# has no coloc information. Probably this is because the gene wasn't tested in any of the QTL
# datasets, although I wonder why this would be the case. An example is LPO at the TSPOAP1-MIR142
# locus. It could also be something to do with the naming of that locus.

colocNet.df = colocNet.df %>%
  left_join(coloc_dataset_groups.df %>% select(dataset_short, celltype_group, celltype_group_broad), by=c("qtl_dataset" = "dataset_short"))

colocNet.groupSummary.df = colocNet.df %>%
  filter(!is.na(celltype_group_broad)) %>%
  group_by(gene_id, celltype_group_broad) %>%
  summarise(locus_name = first(locus_name),
            geneSymbol = first(geneSymbol),
            degree = first(degree),
            page.rank.pctile = first(page.rank.pctile),
            networkScore = first(networkScore),
            totalScore.noColoc = first(totalScore.noColoc),
            meanH4 = if_else(all(is.na(H4)), NaN, mean(H4, na.rm=T)),
            H4 = if_else(all(is.na(H4)), NaN, max(H4, na.rm=T))) %>%
  rename(qtl_dataset = celltype_group_broad) %>%
  filter(qtl_dataset != "maxH4")

colocNet.df = bind_rows(colocNet.groupSummary.df,
                        colocNet.df %>% select(locus_name, gene_id, geneSymbol, degree, page.rank.pctile, networkScore, totalScore.noColoc, qtl_dataset, H4))

colocData.df = bind_rows(colocNet.df %>% filter(qtl_dataset %in% c("brain", "myeloid", "lymphoid", "other", "blood", "maxH4")),
                         network_all.df %>% select(gene_id, page.rank.pctile) %>% mutate(qtl_dataset = "background"))

colocPlot.df = colocData.df %>% filter(H4 > 0.9) %>% mutate(colocSet = "H4 > 0.9")
ggplot(colocPlot.df, aes(x=fct_reorder(qtl_dataset, page.rank.pctile, .fun=median), y=page.rank.pctile, fill=colocSet, alpha=colocSet)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.25), alpha=0.5) + geom_boxplot() +
  theme_bw() + theme(axis.text.x = element_text(angle = 35, hjust = 1)) +
  scale_alpha_manual(values=c("H4 > 0.8" = 0, "H4 > 0.9" = 0.5)) +
  ggtitle("Page rank pctile of colocalising genes by cell type group") +
  xlab("QTL study cell type category")

# Here we plot both genes with coloc H4 > 0.8 and H4 > 0.9 grouped separately in the same plot
colocPlot.df = bind_rows(
#  colocData.df %>% filter(H4 > 0.8) %>% mutate(colocSet = "H4 > 0.8"),
  colocData.df %>% filter(H4 > 0.9) %>% mutate(colocSet = "H4 > 0.9"),
  colocData.df %>% filter(H4 > 0.95) %>% mutate(colocSet = "H4 > 0.95"),
  colocData.df %>% filter(H4 > 0.98) %>% mutate(colocSet = "H4 > 0.98"))
ggplot(colocPlot.df, aes(x=fct_reorder(qtl_dataset, page.rank.pctile, .fun=median), y=page.rank.pctile, fill=colocSet, alpha=colocSet)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.25), alpha=0.5) + geom_boxplot() +
  theme_bw() + theme(axis.text.x = element_text(angle = 35, hjust = 1)) +
#  scale_alpha_manual(values=c("H4 > 0.8" = 0, "H4 > 0.9" = 0.5)) +
  ggtitle("Page rank pctile of colocalising genes by cell type group") +
  xlab("QTL study cell type category")

ggplot(colocPlot.df, aes(x=fct_reorder(qtl_dataset, totalScore.noColoc, .fun=median), y=totalScore.noColoc, fill=colocSet)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.25), alpha=0.5) + geom_boxplot(alpha=0.5) +
  theme_bw() + theme(axis.text.x = element_text(angle = 35, hjust = 1)) +
#  scale_alpha_manual(values=c("H4 > 0.8" = 0, "H4 > 0.9" = 0.5)) +
  ggtitle("Total score (excl coloc) of colocalising genes by cell type group") +
  xlab("QTL study cell type category")

```

There doesn't appear to be any difference between major cell type groupings in terms of the page.rank.pctile for genes that colocalise. There is a hint that the maxH4 score, which can be considered a cell type grouping of ALL datasets, has lower average page.rank.pctile. This could be because by looking across ALL datasets, more genes are included, which have weaker evidence. Stratifying by coloc H4 > 0.8 or 0.9 suggests that this is indeed the case, and that when looking across many QTL datasets, a more stringent H4 cutoff should be used.

Let's check if this is the case, by looking at maxH4 only, and seeing what the page.rank.pctile is for different H4 groups.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
mergenet.df = mergenet.df %>%
  mutate(H4_grp = if_else(colocMaxH4 > 0.9, ">0.9", if_else(colocMaxH4 > 0.8, "0.8 - 0.9", if_else(colocMaxH4 > 0.5, "0.5 - 0.8", "<0.5"))),
         totalScore.noColocDist = totalScore.noColoc - geneDistScore,
         totalScore.networkExpr = networkScore + exprScore) %>%
  mutate(H4_grp = factor(as.character(H4_grp), levels = c("<0.5", "0.5 - 0.8", "0.8 - 0.9", ">0.9")))

ggplot(mergenet.df, aes(x=H4_grp, y=page.rank.pctile, fill=H4_grp, group=H4_grp)) + geom_boxplot(outlier.shape = NA) + geom_jitter(width = 0.2, alpha=0.4) + theme_bw()
ggplot(mergenet.df, aes(x=H4_grp, y=totalScore.noColoc, fill=H4_grp, group=H4_grp)) + geom_boxplot(outlier.shape = NA) + geom_jitter(width = 0.2, alpha=0.4) + theme_bw()
ggplot(mergenet.df, aes(x=H4_grp, y=totalScore.networkExpr, fill=H4_grp, group=H4_grp)) + geom_boxplot(outlier.shape = NA) + geom_jitter(width = 0.2, alpha=0.4) + theme_bw()

pdf(file = paste0(output_root, "coloc.H4.boxplot.totalScore.pdf"), width = 6, height = 4)
ggplot(mergenet.df, aes(x=H4_grp, y=totalScore.noColoc, fill=H4_grp, group=H4_grp)) + geom_boxplot(outlier.shape = NA) + geom_jitter(width = 0.2, alpha=0.4) +
  theme_bw(14) + xlab("Coloc H4 probability") + ylab("Total score w/o coloc")
dev.off()

pdf(file = paste0(output_root, "coloc.H4.boxplot.network.pdf"), width = 6, height = 4)
ggplot(mergenet.df, aes(x=H4_grp, y=page.rank.pctile, fill=H4_grp, group=H4_grp)) + geom_boxplot(outlier.shape = NA) + geom_jitter(width = 0.2, alpha=0.4) +
  theme_bw(14) + xlab("Coloc H4 probability") + ylab("Pagerank percentile")
dev.off()


wilcox.test(mergenet.df %>% filter(H4_grp == ">0.9") %>% .$page.rank.pctile,
            mergenet.df %>% filter(H4_grp == "<0.5") %>% .$page.rank.pctile,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(H4_grp == "0.8 - 0.9") %>% .$page.rank.pctile,
            mergenet.df %>% filter(H4_grp == "<0.5") %>% .$page.rank.pctile,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(H4_grp == "0.5 - 0.8") %>% .$page.rank.pctile,
            mergenet.df %>% filter(H4_grp == "<0.5") %>% .$page.rank.pctile,
            alternative = "greater")

wilcox.test(mergenet.df %>% filter(H4_grp == ">0.9") %>% .$totalScore.noColoc,
            mergenet.df %>% filter(H4_grp == "<0.5") %>% .$totalScore.noColoc,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(H4_grp == "0.8 - 0.9") %>% .$totalScore.noColoc,
            mergenet.df %>% filter(H4_grp == "<0.5") %>% .$totalScore.noColoc,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(H4_grp == "0.5 - 0.8") %>% .$totalScore.noColoc,
            mergenet.df %>% filter(H4_grp == "<0.5") %>% .$totalScore.noColoc,
            alternative = "greater")

wilcox.test(mergenet.df %>% filter(H4_grp == ">0.9") %>% .$totalScore.networkExpr,
            mergenet.df %>% filter(H4_grp == "<0.5") %>% .$totalScore.networkExpr,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(H4_grp == "0.8 - 0.9") %>% .$totalScore.networkExpr,
            mergenet.df %>% filter(H4_grp == "<0.5") %>% .$totalScore.networkExpr,
            alternative = "greater")
wilcox.test(mergenet.df %>% filter(H4_grp == "0.5 - 0.8") %>% .$totalScore.networkExpr,
            mergenet.df %>% filter(H4_grp == "<0.5") %>% .$totalScore.networkExpr,
            alternative = "greater")
```

Clearly for coloc maxH4, the enrichment is all in the very high values of coloc, i.e. > 0.9.

Let's also look at more detailed cell type groupings.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=9}
colocNet.df = mergenet.df %>%
  select(gene_id, page.rank.pctile, seed_gene, degree, networkScore, totalScore.noColoc) %>%
  group_by(gene_id) %>%
  filter(!is.na(networkScore), !gene_id %in% apoe_locus_genes) %>%
  left_join(colocScores.df, by=c("gene_id" = "ensembl_id")) %>%
  gather(key = "qtl_dataset", value = "H4", Alasoo_mac_IFNg:maxH4)
# I notice that after this step there are some loci with a gene that is in the network but that
# has no coloc information. Probably this is because the gene wasn't tested in any of the QTL
# datasets, although I wonder why this would be the case. An example is LPO at the TSPOAP1-MIR142
# locus. It could also be something to do with the naming of that locus.

colocNet.df = colocNet.df %>%
  left_join(coloc_dataset_groups.df %>% select(dataset_short, celltype_group, celltype_group_broad), by=c("qtl_dataset" = "dataset_short"))

colocNet.groupSummary.df = colocNet.df %>%
  filter(!is.na(celltype_group)) %>%
  group_by(gene_id, celltype_group) %>%
  summarise(locus_name = first(locus_name),
            geneSymbol = first(geneSymbol),
            degree = first(degree),
            page.rank.pctile = first(page.rank.pctile),
            totalScore.noColoc = first(totalScore.noColoc),
            networkScore = first(networkScore),
            meanH4 = if_else(all(is.na(H4)), NaN, mean(H4, na.rm=T)),
            H4 = if_else(all(is.na(H4)), NaN, max(H4, na.rm=T))) %>%
  rename(qtl_dataset = celltype_group) %>%
  filter(qtl_dataset != "maxH4")

colocNet.df = bind_rows(colocNet.groupSummary.df %>% filter(qtl_dataset != "microglia"),
                        colocNet.df %>% select(locus_name, gene_id, geneSymbol, degree, page.rank.pctile, networkScore, totalScore.noColoc, qtl_dataset, H4))

colocData.df = bind_rows(colocNet.df %>% filter(qtl_dataset %in% c("brain","microglia","monocyte, macrophage","neutrophil","Bcell, Tcell","LCL","other","blood","NKcell")),
                         network_all.df %>% select(gene_id, page.rank.pctile) %>% mutate(qtl_dataset = "background")) %>%
  mutate(qtl_dataset = gsub("monocyte, macrophage", "monocyte &\nmacrophage", qtl_dataset))

colocPlot.df = colocData.df %>% filter(H4 > 0.9) %>% mutate(colocSet = "H4 > 0.9")
ggplot(colocPlot.df, aes(x=fct_reorder(qtl_dataset, page.rank.pctile, .fun=median), y=page.rank.pctile)) +
  geom_boxplot(fill = "lightblue") + geom_point(position = position_jitter(width = 0.25), alpha=0.5) +
  theme_bw() + theme(axis.text.x = element_text(angle = 35, hjust = 1)) +
  scale_alpha_manual(values=c("H4 > 0.8" = 0, "H4 > 0.9" = 0.5)) +
  ggtitle("Page rank pctile of colocalising genes by cell type group") +
  xlab("QTL study cell type")

colocPlot.df = colocData.df %>% filter(H4 > 0.90) %>% mutate(colocSet = "H4 > 0.9")
p.coloc = ggplot(colocPlot.df, aes(x=fct_reorder(qtl_dataset, totalScore.noColoc, .fun=median), y=totalScore.noColoc)) +
  geom_boxplot(fill = "lightblue", outlier.shape = NA) + geom_point(position = position_jitter(width = 0.25), alpha=0.5) +
  theme_bw() + theme(axis.text.x = element_text(angle = 35, hjust = 1)) +
  scale_alpha_manual(values=c("H4 > 0.8" = 0, "H4 > 0.9" = 0.5)) +
  ggtitle("Total score (excl coloc) of colocalising genes by cell type group") +
  xlab("QTL study cell type")
p.coloc

# Save plot for a supplementary figure
pdf(file = paste0(output_root, "coloc.celltype_vs_totalscore.pdf"), width=8, height=4)
p.coloc
dev.off()

# Do some tests to see if any of the cell type groups differ form each other
wilcox.test(colocPlot.df %>% filter(qtl_dataset == "microglia") %>% .$totalScore.noColoc,
            colocPlot.df %>% filter(qtl_dataset == "other") %>% .$totalScore.noColoc)
wilcox.test(colocPlot.df %>% filter(qtl_dataset == "microglia") %>% .$totalScore.noColoc,
            colocPlot.df %>% filter(qtl_dataset == "blood") %>% .$totalScore.noColoc)
wilcox.test(colocPlot.df %>% filter(qtl_dataset == "microglia") %>% .$totalScore.noColoc,
            colocPlot.df %>% filter(qtl_dataset == "LCL") %>% .$totalScore.noColoc)
wilcox.test(colocPlot.df %>% filter(qtl_dataset == "neutrophil") %>% .$totalScore.noColoc,
            colocPlot.df %>% filter(qtl_dataset == "other") %>% .$totalScore.noColoc)
wilcox.test(colocPlot.df %>% filter(qtl_dataset == "neutrophil") %>% .$totalScore.noColoc,
            colocPlot.df %>% filter(qtl_dataset == "blood") %>% .$totalScore.noColoc)
wilcox.test(colocPlot.df %>% filter(qtl_dataset == "neutrophil") %>% .$totalScore.noColoc,
            colocPlot.df %>% filter(qtl_dataset == "LCL") %>% .$totalScore.noColoc)


# Here we plot both genes with coloc H4 > 0.8 and H4 > 0.9 grouped separately in the same plot
colocPlot.df = bind_rows(
#  colocData.df %>% filter(H4 > 0.8) %>% mutate(colocSet = "H4 > 0.8"),
  colocData.df %>% filter(H4 > 0.9) %>% mutate(colocSet = "H4 > 0.9"),
  colocData.df %>% filter(H4 > 0.95) %>% mutate(colocSet = "H4 > 0.95"),
  colocData.df %>% filter(H4 > 0.98) %>% mutate(colocSet = "H4 > 0.98"))
ggplot(colocPlot.df, aes(x=fct_reorder(qtl_dataset, page.rank.pctile, .fun=median), y=page.rank.pctile, fill=colocSet, alpha=colocSet)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.25), alpha=0.5) + geom_boxplot() +
  theme_bw() + theme(axis.text.x = element_text(angle = 35, hjust = 1)) +
#  scale_alpha_manual(values=c("H4 > 0.8" = 0, "H4 > 0.9" = 0.5)) +
  ggtitle("Page rank pctile of colocalising genes by cell type group") +
  xlab("QTL study cell type")

ggplot(colocPlot.df, aes(x=fct_reorder(qtl_dataset, totalScore.noColoc, .fun=median), y=totalScore.noColoc, fill=colocSet)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.07), alpha=0.25) + geom_boxplot(alpha=0.6) +
  theme_bw() + theme(axis.text.x = element_text(angle = 35, hjust = 1)) +
#  scale_alpha_manual(values=c("H4 > 0.8" = 0, "H4 > 0.9" = 0.5)) +
  ggtitle("Total score (excl coloc) of colocalising genes by cell type group") +
  xlab("QTL study cell type")
```

In general, there isn't much that differentiates any of the cell type groups. That is, when there is a coloc, the page.rank.pctile is generally higher. But note that the number of colocs to compare is low for many of the groupings. If we compare individual groups against maxH4, the strongest difference we get is for neutrophil vs. maxH4 at coloc threshold > 0.8, but this goes away if we use a threshold of 0.9.

```{r, warning=FALSE, message=FALSE, echo=FALSE}

wilcox.test(colocNet.df %>% filter(qtl_dataset == "neutrophil", H4 > 0.8) %>% .$page.rank.pctile,
            colocNet.df %>% filter(qtl_dataset == "maxH4", H4 > 0.8) %>% .$page.rank.pctile,
            alternative = "greater")
wilcox.test(colocNet.df %>% filter(qtl_dataset == "neutrophil", H4 > 0.9) %>% .$page.rank.pctile,
            colocNet.df %>% filter(qtl_dataset == "maxH4", H4 > 0.9) %>% .$page.rank.pctile,
            alternative = "greater")

wilcox.test(colocNet.df %>% filter(qtl_dataset == "microglia", H4 > 0.8) %>% .$page.rank.pctile,
            colocNet.df %>% filter(qtl_dataset == "maxH4", H4 > 0.8) %>% .$page.rank.pctile,
            alternative = "greater")
wilcox.test(colocNet.df %>% filter(qtl_dataset == "microglia", H4 > 0.9) %>% .$page.rank.pctile,
            colocNet.df %>% filter(qtl_dataset == "maxH4", H4 > 0.9) %>% .$page.rank.pctile,
            alternative = "greater")

wilcox.test(colocNet.df %>% filter(qtl_dataset == "brain", H4 > 0.8) %>% .$page.rank.pctile,
            colocNet.df %>% filter(qtl_dataset == "maxH4", H4 > 0.8) %>% .$page.rank.pctile,
            alternative = "greater")
wilcox.test(colocNet.df %>% filter(qtl_dataset == "brain", H4 > 0.9) %>% .$page.rank.pctile,
            colocNet.df %>% filter(qtl_dataset == "maxH4", H4 > 0.9) %>% .$page.rank.pctile,
            alternative = "greater")

```


```{r, warning=FALSE, message=FALSE, echo=FALSE, eval=FALSE}
colocEnrich.df = colocNet.df %>%
  group_by(qtl_dataset) %>%
  mutate(H4 = if_else(is.na(H4), -1, H4)) %>%
  summarise(coloc0.8_networkHi = sum(H4 > 0.8 & page.rank.pctile > 50),
            coloc0.8_networkLo = sum(H4 > 0.8 & page.rank.pctile <= 50),
            colocLo_networkHi = sum(H4 >= 0 & H4 <= 0.8 & page.rank.pctile > 50),
            colocLo_networkLo = sum(H4 >= 0 & H4 <= 0.8 & page.rank.pctile <= 50),
            colocNa_networkHi = sum(H4 == -1 & page.rank.pctile > 50),
            colocNa_networkLo = sum(H4 == -1 & page.rank.pctile <= 50),
            colocLoNa_networkHi = sum(H4 <= 0.8 & page.rank.pctile > 50),
            colocLoNa_networkLo = sum(H4 <= 0.8 & page.rank.pctile <= 50))

colocEnrich.df = colocNet.df %>%
  group_by(qtl_dataset) %>%
  mutate(H4 = if_else(is.na(H4), -1, H4)) %>%
  summarise(coloc0.8_networkHi = sum(H4 > 0.5 & page.rank.pctile > 50),
            coloc0.8_networkLo = sum(H4 > 0.5 & page.rank.pctile <= 50),
            colocLo_networkHi = sum(H4 >= 0 & H4 <= 0.5 & page.rank.pctile > 50),
            colocLo_networkLo = sum(H4 >= 0 & H4 <= 0.5 & page.rank.pctile <= 50),
            colocNa_networkHi = sum(H4 == -1 & page.rank.pctile > 50),
            colocNa_networkLo = sum(H4 == -1 & page.rank.pctile <= 50),
            colocLoNa_networkHi = sum(H4 <= 0.5 & page.rank.pctile > 50),
            colocLoNa_networkLo = sum(H4 <= 0.5 & page.rank.pctile <= 50))

colocEnrich.df = colocNet.df %>%
  group_by(qtl_dataset) %>%
  mutate(H4 = if_else(is.na(H4), -1, H4)) %>%
  summarise(coloc0.8_networkHi = sum(H4 > 0.8 & page.rank.pctile > 80),
            coloc0.8_networkLo = sum(H4 > 0.8 & page.rank.pctile <= 50),
            colocLo_networkHi = sum(H4 >= 0 & H4 <= 0.8 & page.rank.pctile > 80),
            colocLo_networkLo = sum(H4 >= 0 & H4 <= 0.8 & page.rank.pctile <= 50),
            colocNa_networkHi = sum(H4 == -1 & page.rank.pctile > 80),
            colocNa_networkLo = sum(H4 == -1 & page.rank.pctile <= 50),
            colocLoNa_networkHi = sum(H4 <= 0.8 & page.rank.pctile > 80),
            colocLoNa_networkLo = sum(H4 <= 0.8 & page.rank.pctile <= 50))

colocEnrich.df = colocNet.df %>%
  group_by(qtl_dataset) %>%
  mutate(H4 = if_else(is.na(H4), -1, H4)) %>%
  summarise(coloc0.8_networkHi = sum(H4 > 0.5 & page.rank.pctile > 80),
            coloc0.8_networkLo = sum(H4 > 0.5 & page.rank.pctile <= 50),
            colocLo_networkHi = sum(H4 >= 0 & H4 <= 0.5 & page.rank.pctile > 80),
            colocLo_networkLo = sum(H4 >= 0 & H4 <= 0.5 & page.rank.pctile <= 50),
            colocNa_networkHi = sum(H4 == -1 & page.rank.pctile > 80),
            colocNa_networkLo = sum(H4 == -1 & page.rank.pctile <= 50),
            colocLoNa_networkHi = sum(H4 <= 0.5 & page.rank.pctile > 80),
            colocLoNa_networkLo = sum(H4 <= 0.5 & page.rank.pctile <= 50))
```



### Genome-wide enrichment of annotations for genes with low p-values nearby

Another way to determine enrichments is to look genome-wide. We only have gene expression and network scores for (nearly) all genes, and so we can't use locus-specific scores such as coloc or distance.

To do this, for each gene we consider the minimum p value among SNPs within 10 kb. We then look at different p-value thresholds and see what the enrichment is for high network scores at each threshold.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
network_excl.df = network_all.df %>%
  filter(chr != 19 | (abs(start - 45409011) > 1e6 & abs(end - 45409011) > 1e6),
         !is.na(minp), !is.na(page.rank.pctile))
# Ignore genes near APOE
ggplot(network_excl.df %>% filter(minp > 1e-8), aes(x=minp)) +
  geom_histogram(bins=100)
ggplot(network_excl.df %>% filter(minp > 1e-8), aes(x=-log10(minp))) +
  geom_histogram(bins=100)
sum(is.na(network_excl.df$minp))
sum(network_excl.df$minp < 5e-8, na.rm=T)
quantile(network_excl.df$minp, probs=seq(0, 1, 0.01), na.rm=T)

```


```{r, warning=FALSE, message=FALSE, echo=FALSE}
network_excl.df$page.rank.pctile.cat = categorize(network_excl.df$page.rank.pctile, thresholds = c(50, 60, 70, 80, 90, 95))
network_excl.df$page.rank.pctile.cat = factor(network_excl.df$page.rank.pctile.cat, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))

ggplot(network_excl.df, aes(x=page.rank.pctile.cat, y=minp.quantile)) +
  geom_boxplot() + geom_jitter(width=0.25, alpha=0.2)
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
bpy.colors = function (n = 100, cutoff.tails = 0.1, alpha = 1) {
    n <- as.integer(n[1])
    if (n <= 0)
        return(character(0))
	if (cutoff.tails >= 1 || cutoff.tails < 0)
		stop("cutoff.tails should be in [0, 1]")
	i = seq(0.5 * cutoff.tails, 1 - 0.5 * cutoff.tails, length = n)
    r = ifelse(i < .25, 0, ifelse(i < .57, i / .32 - .78125, 1))
    g = ifelse(i < .42, 0, ifelse(i < .92, 2 * i - .84, 1))
    b = ifelse(i < .25, 4 * i, ifelse(i < .42, 1,
        ifelse(i < .92, -2 * i + 1.84, i / .08 - 11.5)))
    rgb(r, g, b, alpha)
}
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
network_excl.df$minp_cat = categorize(network_excl.df$minp, thresholds = c(0.01, 1e-4, 1e-6, 1e-8))
minp_cat_levels = c("> 0.01", "1e-04-0.01", "1e-06-1e-04", "1e-08-1e-06", "<= 1e-08")
network_excl.df$minp_cat = factor(network_excl.df$minp_cat, levels = minp_cat_levels)

network_excl.df$page.rank.pctile.cat = categorize(network_excl.df$page.rank.pctile, thresholds = c(50, 60, 70, 80, 90))
page.rank.pct_cat_levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "> 90")
network_excl.df$page.rank.pctile.cat = factor(network_excl.df$page.rank.pctile.cat, levels = page.rank.pct_cat_levels)
xtabs(~ minp_cat + page.rank.pctile.cat, data = network_excl.df)
or.df = stratifiedOddsRatios(network_excl.df, "minp_cat", "page.rank.pctile.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
or.df$cat2 = factor(or.df$cat2, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "> 90"))
stratifiedOddsRatioPlot(or.df, "Minimum p value", "Pagerank\npercentile")

myCols = bpy.colors(8)
prPctColors = c("50-60"=myCols[2], "60-70"=myCols[3], "70-80"=myCols[4], "80-90"=myCols[5], "> 90"=myCols[6])

# Save a version for a figure
pdf(file.path(ad_dir, "genes/network_score.minp_enrichment.pdf"), width=3.1*1.5, height=2.5*1.5)
or.df = or.df %>%
  mutate(cat1 = gsub("1e-04-0.01", "0.01 - 1e-4", cat1),
         cat1 = gsub("1e-06-1e-04", "1e-4 - 1e-6", cat1),
         cat1 = gsub("1e-08-1e-06", "1e-6 - 1e-8", cat1),
         cat1 = gsub("<= 1e-08", "<= 1e-8", cat1)) %>%
  mutate(cat1 = factor(as.character(cat1), levels = c("0.01 - 1e-4", "1e-4 - 1e-6", "1e-6 - 1e-8", "<= 1e-8")))
stratifiedOddsRatioPlot(or.df, "Minimum p value", "Pagerank\npercentile") +
  coord_cartesian(ylim = c(0.5, 10)) +
  theme_classic(13) +
  theme(legend.justification = c(0, 1), legend.position = c(0.02, 1), legend.key.size = unit(0.4, "cm")) +
  scale_fill_manual(name = "Pagerank\npercentile", values = prPctColors)
dev.off()


xt = xtabs(~ minp_cat + page.rank.pctile.cat, data = network_excl.df)
or.df = stratifiedOddsRatios(network_excl.df, "minp_cat", "page.rank.pctile.cat")
# or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
# or.df$cat2 = factor(or.df$cat2, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-97", "> 97"))
# stratifiedOddsRatioPlot(or.df, "Min_p_value range", "Pagerank %ile")
# 
# Can we determine the "excess" number of high page.rank.pctile genes
# in low p-value categories? Can we assume that all of these excess are
# true AD causal genes that we aren't able to discover yet with current
# sample sizes?
# First add the counts of genes in each category, and compute the excess
# by assuming that there are NO causal genes in the category p > 0.01.
or.count.df = or.df %>% left_join(as.data.frame(xt), by=c("cat1" = "minp_cat", "cat2" = "page.rank.pctile.cat")) %>%
  mutate(excess = as.integer((estimate - 1) / estimate * Freq))
# The total excess in all categories with p < 0.01
sum(or.count.df %>% .$excess)
# The excess excluding the worst p value category 1e-04-0.01
sum(or.count.df %>% filter(cat1 != "1e-04-0.01") %>% .$excess)
```


```{r, warning=FALSE, message=FALSE, echo=FALSE}
wilcox.test(network_excl.df %>% filter(page.rank.pctile.cat == "> 90") %>% .$minp.quantile,
            network_excl.df %>% filter(page.rank.pctile.cat == "<= 50") %>% .$minp.quantile,
            alternative = "greater")
wilcox.test(network_excl.df %>% filter(page.rank.pctile.cat == "80-90") %>% .$minp.quantile,
            network_excl.df %>% filter(page.rank.pctile.cat == "<= 50") %>% .$minp.quantile,
            alternative = "greater")
wilcox.test(network_excl.df %>% filter(page.rank.pctile.cat == "70-80") %>% .$minp.quantile,
            network_excl.df %>% filter(page.rank.pctile.cat == "<= 50") %>% .$minp.quantile,
            alternative = "greater")
wilcox.test(network_excl.df %>% filter(page.rank.pctile.cat == "60-70") %>% .$minp.quantile,
            network_excl.df %>% filter(page.rank.pctile.cat == "<= 50") %>% .$minp.quantile,
            alternative = "greater")
wilcox.test(network_excl.df %>% filter(page.rank.pctile.cat == "50-60") %>% .$minp.quantile,
            network_excl.df %>% filter(page.rank.pctile.cat == "<= 50") %>% .$minp.quantile,
            alternative = "greater")
```

Now do something similar for expression.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
vecQuantiles = function(x) {
  sapply(x, function(val) if (is.na(val)) { NA } else { (sum(x <= val, na.rm = T) - 1) / (sum(!is.na(x)) - 1) })
}

genes.minp.excl.df = genes.minp.df %>%
  filter(chr != 19 | (abs(gene_window_start - 45409011) > 1e6 & abs(gene_window_end - 45409011) > 1e6),
         !is.na(minp)) %>%
  mutate(minp.quantile = vecQuantiles(-minp))

# Ignore genes near APOE
ggplot(genes.minp.excl.df %>% filter(minp > 1e-8), aes(x=minp)) +
  geom_histogram(bins=100)
ggplot(genes.minp.excl.df %>% filter(minp > 1e-8), aes(x=-log10(minp))) +
  geom_histogram(bins=100)
sum(is.na(genes.minp.excl.df$minp))
sum(genes.minp.excl.df$minp < 5e-8, na.rm=T)
quantile(genes.minp.excl.df$minp, probs=seq(0, 1, 0.01), na.rm=T)
```


```{r, warning=FALSE, message=FALSE, echo=FALSE}
# Annotate expression to the genes that we have minp values for
expr.minp.df = genes.minp.excl.df %>%
  left_join(bulk_expr.rel.df %>% select(gene_id, symbol, primary_microglia, Brain_Hippocampus, Brain_Cortex))


bulk_expr.tpm.mat = bulk_expr.tpm.df %>% select(-gene_id, -symbol) %>% as.matrix()
bulk_expr.tpm.df$maxExpr = apply(bulk_expr.tpm.mat, MARGIN = 1, function(x) max(x, na.rm = T))
expressedGenes = bulk_expr.tpm.df %>% filter(maxExpr > 1) %>% .$gene_id

sum(expr.minp.df$gene_id %in% expressedGenes)
expr.minp.df = expr.minp.df %>% filter(gene_id %in% expressedGenes)

expr.minp.df$primary_microglia.cat = categorize(expr.minp.df$primary_microglia, thresholds = c(50, 60, 70, 80, 90, 95))
expr.minp.df$primary_microglia.cat = factor(expr.minp.df$primary_microglia.cat, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))

ggplot(expr.minp.df, aes(x=primary_microglia.cat, y=minp.quantile)) +
  geom_boxplot() + geom_jitter(width=0.25, alpha=0.2) +
  ggtitle("Primary microglia expression percentiles")

expr.minp.df$Brain_Hippocampus.cat = categorize(expr.minp.df$Brain_Hippocampus, thresholds = c(50, 60, 70, 80, 90, 95))
expr.minp.df$Brain_Hippocampus.cat = factor(expr.minp.df$Brain_Hippocampus.cat, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))

ggplot(expr.minp.df, aes(x=Brain_Hippocampus.cat, y=minp.quantile)) +
  geom_boxplot() + geom_jitter(width=0.25, alpha=0.2) +
  ggtitle("Brain Hippocampus expression percentiles")
```


```{r, warning=FALSE, message=FALSE, echo=FALSE}
expr.minp.df$minp_cat = categorize(expr.minp.df$minp, thresholds = c(0.01, 1e-4, 1e-6, 1e-8))
minp_cat_levels = c("> 0.01", "1e-04-0.01", "1e-06-1e-04", "1e-08-1e-06", "<= 1e-08")
expr.minp.df$minp_cat = factor(expr.minp.df$minp_cat, levels = minp_cat_levels)

xtabs(~ minp_cat + primary_microglia.cat, data = expr.minp.df)
or.df = stratifiedOddsRatios(expr.minp.df, "minp_cat", "primary_microglia.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
or.df$cat2 = factor(or.df$cat2, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))
stratifiedOddsRatioPlot(or.df, "Min_p_value range", "Primary microglia %ile")

xtabs(~ minp_cat + Brain_Hippocampus.cat, data = expr.minp.df)
or.df = stratifiedOddsRatios(expr.minp.df, "minp_cat", "Brain_Hippocampus.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
or.df$cat2 = factor(or.df$cat2, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))
stratifiedOddsRatioPlot(or.df, "Min_p_value range", "Brain Hippocampus %ile")

```


```{r, warning=FALSE, message=FALSE, echo=FALSE}
expr.minp.df = genes.minp.excl.df %>%
  left_join(sc_brain_expr.rel.df %>% select(gene_id, symbol, Microglia, Astrocyte, Oligodendrocyte, Endothelial))

sc_brain_expr.tpm.mat = sc_brain_expr.tpm.df %>% select(-gene_id, -hgnc_symbol) %>% as.matrix()
sc_brain_expr.tpm.df$maxExpr = apply(sc_brain_expr.tpm.mat, MARGIN = 1, function(x) max(x, na.rm = T))
expressedGenes = sc_brain_expr.tpm.df %>% filter(maxExpr > 1) %>% .$gene_id

sum(expr.minp.df$gene_id %in% expressedGenes)
expr.minp.df = expr.minp.df %>% filter(gene_id %in% expressedGenes)


expr.minp.df$Microglia.cat = categorize(expr.minp.df$Microglia, thresholds = c(50, 60, 70, 80, 90, 95))
expr.minp.df$Microglia.cat = factor(expr.minp.df$Microglia.cat, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))

ggplot(expr.minp.df, aes(x=Microglia.cat, y=minp.quantile)) +
  geom_boxplot() + geom_jitter(width=0.25, alpha=0.2) +
  ggtitle("Brain single-cell microglia expression percentiles")

expr.minp.df$Astrocyte.cat = categorize(expr.minp.df$Astrocyte, thresholds = c(50, 60, 70, 80, 90, 95))
expr.minp.df$Astrocyte.cat = factor(expr.minp.df$Astrocyte.cat, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))

ggplot(expr.minp.df, aes(x=Astrocyte.cat, y=minp.quantile)) +
  geom_boxplot() + geom_jitter(width=0.25, alpha=0.2) +
  ggtitle("Brain single-cell astrocyte expression percentiles")

expr.minp.df$Oligodendrocyte.cat = categorize(expr.minp.df$Oligodendrocyte, thresholds = c(50, 60, 70, 80, 90, 95))
expr.minp.df$Oligodendrocyte.cat = factor(expr.minp.df$Oligodendrocyte.cat, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))

ggplot(expr.minp.df, aes(x=Oligodendrocyte.cat, y=minp.quantile)) +
  geom_boxplot() + geom_jitter(width=0.25, alpha=0.2) +
  ggtitle("Brain single-cell oligodendrocyte expression percentiles")

expr.minp.df$Endothelial.cat = categorize(expr.minp.df$Endothelial, thresholds = c(50, 60, 70, 80, 90, 95))
expr.minp.df$Endothelial.cat = factor(expr.minp.df$Endothelial.cat, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))

ggplot(expr.minp.df, aes(x=Endothelial.cat, y=minp.quantile)) +
  geom_boxplot() + geom_jitter(width=0.25, alpha=0.2) +
  ggtitle("Brain single-cell endothelial expression percentiles")


```


```{r, warning=FALSE, message=FALSE, echo=FALSE}
expr.minp.df$minp_cat = categorize(expr.minp.df$minp, thresholds = c(0.01, 1e-4, 1e-6, 1e-8))
minp_cat_levels = c("> 0.01", "1e-04-0.01", "1e-06-1e-04", "1e-08-1e-06", "<= 1e-08")
expr.minp.df$minp_cat = factor(expr.minp.df$minp_cat, levels = minp_cat_levels)

xtabs(~ minp_cat + Microglia.cat, data = expr.minp.df)
or.df = stratifiedOddsRatios(expr.minp.df, "minp_cat", "Microglia.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
or.df$cat2 = factor(or.df$cat2, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))
stratifiedOddsRatioPlot(or.df, "Min_p_value range", "Brain single-cell microglia %ile")

xtabs(~ minp_cat + Astrocyte.cat, data = expr.minp.df)
or.df = stratifiedOddsRatios(expr.minp.df, "minp_cat", "Astrocyte.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
or.df$cat2 = factor(or.df$cat2, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))
stratifiedOddsRatioPlot(or.df, "Min_p_value range", "Brain single-cell astrocyte %ile")

xtabs(~ minp_cat + Oligodendrocyte.cat, data = expr.minp.df)
or.df = stratifiedOddsRatios(expr.minp.df, "minp_cat", "Oligodendrocyte.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
or.df$cat2 = factor(or.df$cat2, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))
stratifiedOddsRatioPlot(or.df, "Min_p_value range", "Brain single-cell oligodendrocyte %ile")

xtabs(~ minp_cat + Endothelial.cat, data = expr.minp.df)
or.df = stratifiedOddsRatios(expr.minp.df, "minp_cat", "Endothelial.cat")
or.df$cat1 = factor(or.df$cat1, levels = minp_cat_levels)
or.df$cat2 = factor(or.df$cat2, levels = c("<= 50", "50-60", "60-70", "70-80", "80-90", "90-95", "> 95"))
stratifiedOddsRatioPlot(or.df, "Min_p_value range", "Brain single-cell endothelial %ile")
```

## Expression specificity vs. level

Here we look at whether genes with top scores (excluding expression) are more highly enriched in genes with high expression specificity or high expression level in different cell/tissue types.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.width=7, fig.height=5}
candidate_genes = readr::read_tsv(file.path(ad_dir, "candidate_genes.tsv"), col_names = c("symbol"))$symbol

mergenet.expr.df = mergenet.df %>%
  mutate(is_candidate = symbol %in% candidate_genes) %>%
  rowwise() %>%
  mutate(label = if_else(is_candidate, symbol, "")) %>%
  mutate(colocScore = colocMaxH4,
         primaryMicrogliaScore = max(0, bulk_primary_microglia - 50) * 2,
         scMicrogliaScore = max(0, sc_Microglia - 50) * 2,
         scAstrocyteScore = max(0, sc_Astrocyte - 50) * 2,
         bulkHippocampusScore = max(0, bulk_Brain_Hippocampus - 50) * 2,
         bulkBrainCortexScore = max(0, bulk_Brain_Cortex - 50) * 2,
         primaryMicrogliaScore2 = max(0, bulk_primary_microglia - 70) * 2,
         scMicrogliaScore2 = max(0, sc_Microglia - 70) * 2,
         scAstrocyteScore2 = max(0, sc_Astrocyte - 70) * 2,
         bulkHippocampusScore2 = max(0, bulk_Brain_Hippocampus - 70) * 2,
         bulkBrainCortexScore2 = max(0, bulk_Brain_Cortex - 70) * 2,
         totalScore.noExpr = geneDistScore + colocScore + networkScore + codingScore,
         totalScore.noNetwork = geneDistScore + colocScore + codingScore + exprScore,
         totalScore.noColoc = geneDistScore + networkScore + codingScore + exprScore,
         totalScore.noDist = colocScore + networkScore + codingScore + exprScore)

p1 = ggplot(mergenet.expr.df, aes(x=totalScore.noNetwork, y=networkScore, col=is_candidate)) +
  geom_point(aes(alpha = is_candidate)) +
  geom_text(mapping = aes(label = label), size=2.2, nudge_x = 0.05, hjust = 0) +
  scale_alpha_manual(values = c("TRUE"=0.75, "FALSE"=0.5), guide=F) +
  scale_color_manual(values = c("TRUE"="dodgerblue", "FALSE"="black"), guide=F) +
  ggtitle("NetworkScore vs. Non-network score")
ggMarginal(p1, margins = 'both', type="density", size=10)

cor.test(mergenet.expr.df$primaryMicrogliaScore, mergenet.expr.df$totalScore.noExpr, method = "spearman")
cor.test(mergenet.expr.df$scMicrogliaScore, mergenet.expr.df$totalScore.noExpr, method = "spearman")
cor.test(mergenet.expr.df$scAstrocyteScore, mergenet.expr.df$totalScore.noExpr, method = "spearman")
cor.test(mergenet.expr.df$bulkHippocampusScore, mergenet.expr.df$totalScore.noExpr, method = "spearman")
cor.test(mergenet.expr.df$bulkBrainCortexScore, mergenet.expr.df$totalScore.noExpr, method = "spearman")

cor.test(mergenet.expr.df$primaryMicrogliaScore2, mergenet.expr.df$totalScore.noExpr, method = "spearman")
cor.test(mergenet.expr.df$scMicrogliaScore2, mergenet.expr.df$totalScore.noExpr, method = "spearman")
cor.test(mergenet.expr.df$scAstrocyteScore2, mergenet.expr.df$totalScore.noExpr, method = "spearman")
cor.test(mergenet.expr.df$bulkHippocampusScore2, mergenet.expr.df$totalScore.noExpr, method = "spearman")
cor.test(mergenet.expr.df$bulkBrainCortexScore2, mergenet.expr.df$totalScore.noExpr, method = "spearman")


bulk_expr.tpm.norm.mat = bulk_expr.tpm.df %>% select(-gene_id, -symbol) %>% as.matrix()
bulk_expr.tpm.norm.mat[is.na(bulk_expr.tpm.norm.mat)] = 0
N = nrow(bulk_expr.tpm.norm.mat)
bulk_expr.tpm.norm.mat = apply(bulk_expr.tpm.norm.mat, MARGIN=2, FUN=function(x) rank(x, ties.method="average") / N) * 100
bulk_expr.tpm.norm.df = cbind(bulk_expr.tpm.df %>% select(gene_id, symbol), bulk_expr.tpm.norm.mat)

mergenet.norm_expr.df = mergenet.expr.df %>%
  select(leadSNP:networkScore, is_candidate, label, starts_with("total")) %>%
  left_join(bulk_expr.tpm.norm.df %>% select(-symbol), by="gene_id") %>%
  rowwise() %>%
  mutate(colocScore = colocMaxH4,
         primaryMicrogliaScore = max(0, primary_microglia - 50) * 2,
         Brain_HippocampusScore = max(0, Brain_Hippocampus - 50) * 2,
         Brain_CortexScore = max(0, Brain_Cortex - 50) * 2,
         primaryMicrogliaScore2 = max(0, primary_microglia - 70) * 2,
         Brain_HippocampusScore2 = max(0, Brain_Hippocampus - 70) * 2,
         Brain_CortexScore2 = max(0, Brain_Cortex - 70) * 2)

cor.test(mergenet.norm_expr.df$primaryMicrogliaScore, mergenet.norm_expr.df$totalScore.noExpr, method = "spearman")
cor.test(mergenet.norm_expr.df$Brain_HippocampusScore, mergenet.norm_expr.df$totalScore.noExpr, method = "spearman")
cor.test(mergenet.norm_expr.df$Brain_CortexScore, mergenet.norm_expr.df$totalScore.noExpr, method = "spearman")

topN = 40
merged.top40PrimaryMicrogliaExprSpecificity = mergenet.expr.df %>%
  arrange(desc(primaryMicrogliaScore)) %>%
  .[1:topN,] %>%
  mutate(scoreGroup = "Top 40 primary microglia expr specificity", scoreType = "specificity")

merged.top40HippocampusExprSpecificity = mergenet.expr.df %>%
  arrange(desc(bulkHippocampusScore)) %>%
  .[1:topN,] %>%
  mutate(scoreGroup = "Top 40 brain hippocampus expr specificity", scoreType = "specificity")

merged.top40BrainCortexExprSpecificity = mergenet.expr.df %>%
  arrange(desc(bulkBrainCortexScore)) %>%
  .[1:topN,] %>%
  mutate(scoreGroup = "Top 40 brain cortex expr specificity", scoreType = "specificity")

merged.top40ScMicrogliaExprSpecificity = mergenet.expr.df %>%
  arrange(desc(scMicrogliaScore)) %>%
  .[1:topN,] %>%
  mutate(scoreGroup = "Top 40 sc microglia expr specificity", scoreType = "specificity")

merged.top40ScAstrocyteExprSpecificity = mergenet.expr.df %>%
  arrange(desc(scAstrocyteScore)) %>%
  .[1:topN,] %>%
  mutate(scoreGroup = "Top 40 sc astrocyte expr specificity", scoreType = "specificity")


merged.top40PrimaryMicrogliaExprLevel = mergenet.norm_expr.df %>%
  arrange(desc(primaryMicrogliaScore)) %>%
  .[1:topN,] %>%
  mutate(scoreGroup = "Top 40 primary microglia expr level", scoreType = "level")

merged.top40HippocampusExprLevel = mergenet.norm_expr.df %>%
  arrange(desc(Brain_HippocampusScore)) %>%
  .[1:topN,] %>%
  mutate(scoreGroup = "Top 40 brain hippocampus expr level", scoreType = "level")

merged.top40BrainCortexExprLevel = mergenet.norm_expr.df %>%
  arrange(desc(Brain_CortexScore)) %>%
  .[1:topN,] %>%
  mutate(scoreGroup = "Top 40 brain cortex expr level", scoreType = "level")


merged.df4 = bind_rows(
  mergenet.expr.df %>% mutate(scoreGroup = "All genes"),
  merged.top40PrimaryMicrogliaExprSpecificity,
  merged.top40PrimaryMicrogliaExprLevel,
  merged.top40HippocampusExprSpecificity,
  merged.top40HippocampusExprLevel,
  merged.top40BrainCortexExprSpecificity,
  merged.top40BrainCortexExprLevel
)

library(egg) # needed for theme_article()

ggplot(merged.df4, aes(x=scoreGroup, y=totalScore.noExpr, fill=scoreType)) +
  geom_violin(alpha = 0.9) + geom_jitter(width = 0.25, alpha = 0.4) + geom_boxplot(width = 0.05, outlier.shape = NA, fill="white") +
  ggtitle("Scores for top genes in different groups") +
  theme_classic(13) + theme(axis.text.x = element_text(angle = 25, hjust = 1), axis.title.x = element_blank()) +
  scale_fill_discrete(guide=F) + ylab("Total score (without expression)")

wilcox.test(merged.top40PrimaryMicrogliaExprSpecificity$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40ScMicrogliaExprSpecificity$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40HippocampusExprSpecificity$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40BrainCortexExprSpecificity$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40ScAstrocyteExprSpecificity$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)

wilcox.test(merged.top40PrimaryMicrogliaExprLevel$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40HippocampusExprLevel$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40BrainCortexExprLevel$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)

expr.plot.df = mergenet.expr.df %>%
  arrange(desc(totalScore.noExpr)) %>%
  select(gene_id, primaryMicrogliaScore, scMicrogliaScore, bulkHippocampusScore, bulkBrainCortexScore, scAstrocyteScore) %>%
  .[1:topN,] %>%
  gather(key = "Expression_dataset", value = "expr", -gene_id)

ggplot(expr.plot.df, aes(x=Expression_dataset, y=expr, fill=Expression_dataset)) +
  geom_violin(alpha = 0.9) + geom_jitter(width = 0.25, alpha = 0.4) + geom_boxplot(width = 0.05, outlier.shape = NA, fill="white") +
  ggtitle("Expression scores for top genes by total score") +
  theme_classic(14) + theme(axis.text.x = element_text(angle = 30, hjust = 1), axis.title.x = element_blank()) +
  scale_fill_discrete(guide=F) + ylab("Expression score")

```

```{r, warning=FALSE, message=FALSE, echo=FALSE, eval=FALSE}
# This section was checking whether we get a clearer separation if, instead of
# taking the top N genes by expression (specificity / level), we take the top N
# by expression rank within each locus. It is largely comparable.

mergenet.expr.df2 = mergenet.expr.df %>%
  group_by(locus) %>%
  mutate(primaryMicrogliaRank = rank(-primaryMicrogliaScore),
         hippocampusRank = rank(-bulkHippocampusScore),
         brainCortexRank = rank(-bulkBrainCortexScore),
         scMicrogliaRank = rank(-scMicrogliaScore),
         astrocyteRank = rank(-scAstrocyteScore))

merged.top40PrimaryMicrogliaExprSpecificity = mergenet.expr.df2 %>%
  arrange(primaryMicrogliaRank) %>%
  .[1:40,] %>%
  mutate(scoreGroup = "Top 40 primary microglia expr specificity", scoreType = "specificity")

merged.top40HippocampusExprSpecificity = mergenet.expr.df2 %>%
  arrange(hippocampusRank) %>%
  .[1:40,] %>%
  mutate(scoreGroup = "Top 40 brain hippocampus expr specificity", scoreType = "specificity")

merged.top40BrainCortexExprSpecificity = mergenet.expr.df2 %>%
  arrange(brainCortexRank) %>%
  .[1:40,] %>%
  mutate(scoreGroup = "Top 40 brain cortex expr specificity", scoreType = "specificity")

merged.top40ScMicrogliaExprSpecificity = mergenet.expr.df2 %>%
  arrange(scMicrogliaRank) %>%
  .[1:40,] %>%
  mutate(scoreGroup = "Top 40 sc microglia expr specificity", scoreType = "specificity")

merged.top40ScAstrocyteExprSpecificity = mergenet.expr.df2 %>%
  arrange(astrocyteRank) %>%
  .[1:40,] %>%
  mutate(scoreGroup = "Top 40 sc astrocyte expr specificity", scoreType = "specificity")


merged.top40PrimaryMicrogliaExprLevel = mergenet.norm_expr.df %>%
  arrange(desc(primaryMicrogliaScore)) %>%
  .[1:40,] %>%
  mutate(scoreGroup = "Top 40 primary microglia expr level", scoreType = "level")

merged.top40HippocampusExprLevel = mergenet.norm_expr.df %>%
  arrange(desc(Brain_HippocampusScore)) %>%
  .[1:40,] %>%
  mutate(scoreGroup = "Top 40 brain hippocampus expr level", scoreType = "level")

merged.top40BrainCortexExprLevel = mergenet.norm_expr.df %>%
  arrange(desc(Brain_CortexScore)) %>%
  .[1:40,] %>%
  mutate(scoreGroup = "Top 40 brain cortex expr level", scoreType = "level")


merged.df4 = bind_rows(
  mergenet.expr.df %>% mutate(scoreGroup = "All genes"),
  merged.top40PrimaryMicrogliaExprSpecificity,
  merged.top40HippocampusExprSpecificity,
  merged.top40BrainCortexExprSpecificity,
  merged.top40ScMicrogliaExprSpecificity,
  merged.top40ScAstrocyteExprSpecificity
)

library(egg) # needed for theme_article()

ggplot(merged.df4, aes(x=scoreGroup, y=totalScore.noExpr, fill=scoreType)) +
  geom_violin(alpha = 0.9) + geom_jitter(width = 0.25, alpha = 0.4) + geom_boxplot(width = 0.05, outlier.shape = NA, fill="white") +
  ggtitle("Scores for top genes in different groups") +
  theme_classic(14) + theme(axis.text.x = element_text(angle = 20, hjust = 1), axis.title.x = element_blank()) +
  scale_fill_discrete(guide=F) + ylab("Total score (without expression)")

wilcox.test(merged.top40PrimaryMicrogliaExprSpecificity$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40ScMicrogliaExprSpecificity$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40HippocampusExprSpecificity$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40BrainCortexExprSpecificity$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40ScAstrocyteExprSpecificity$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)

wilcox.test(merged.top40PrimaryMicrogliaExprLevel$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40HippocampusExprLevel$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)
wilcox.test(merged.top40BrainCortexExprLevel$totalScore.noExpr, mergenet.expr.df$totalScore.noExpr)

expr.plot.df = mergenet.expr.df %>%
  arrange(desc(totalScore.noExpr)) %>%
  select(gene_id, primaryMicrogliaScore, scMicrogliaScore, bulkHippocampusScore, bulkBrainCortexScore, scAstrocyteScore) %>%
  .[1:40,] %>%
  gather(key = "Expression_dataset", value = "expr", -gene_id)

ggplot(expr.plot.df, aes(x=Expression_dataset, y=expr, fill=Expression_dataset)) +
  geom_violin(alpha = 0.9) + geom_jitter(width = 0.25, alpha = 0.4) + geom_boxplot(width = 0.05, outlier.shape = NA, fill="white") +
  ggtitle("Expression scores for top genes by total score") +
  theme_classic(14) + theme(axis.text.x = element_text(angle = 20, hjust = 1), axis.title.x = element_blank()) +
  scale_fill_discrete(guide=F) + ylab("Expression score")

```


## Network enrichment plots

Make some plots of network scores for supplementary figures.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
seed_gene_loci = merged.df %>% filter(seed_gene) %>% pull(locus)
merged.seed_loci.df = merged.df %>% filter(locus %in% seed_gene_loci)

p.1mb = ggplot(merged.seed_loci.df %>% filter(!seed_gene), aes(x=locus, y=page.rank.pctile, col=seed_gene, size=seed_gene)) +
  geom_point(alpha=0.5) +
  geom_point(data=merged.seed_loci.df %>% filter(seed_gene), alpha=0.7) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("PageRank percentile per locus (1 Mb window)") +
  ylab("Pagerank percentile") +
  scale_color_manual(name = "Seed gene", values = c("FALSE"="grey20", "TRUE"="blue")) +
  scale_size_manual(name = "Seed gene", values=c("TRUE"=2.5, "FALSE"=1.5))
p.1mb

p.200kb = ggplot(merged.seed_loci.df %>% filter(!seed_gene, abs(dist) < 1e5), aes(x=locus, y=page.rank.pctile, col=seed_gene, size=seed_gene)) +
  geom_point(alpha=0.5) +
  geom_point(data=merged.seed_loci.df %>% filter(seed_gene), alpha=0.7) +
  theme_bw() + theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("PageRank percentile per locus (200 kb window)") +
  ylab("Pagerank percentile") +
  scale_color_manual(name = "Seed gene", values = c("FALSE"="grey20", "TRUE"="blue")) +
  scale_size_manual(name = "Seed gene", values = c("TRUE"=2.5, "FALSE"=1.5))
p.200kb

pdf(file = paste0(output_root, "network.seed_gene_pagerank_pctile.1mb.pdf"), width = 6*1.4, height = 2.9*1.4)
p.1mb
dev.off()

p.violin = ggplot(merged.df %>% filter(!is.na(page.rank.pctile)), aes(x=seed_gene, fill=seed_gene, y=page.rank.pctile)) +
  geom_violin() +
  geom_jitter(width = 0.1, alpha = 0.5) +
  geom_boxplot(width = 0.05, fill="white", alpha = 0.8) +
  theme_classic(14) +
  xlab("Seed gene") +
  ylab("Pagerank percentile") +
  scale_fill_discrete(name = "Seed gene", guide = F)
p.violin

pdf(file = paste0(output_root, "network.seed_gene.violin.pdf"), width = 3, height = 3.5)
p.violin
dev.off()

```

Which is the top network gene for each locus?

```{r, warning=FALSE, message=FALSE, echo=FALSE}
# Note that AC113554.1 is the top gene at the ACE locus, but this is
# a transcript overalpping ACE.
merged.df %>% arrange(desc(page.rank.pctile)) %>% filter(biotype == "protein_coding") %>% filter(!duplicated(locus)) %>% select(locus, symbol, page.rank.pctile, totalScore)
#View(merged.df %>% arrange(desc(page.rank.pctile)) %>% filter(biotype == "protein_coding") %>% filter(!duplicated(locus)) %>% select(locus, symbol, totalScore, page.rank.pctile, dist, everything()))
#View(merged.df %>% arrange(desc(page.rank.pctile)) %>% select(locus, symbol, totalScore, page.rank.pctile, dist, everything()))
#View(network_all.df %>% filter(geneSymbol %in% c("ACE", "AC113554.1")))
```

Test the significance of seed genes having higher PR pctile than other genes.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
wilcox.test(merged.df %>% filter(seed_gene) %>% pull(page.rank.pctile), merged.df %>% filter(!seed_gene) %>% pull(page.rank.pctile), alternative = "greater")
# The below compares all seed genes, not just at AD loci, and compares
# against all genes, not just those at AD loci.
#wilcox.test(network_all.df %>% filter(seed_gene) %>% pull(page.rank.pctile), network_all.df %>% filter(!seed_gene) %>% pull(page.rank.pctile))

wilcox.test(network_all.df %>% arrange(desc(page.rank.pctile)) %>% .[1:1000,] %>% .$minp,
            network_all.df %>% arrange(desc(page.rank.pctile)) %>% .[-1:-1000,] %>% .$minp)
```


## Glmnet models

Let's first to simple GLM logistic regression models.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(glmnet)

#merged.df = getGeneScores(geneExpr.df, useMaxH4 = T)
merged.df = merged.df %>% ungroup() %>%
  mutate(is_candidate = symbol %in% candidate_genes)

#sapply(merged.df, FUN=function(x) sum(is.na(x)))
#sapply(merged.na_zero.df, FUN=function(x) sum(is.na(x)))

merged.df$gene_dist_cat = categorize(abs(merged.df$dist), thresholds = c(10000, 100000))
merged.df$gene_dist_cat = factor(as.character(merged.df$gene_dist_cat), levels = c("> 1e+05", "10000-1e+05", "<= 10000"))

merged.df$coloc_cat = categorize(abs(merged.df$colocMaxH4), thresholds = c(0.8, 0.95))
merged.df$coloc_cat = factor(merged.df$coloc_cat, levels = c("<= 0.8", "0.8-0.95", "> 0.95"))

merged.df$page.rank.pctile_cat = categorize(merged.df$page.rank.pctile, thresholds = c(50, 80, 90))
pr_pctile_cat_levels = c("<= 50", "50-80", "80-90", "> 90")
merged.df$page.rank.pctile_cat = factor(merged.df$page.rank.pctile_cat, levels = pr_pctile_cat_levels)

merged.df$gene_dist_cat_fine = categorize(abs(merged.df$dist), thresholds = c(0, 10000, 100000, 200000))
gene_dist_cat_fine_levels = c("> 2e+05", "1e+05-2e+05", "10000-1e+05", "0-10000", "<= 0")
merged.df$gene_dist_cat_fine = factor(merged.df$gene_dist_cat_fine, levels = gene_dist_cat_fine_levels)
xtabs(~ gene_dist_cat_fine + page.rank.pctile_cat, data = merged.df)

merged.df$coloc_cat_fine = categorize(merged.df$colocMaxH4, thresholds = c(0.5, 0.8, 0.9, 0.95))
coloc_cat_fine_levels = c("<= 0.5", "0.5-0.8", "0.8-0.9", "0.9-0.95", "> 0.95")
merged.df$coloc_cat_fine = factor(merged.df$coloc_cat_fine, levels = coloc_cat_fine_levels)

merged.df$sc_microglia_cat = categorize(abs(merged.df$sc_Microglia), thresholds = c(50, 80, 90))
sc_expr_cat_levels = c("<= 50", "50-80", "80-90", "> 90")
merged.df$sc_microglia_cat = factor(merged.df$sc_microglia_cat, levels = sc_expr_cat_levels)

merged.df$sc_astrocyte_cat = categorize(abs(merged.df$sc_Astrocyte), thresholds = c(50, 80, 90))
merged.df$sc_astrocyte_cat = factor(merged.df$sc_astrocyte_cat, levels = sc_expr_cat_levels)

merged.df$bulk_microglia_cat = categorize(abs(merged.df$bulk_primary_microglia), thresholds = c(50, 80, 90))
merged.df$bulk_microglia_cat = factor(merged.df$bulk_microglia_cat, levels = sc_expr_cat_levels)

merged.df$bulk_Brain_Hippocampus_cat = categorize(abs(merged.df$bulk_Brain_Hippocampus), thresholds = c(50, 80, 90))
brain_expr_cat_levels = c("<= 50", "50-80", "80-90", "> 90")
merged.df$bulk_Brain_Hippocampus_cat = factor(merged.df$bulk_Brain_Hippocampus_cat, levels = brain_expr_cat_levels)

merged.df$coding_cat = categorize(abs(merged.df$codingScore), thresholds = c(0.05))
coding_cat_levels = c("<= 0.05", "> 0.05")
merged.df$coding_cat = factor(merged.df$coding_cat, levels = coding_cat_levels)


merged.ml.df = merged.df %>% group_by(gene_id) %>%
  mutate_at(vars(starts_with("sc_"), -sc_microglia_cat, -sc_astrocyte_cat), .funs = function(x) if_else(is.na(x), 0, x)) %>%
  ungroup()


mergenet.ml.df = merged.ml.df %>% filter(!is.na(page.rank.pctile), locus != "APOE") %>%
  na.omit()

mergenet.near_vs_far = mergenet.ml.df %>% filter(gene_dist_cat != "10000-1e+05")

mergenet.100k.dist.df = mergenet.ml.df %>% filter(abs(dist) < 100000)
mergenet.100k.dist.df$gene_dist_cat = categorize(abs(mergenet.100k.dist.df$dist), thresholds = c(10000))
mergenet.100k.dist.df$gene_dist_cat = factor(as.character(mergenet.100k.dist.df$gene_dist_cat), levels = c("> 10000", "<= 10000"))

mergenet.network_hi_vs_lo = mergenet.ml.df
mergenet.network_hi_vs_lo$page.rank.pctile_cat = categorize(mergenet.network_hi_vs_lo$page.rank.pctile, thresholds = c(50, 80))
mergenet.network_hi_vs_lo = mergenet.network_hi_vs_lo %>% filter(page.rank.pctile_cat != "50-80")
pr_pctile_cat_levels = c("<= 50", "> 80")
mergenet.network_hi_vs_lo$page.rank.pctile_cat = factor(mergenet.network_hi_vs_lo$page.rank.pctile_cat, levels = pr_pctile_cat_levels)

xtabs(~ gene_dist_cat + coloc_cat, data = mergenet.near_vs_far)
xtabs(~ gene_dist_cat + coloc_cat_fine, data = mergenet.near_vs_far)
xtabs(~ gene_dist_cat + coding_cat, data = mergenet.near_vs_far)
xtabs(~ gene_dist_cat + sc_microglia_cat, data = mergenet.near_vs_far)
xtabs(~ gene_dist_cat + sc_astrocyte_cat, data = mergenet.near_vs_far)
xtabs(~ gene_dist_cat + bulk_microglia_cat, data = mergenet.near_vs_far)
xtabs(~ gene_dist_cat + bulk_Brain_Hippocampus_cat, data = mergenet.near_vs_far)
xtabs(~ gene_dist_cat + page.rank.pctile_cat, data = mergenet.near_vs_far)

xtabs(~ page.rank.pctile_cat + coloc_cat, data = mergenet.network_hi_vs_lo)
xtabs(~ page.rank.pctile_cat + coloc_cat_fine, data = mergenet.network_hi_vs_lo)
xtabs(~ page.rank.pctile_cat + coding_cat, data = mergenet.network_hi_vs_lo)
xtabs(~ page.rank.pctile_cat + sc_microglia_cat, data = mergenet.network_hi_vs_lo)
xtabs(~ page.rank.pctile_cat + sc_astrocyte_cat, data = mergenet.network_hi_vs_lo)
xtabs(~ page.rank.pctile_cat + bulk_microglia_cat, data = mergenet.network_hi_vs_lo)
xtabs(~ page.rank.pctile_cat + bulk_Brain_Hippocampus_cat, data = mergenet.network_hi_vs_lo)
xtabs(~ page.rank.pctile_cat + gene_dist_cat, data = mergenet.network_hi_vs_lo)
```

First we do basic glm logistic regression, and compare using discrete predictor categories to using quantitative scores.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
set.seed(123)
mergenet.near_vs_far = mergenet.near_vs_far %>% ungroup()
mergenet.glm.df = mergenet.near_vs_far %>%
  select(gene_dist_cat, colocScore, networkScore, codingScore, exprScore,
         colocMaxH4, coloc_cat, coloc_cat_fine,
         codingScore, coding_cat,
         page.rank.pctile, page.rank.pctile_cat,
         starts_with("sc_"), starts_with("bulk_")) %>%
  filter(gene_dist_cat != "10000-1e+05")
# Can't run glmnet on rows with NAs
mergenet.near_vs_far = na.omit(mergenet.near_vs_far)
mergenet.glm.df = mergenet.near_vs_far
numExcludedNA = length(attr(mergenet.near_vs_far, "na.action"))

glm_model = glm(gene_dist_cat ~ coloc_cat + page.rank.pctile_cat + coding_cat + bulk_microglia_cat + sc_microglia_cat, data = mergenet.glm.df, family = "binomial")
drop1(glm_model)

glm_model = glm(gene_dist_cat ~ colocScore + networkScore + codingScore + exprScore, data = mergenet.glm.df, family = "binomial")
drop1(glm_model)
```

Models with all quantitative predictors are better than those with categorical ones, as determined by lower AIC.

Let's see if this is true for each individual component.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
glm_model = glm(gene_dist_cat ~ coloc_cat + page.rank.pctile_cat + coding_cat + bulk_microglia_cat, data = mergenet.glm.df, family = "binomial")
summary(glm_model)

glm_model = glm(gene_dist_cat ~ colocScore + page.rank.pctile_cat + coding_cat + bulk_microglia_cat, data = mergenet.glm.df, family = "binomial")
summary(glm_model)

glm_model = glm(gene_dist_cat ~ coloc_cat + networkScore + coding_cat + bulk_microglia_cat, data = mergenet.glm.df, family = "binomial")
summary(glm_model)

glm_model = glm(gene_dist_cat ~ coloc_cat + page.rank.pctile_cat + codingScore + bulk_microglia_cat, data = mergenet.glm.df, family = "binomial")
summary(glm_model)

glm_model = glm(gene_dist_cat ~ coloc_cat + page.rank.pctile_cat + coding_cat + exprScore, data = mergenet.glm.df, family = "binomial")
summary(glm_model)

```

Now let's do a proper glmnet cross-validated model, and confirm that quantitative predictors are better.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
testGlmnetModelDist = function(full.df, fit.df, type.measure = "deviance") {
  set.seed(123) 
  # Dummy code categorical predictor variables
  x <- model.matrix(gene_dist_cat ~ ., fit.df)[,-1]
  # Convert the outcome (class) to a numerical variable
  y <- ifelse(fit.df$gene_dist_cat == "<= 10000", 1, 0)
  
  set.seed(123) 
  cv.lasso <- cv.glmnet(x, y, alpha = 1, family = "binomial", type.measure=type.measure)
  plot(cv.lasso)
  print(coef(cv.lasso, cv.lasso$lambda.min))
  
  x.full = model.matrix(gene_dist_cat ~ ., fit.df)[,-1]
  probabilities <- cv.lasso %>% predict(newx = x.full, type = "response", s = "lambda.min")
  
  out.df = bind_cols(prob = probabilities,
                     full.df[, colnames(fit.df)],
                     full.df[,] %>% select(-one_of(colnames(fit.df))))
  
  #print(ggplot(out.df, aes(x = prob)) + geom_histogram() + ggtitle("Training set predictions"))
  return(list(out.df = out.df, cv.lasso = cv.lasso))
}

ViewPredDist = function(full.df, fit.df, cv.lasso) {
  x.fit = model.matrix(gene_dist_cat ~ ., fit.df)[,-1]
  full.df$prob = cv.lasso %>% predict(newx = x.fit, type = "response", s = "lambda.min")
  View(full.df %>% select(prob, one_of(colnames(fit.df)), everything(), -leadSNP, -chr:-lead_p))
}

#####################
res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, coloc_cat, page.rank.pctile_cat, coding_cat, bulk_microglia_cat, sc_microglia_cat),
                      type.measure = "mse")
min(res.dist$cv.lasso$cvm)

res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, colocScore, networkScore, codingScore, exprScore),
                      type.measure = "mse")
min(res.dist$cv.lasso$cvm)
```

Next, examine whether other cell type expression scores add anything.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, colocScore, networkScore, codingScore, exprScore),
                      type.measure = "mse")
min(res.dist$cv.lasso$cvm)

res.dist.final = res.dist # This model (with all 4 predictors) is the best / final one
#ViewPredDist(mergenet.near_vs_far, mergenet.near_vs_far %>% select(gene_dist_cat, colocScore, networkScore, codingScore, exprScore), res.dist.final$cv.lasso)

res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, colocScore, networkScore, codingScore, exprScore, scAstrocyteScore, bulkHippocampusScore),
                      type.measure = "mse")
min(res.dist$cv.lasso$cvm)
```

Neither sc Astrocyte nor GTEx hippocampus gets added to the model - coefficients are zero, and a model without them has lower MSE.

Let's drop one term at a time to check how important each is to the predictions, as measured by the change in the MSE when the predictor is excluded.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, colocScore, networkScore, codingScore, exprScore),
                      type.measure = "mse")
min(res.dist$cv.lasso$cvm)

res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, networkScore, codingScore, exprScore),
                      type.measure = "mse")
min(res.dist$cv.lasso$cvm)

res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, colocScore, codingScore, exprScore),
                      type.measure = "mse")
min(res.dist$cv.lasso$cvm)

res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, colocScore, networkScore, exprScore),
                      type.measure = "mse")
min(res.dist$cv.lasso$cvm)

res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, colocScore, networkScore, codingScore),
                      type.measure = "mse")
min(res.dist$cv.lasso$cvm)

```


```{r, warning=FALSE, message=FALSE, echo=FALSE, eval=FALSE}
# See what the results look like when we use AUC or deviance instead of MSE
res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, networkScore, codingScore, primaryMicrogliaScore, scMicrogliaScore),
                      type.measure = "auc")
min(res.dist$cv.lasso$cvm)

res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, colocScore, codingScore, primaryMicrogliaScore, scMicrogliaScore),
                      type.measure = "deviance")
min(res.dist$cv.lasso$cvm)
res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, colocScore, networkScore, primaryMicrogliaScore, scMicrogliaScore),
                      type.measure = "deviance")
min(res.dist$cv.lasso$cvm)
res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, colocScore, networkScore, codingScore),
                      type.measure = "deviance")
min(res.dist$cv.lasso$cvm)
res.dist = testGlmnetModelDist(mergenet.near_vs_far,
                      mergenet.near_vs_far %>% select(gene_dist_cat, primaryMicrogliaScore, scMicrogliaScore),
                      type.measure = "deviance")
min(res.dist$cv.lasso$cvm)

merged.fit.df = merged.ml.df %>% select(gene_dist_cat, coloc_cat, page.rank.pctile_cat, coding_cat, bulk_microglia_cat, sc_microglia_cat, sc_astrocyte_cat, bulk_Brain_Hippocampus_cat)
x.fit = model.matrix(gene_dist_cat ~ ., merged.fit.df)[,-1]
```

Build a model to predict genes with high network scores.
First, we see if a model with quantitative scores is more predictive than one with categorical.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
# Predict high network score genes, relative to low

testGlmnetModelNetwork = function(full.df, fit.df, type.measure = "deviance") {
  set.seed(123) 
  # Dummy code categorical predictor variables
  x <- model.matrix(page.rank.pctile_cat ~ ., fit.df)[,-1]
  # Convert the outcome (class) to a numerical variable
  y <- ifelse(fit.df$page.rank.pctile_cat == "> 80", 1, 0)
  
  set.seed(123) 
  #cv.lasso <- cv.glmnet(x, y, alpha = 1, family = "binomial")
  cv.lasso <- cv.glmnet(x, y, alpha = 0, family = "binomial", type.measure=type.measure)
  plot(cv.lasso)
  print(coef(cv.lasso, cv.lasso$lambda.min))
  #print(coef(cv.lasso, cv.lasso$lambda.1se))
  x.full = model.matrix(page.rank.pctile_cat ~ ., fit.df)[,-1]
  probabilities <- cv.lasso %>% predict(newx = x.full, type = "response", s = "lambda.min")
  
  out.df = bind_cols(prob = probabilities,
                     full.df[, colnames(fit.df)],
                     full.df[,] %>% select(-one_of(colnames(fit.df))))
  
  #print(ggplot(out.df, aes(x = prob)) + geom_histogram() + ggtitle("Training set predictions"))
  return(list(out.df = out.df, cv.lasso = cv.lasso))
}

ViewPredNetwork = function(full.df, fit.df, cv.lasso) {
  x.fit = model.matrix(page.rank.pctile_cat ~ ., fit.df)[,-1]
  full.df$prob = cv.lasso %>% predict(newx = x.fit, type = "response", s = "lambda.min")
  View(full.df %>% select(prob, is_candidate, one_of(colnames(fit.df)), everything(), -leadSNP, -chr:-lead_p))
}

# res.network = testGlmnetModel(mergenet.network_hi_vs_lo,
#                       mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, gene_dist_cat, coloc_cat, coding_cat, bulk_microglia_cat, sc_microglia_cat, sc_astrocyte_cat, bulk_Brain_Hippocampus_cat),
#                       type.measure = "deviance")
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, gene_dist_cat, coloc_cat, coding_cat, bulk_microglia_cat, sc_microglia_cat),
                      type.measure = "mse")
min(res.network$cv.lasso$cvm)

# Comparing gene_dist_cat with a geneDistScore shows that the quantitative score can
# do better... when it is steep, i.e. more heavily penalises distal genes.
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, coloc_cat, coding_cat, bulk_microglia_cat, sc_microglia_cat),
                      type.measure = "mse")
min(res.network$cv.lasso$cvm)

res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, colocScore, codingScore, exprScore),
                      type.measure = "mse")
min(res.network$cv.lasso$cvm)

```

Quantitative predictors are better.

Next, see what the model MSE is when each predictor is dropped singly.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, colocScore, codingScore, exprScore),
                      type.measure = "mse")
min(res.network$cv.lasso$cvm)


res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, colocScore, codingScore, exprScore),
                      type.measure = "mse")
min(res.network$cv.lasso$cvm)
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, codingScore, exprScore),
                      type.measure = "mse")
min(res.network$cv.lasso$cvm)
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, colocScore, exprScore),
                      type.measure = "mse")
min(res.network$cv.lasso$cvm)
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, colocScore, codingScore),
                      type.measure = "mse")
min(res.network$cv.lasso$cvm)

# Drop both coloc and coding
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, exprScore),
                      type.measure = "mse")
min(res.network$cv.lasso$cvm)

res.network.final = res.network
#ViewPredNetwork(mergenet.network_hi_vs_lo, mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, exprScore), res.network.final$cv.lasso)

```

When predicting high-network genes, including colocScore and codingScore result in worse models based on MSE.



```{r, warning=FALSE, message=FALSE, echo=FALSE, eval=FALSE}
# OLD
# Conclusion from this section - basing training on AUC doesn't work well, as it seems
# to be random what AUC you get as the penalty changes, and you end up with gene
# predictions that vary only slightly from each other.
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, colocScore, codingScore, exprScore),
                      type.measure = "auc")
max(res.network$cv.lasso$cvm) # For AUC we want to find the max score


res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, colocScore, codingScore, exprScore),
                      type.measure = "auc")
max(res.network$cv.lasso$cvm)
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, codingScore, exprScore),
                      type.measure = "auc")
max(res.network$cv.lasso$cvm)
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, colocScore, exprScore),
                      type.measure = "auc")
max(res.network$cv.lasso$cvm)
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, colocScore, codingScore),
                      type.measure = "auc")
max(res.network$cv.lasso$cvm)

# Drop both coloc and coding
res.network = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, exprScore),
                      type.measure = "auc")
max(res.network$cv.lasso$cvm)

res.network.final = res.network
```



Next, let's see if adding in brain or astrocyte expression improves the model.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
res.network1 = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, exprScore, bulkHippocampusScore),
                      type.measure = "mse")
min(res.network1$cv.lasso$cvm)

res.network2 = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, exprScore, scAstrocyteScore),
                      type.measure = "mse")
min(res.network2$cv.lasso$cvm)

res.network3 = testGlmnetModelNetwork(mergenet.network_hi_vs_lo,
                      mergenet.network_hi_vs_lo %>% select(page.rank.pctile_cat, geneDistScore, exprScore, bulkHippocampusScore, scAstrocyteScore),
                      type.measure = "mse")
min(res.network3$cv.lasso$cvm)

```

The model has worse MSE when brain or astrocyte expression score are included as predictors.

Let's now make predictions using each model, across all locus genes. We'll look at the correlation between the gene predictions, and create a final "model score" that is the average of the two model predictions.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
#res.dist.final
#res.network.final

merged.tofit.df = merged.df %>% ungroup() %>% select(colocScore, networkScore, codingScore, exprScore) %>% as.matrix()
merged.tofit.df[is.na(merged.tofit.df)] = 0
geneProbs.dist <- as.numeric(res.dist.final$cv.lasso %>% predict(newx = merged.tofit.df, type = "response", s = "lambda.min"))

merged.tofit.df = merged.df %>% ungroup() %>% select(geneDistScore, exprScore) %>% as.matrix()
merged.tofit.df[is.na(merged.tofit.df)] = 0
geneProbs.network <- as.numeric(res.network.final$cv.lasso %>% predict(newx = merged.tofit.df, type = "response", s = "lambda.min"))

merged.df$model_prob_dist = geneProbs.dist
merged.df$model_prob_network = geneProbs.network
merged.df$model_prob = (merged.df$model_prob_dist + merged.df$model_prob_network) / 2

merged.label.df = merged.df %>% select(locus, symbol, model_prob, model_prob_network, model_prob_dist, totalScore) %>%
  filter(model_prob_network > 0.5 | model_prob_dist > 0.25) %>%
  group_by(locus) %>%
  arrange(desc(model_prob)) %>%
  mutate(top_at_locus = (row_number() == 1))
ggplot(merged.df, aes(x=model_prob_network, y=model_prob_dist)) +
  geom_point(alpha = 0.5) +
  geom_text(data = merged.label.df, mapping = aes(label = symbol), size=2, hjust=0, nudge_x=0.005)
# View(merged.df %>%
#        select(model_prob, model_prob_dist, model_prob_network, locus, symbol, geneDistScore, colocScore, networkScore, codingScore, exprScore) %>%
#        arrange(desc(model_prob)) %>%
#        group_by(locus) %>%
#        mutate(rank = row_number()) %>%
#        arrange(locus, rank)
#      )
# View(merged.df %>%
#        select(model_prob, model_prob_dist, model_prob_network, locus, symbol, geneDistScore, colocScore, networkScore, codingScore, exprScore) %>%
#        arrange(desc(model_prob)) %>%
#        group_by(locus) %>%
#        mutate(rank = row_number()) %>%
#        filter(rank < 4) %>%
#        arrange(locus, rank)
#      )

```

Let's also look at a scatterplot and correlation of the model score predictions and the total score.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
cor.val = cor(merged.df$model_prob, merged.df$totalScore, method = "spearman", use = "pairwise.complete.obs")
x = cor.test(merged.df$model_prob, merged.df$totalScore, method = "spearman")

ggplot(merged.df, aes(x=model_prob, y=totalScore)) +
  geom_point(alpha = 0.5) +
  geom_text(data = merged.label.df %>% filter(model_prob > 0.37), mapping = aes(label = symbol), size=2, hjust=0, nudge_x=0.005) +
  annotate("text", x=0.15, y=4.25, label = sprintf("Spearman rho = %.3f", cor.val), hjust = 0)
#  scale_color_manual(values = c("FALSE" = "black", "TRUE" = "blue"))

#View(merged.df %>% select(locus, symbol, model_prob, totalScore))
#View(merged.df %>% select(locus, symbol, model_prob, totalScore, starts_with("bulk"), everything()))
```

How many of the top 20 genes by model probability are in the top 40 by total score?

```{r, warning=FALSE, message=FALSE}
top_by_model = merged.df %>% arrange(desc(model_prob)) %>% .$symbol
top_by_totalscore = merged.df %>% arrange(desc(totalScore)) %>% .$symbol

#top_by_model[1:20]
#top_by_totalscore[1:20]
sum(top_by_model[1:20] %in% top_by_totalscore[1:20])
sum(top_by_model[1:40] %in% top_by_totalscore[1:40])
sum(is.na(merged.df$page.rank.pctile))

sum(merged.df$biotype == "protein_coding")
sum(is.na(merged.df %>% filter(biotype == "protein_coding") %>% .$page.rank.pctile))
sum(!is.na(merged.df %>% filter(biotype == "protein_coding") %>% .$page.rank.pctile))

```


Finally, we save the results in a table.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
bulk_expr.tpm.df = readr::read_tsv(bulk_expr_tpm_file)
sc_brain_expr.tpm.df = readr::read_tsv(sc_expr_brain_tpm_file)
geneExpr.df = readr::read_tsv(gene_expr_file) %>%
  rename(gene_id = geneID) %>%
  select(gene_id, ROSMAP_brain_TPM = ROSMAP_brain)


merged.tosave = merged.df %>%
  ungroup() %>%
  select(locus, gene_id, symbol, biotype, chr, gene_start, gene_end, dist, model_prob, totalScore, geneDistScore, codingScore, colocScore, networkScore, exprScore,
         Primary_microglia_pctile = bulk_primary_microglia,
         Single_cell_Microglia_pctile = sc_Microglia) %>%
  left_join(bulk_expr.tpm.df %>% select(gene_id, Primary_microglia_TPM = primary_microglia,
                                        Brain_Hippocampus_TPM = Brain_Hippocampus), by="gene_id") %>%
  left_join(geneExpr.df, by="gene_id") %>%
  left_join(sc_brain_expr.tpm.df %>% select(gene_id, Single_cell_Microglia_TPM = Microglia)) %>%
  select(locus:Primary_microglia_TPM, Single_cell_Microglia_TPM, Brain_Hippocampus_TPM, ROSMAP_brain_TPM)

write_tsv(merged.tosave, path = paste0(output_root, "AD.loci.1Mb_window.all.geneScores.tsv"))

roundToString = function(x) {
  if (is.na(x)) { return("") }
  sprintf("%.3g", round(x, digits = 3))
}

merged.rounded = cbind(merged.tosave %>% select(locus:dist),
                       apply(merged.tosave %>% select(model_prob:ROSMAP_brain_TPM),
                             MARGIN=c(1, 2), FUN=roundToString))
write_tsv(merged.rounded, path = paste0(output_root, "AD.loci.1Mb_window.all.geneScores.rounded.tsv"))
```


