---
title: "Gene prioritization at AD GWAS loci"
output: html_document
---

## Introduction
To prioritise genes at Alzheimer's disease risk loci, we integrate multiple sources of evidence:
- distance from gene to genomic window defined by independent lead SNPs
- colocalisation score integrated across multiple QTL datasets
- fine-mapping probability of coding variants
- specificity of expression in either microglia or brain, relative to all GTEx tissues
- network-based prioritisation of genes

Each of these inputs (except for distance) has been processed in a separate analysis, and is integrated here.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
library(tidyverse)
library(ggExtra)
library(cowplot)
library(scales)
library(pROC)
knitr::opts_chunk$set(fig.width=7, fig.height=4.8)

theme_set(theme_bw())

savePlots = T

ad_dir = "/Users/jeremys/work/opentargets/AD_finemap"
annotated_snps_file = paste0(ad_dir, "/annotated/AD.meta.annotated.selected.probable.paintor.tsv")
coloc_scores_file = paste0(ad_dir, "/coloc/coloc.AD.meta.colocScores.per_gene.tsv")
network_file = paste0(ad_dir, "/network/ad_network.loci.500kb.tsv")
gene_expr_file = paste0(ad_dir, "/genes/AD.loci.1Mb_window.gene_overlaps.all.selected_tpms.tsv")
output_root = "/Users/jeremys/work/opentargets/AD_finemap/genes/"

annot.df = readr::read_tsv(annotated_snps_file, col_names = T,
                           col_types = cols(.default = col_character(), mean_prob="d", paintor_pp="d",
                                            finemap_prob_nc="d", gcta_maxCondProb="d", spliceai_max_DS="d"))
colocScores.df = readr::read_tsv(coloc_scores_file) %>%
  filter(!is.na(ensembl_id))
colocScores.df$maxH4 = apply(colocScores.df %>% select(-locus_name, -ensembl_id, -geneSymbol, -colocScore, -scoreRank), MARGIN = 1,
                             function(x) max(x, na.rm=T))

network.df = readr::read_tsv(network_file)

# The gene expression table includes genes within 500 kb of AD GWAS peaks, and is
# the basis for our gene annotations
geneExpr.df = readr::read_tsv(gene_expr_file)

geneExpr.df$maxExpr = apply(geneExpr.df[,16:ncol(geneExpr.df)], MARGIN=1, FUN=function(x) max(x, na.rm=T))
geneExpr.df = geneExpr.df %>%
  filter(grepl("protein_coding", biotype) | maxExpr > 1) %>%
  select(-maxExpr, -lead_neglog10p)

write.table(geneExpr.df, file=paste0(output_root, "AD.loci.1Mb_window.expressed_genes.all.tpms.tsv"),
            col.names = T, row.names = F, quote=F, na="", sep="\t")
```


## Gene distance score

First we develop a function that captures intuition about the likelihood of a gene to be functionally relevant based on its distance from the GWAS peak. The function used is as follows.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 3, fig.width = 5}
###############################################################################
# Gene distance evidence
geneDistScore = function(x) {
  distBias = 6355
  maxDist = 5e5
  scores = (log10(maxDist) - log10((abs(x)+distBias))) / (log10(maxDist) - log10(distBias))
  scores[scores < 0] = 0
  scores
}

#df = data.frame(dist = c(0, 100, 1000, 10000, 50000, 100000, 250000))
#df$score = geneDistScore(df$dist)
#df
p.dist = ggplot(data.frame(x = c(1, 500000)), aes(x = x)) +
  stat_function(fun = geneDistScore) +
  scale_x_log10() + xlab("Gene distance") + ylab("Score")
p.dist

if (savePlots) {
  pdf(file = paste0(output_root, "score.geneDistFunction.pdf"), width = 5, height = 3.5)
  print(p.dist)
  dev.off()
}
```

The function is defined to be smooth over the interval from 0 to 500kb. Below are examples of the scores that would be obtained for genes at different distances.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
dist_scores = data.frame(distance = c(0, 100, 1000, 10000, 50000, 100000, 250000))
dist_scores$score = geneDistScore(dist_scores$distance)
print(dist_scores)
```


```{r, warning=FALSE, message=FALSE, echo=FALSE}

###############################################################################
# Coding variant fine-mapping evidence
codingScores.df = annot.df %>%
  filter(grepl("missense|frameshift_variant|splice_donor_variant|stop_gained|stop_lost", Consequence)) %>%
  group_by(Gene) %>%
  summarise(codingScore = sum(paintor_pp))

codingScores.df = annot.df %>%
  filter(grepl("missense|frameshift_variant|splice_donor_variant|stop_gained|stop_lost", Consequence)) %>%
  group_by(Gene) %>%
  summarise(codingScore = sum(mean_prob) * 2)
#sum(codingScores.df$codingScore)

###############################################################################
# Merge scores together
getGeneScores = function(geneExpr.df, useMaxH4 = F) {
  merged.df = geneExpr.df %>%
    left_join(colocScores.df %>% select(locus_name, ensembl_id, colocScore, colocScoreRank = scoreRank, colocMaxH4 = maxH4),
              by = c("locus" = "locus_name", "geneID" = "ensembl_id")) %>%
    group_by(locus, geneID) %>%
    mutate(colocScore = ifelse(is.na(colocScore), 0, colocScore),
           colocMaxH4 = ifelse(is.na(colocMaxH4), 0, colocMaxH4)) %>%
    left_join(codingScores.df, by=c("geneID"="Gene")) %>%
    mutate(codingScore = ifelse(is.na(codingScore), 0, codingScore),
           geneDistScore = geneDistScore(dist)) %>%
    ungroup() %>%
    mutate(colocScoreRank = rank(-colocScore),
           colocMaxH4Rank = rank(-colocMaxH4))
  
  # First version of expression scores
  # merged.df = merged.df %>%
  #   group_by(locus, geneID) %>%
  #   mutate(brainExprSpecificityScore = max(0, brain_quantile - 0.5) * 2,
  #          microgliaExprSpecificityScore = max(0, microglia_quantile - 0.5) * 2,
  #          brainExprScore = if_else(is.na(brainExprSpecificityScore), 0, brainExprSpecificityScore),
  #          microgliaExprScore = if_else(is.na(microgliaExprSpecificityScore), 0, microgliaExprSpecificityScore),
  #          exprScore = max(microgliaExprScore, brainExprScore)
  #          )
  merged.df = merged.df %>%
    group_by(locus, geneID) %>%
    mutate(brainExprSpecificityScore = max(0, brain_quantile - 0.5) * 2,
           microgliaExprSpecificityScore = max(0, microglia_quantile - 0.5) * 2,
           brainExprScore = if_else(is.na(brainExprSpecificityScore), 0, brainExprSpecificityScore),
           microgliaExprScore = if_else(is.na(microgliaExprSpecificityScore), 0, microgliaExprSpecificityScore),
           exprScore = max(microgliaExprScore, brainExprScore),
           meanBrainExp = mean(c(ROSMAP_brain, `BrainSeq brain`, GTEx_hippocampus, `Brain - Cerebellum`, `Brain - Cortex`), na.rm=T),
           brainIsExpressedScore = if_else(is.na(meanBrainExp) | meanBrainExp < 1, 0, 1),
           microgliaIsExpressedScore = if_else(is.na(primary_microglia) | primary_microglia < 1, 0, 1),
           isExpressedScore = max(microgliaIsExpressedScore, brainIsExpressedScore),
           microgliaExprLevelScore = if_else(is.na(primary_microglia) | primary_microglia < 1, 0, log10(primary_microglia) / 3),
           brainExprLevelScore = if_else(is.na(meanBrainExp) | meanBrainExp < 1, 0, log10(meanBrainExp) / 3),
           exprLevelScore = max(microgliaExprLevelScore, brainExprLevelScore)
           )
  # merged.df = merged.df %>%
  #   group_by(locus, geneID) %>%
  #   mutate(meanBrainExp = mean(c(ROSMAP_brain, `BrainSeq brain`, GTEx_hippocampus, `Brain - Cerebellum`, `Brain - Cortex`), na.rm=T),
  #          brainIsExpressedScore = if_else(is.na(meanBrainExp) | meanBrainExp < 1, 0, 1),
  #          microgliaIsExpressedScore = if_else(is.na(primary_microglia) | primary_microglia < 1, 0, 1),
  #          brainExprSpecificityScore = max(0, brain_quantile - 0.5) * 2,
  #          microgliaExprSpecificityScore = max(0, microglia_quantile - 0.5) * 2,
  #          brainExprSpecificityScore = if_else(is.na(brainExprSpecificityScore), 0, brainExprSpecificityScore),
  #          microgliaExprSpecificityScore = if_else(is.na(microgliaExprSpecificityScore), 0, microgliaExprSpecificityScore),
  #          brainExprScore = brainIsExpressedScore * 0.25 + brainExprSpecificityScore * 0.75,
  #          microgliaExprScore = microgliaIsExpressedScore * 0.25 + microgliaExprSpecificityScore * 0.75)
  merged.df = merged.df %>%
    mutate(exprScore = max(microgliaExprScore, brainExprScore))
  
  merged.df = merged.df %>%
    left_join(network.df %>% select(geneID, page.rank.pctile), by="geneID") %>%
    mutate(networkScore = max(0, page.rank.pctile - 50) / 100 * 2) %>%
    mutate(networkScore = if_else(is.na(networkScore), 0, networkScore))
  
  if (useMaxH4) {
    merged.df = merged.df %>% mutate(colocScore = colocMaxH4, colocScoreRank = colocMaxH4Rank)
  }
  
  merged.df = merged.df %>%
    mutate(totalScore = codingScore + colocScore + geneDistScore + exprScore + networkScore,
           totalScore.noNetwork = codingScore + colocScore + geneDistScore + exprScore,
           totalScore.noColoc = codingScore + geneDistScore + exprScore + networkScore,
           totalScore.noCoding = colocScore + geneDistScore + exprScore + networkScore,
           totalScore.noDist = codingScore + colocScore + exprScore + networkScore,
           totalScore.noExpr = codingScore + colocScore + geneDistScore + networkScore) %>%
    ungroup()
  merged.df
}

merged.df = getGeneScores(geneExpr.df, useMaxH4 = F)
#View(merged.df[, -c(6:11)])
```

## Score contributions

Let's see what the total score contributions are for each category across all genes.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
#View(merged.df %>% select(locus, biotype, symbol, dist, contains("Score")))
totals.df = merged.df %>%
  ungroup() %>%
  summarise(codingScore = sum(codingScore),
            colocScore = sum(colocScore),
            maxH4Score = sum(colocMaxH4),
            networkScore = sum(networkScore),
            geneDistScore = sum(geneDistScore),
            exprScore = sum(exprScore)) %>% t()
colnames(totals.df)[1] = "Score total"
print(totals.df)
```

It might seem like the coding score is contributing far too little, but the reason is that this comes from fine-mapping, and the great majority of genes do not have a fine-mapped coding SNP. Genes that do (APH1B, CD33) are already boosted to a very significant degree by this score. The expression score contributes the most because every gene in the 1-Mb window can get some level of score, if it is expressed in microglia or brain above the median across tissues.

## Compare colocScore and maxH4

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 4, fig.width = 5}
p = ggplot(merged.df, aes(x=colocScore, y=colocMaxH4)) +
  geom_point(alpha=0.4) + geom_smooth(method="lm")
print(ggMarginal(p, margins = 'both', type="density", size=10))
```

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 4, fig.width = 6}
# Plot colocScore and maxH4 ranks against each other
#p = ggplot(merged.df %>% filter(!is.na(colocScore)), aes(x=colocScoreRank, y=colocMaxH4Rank)) +
#  geom_point() + geom_abline(slope=1, intercept=0, col="red")

ggplot(merged.df %>% gather("scoreType", "score", c(colocScore, colocMaxH4)), aes(x=score, fill=scoreType)) +
  geom_density(alpha=0.5)

print("colocScore:")
summary(merged.df$colocScore)
print("maxH4:")
summary(merged.df$colocMaxH4)
```

ColocScore and maxH4 are highly correlated. On average maxH4 is slightly higher.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 4, fig.width = 6}
# See how many of the top AD genes are recovered by either colocScore or maxH4
#top_candidate_genes = readr::read_tsv(file.path(ad_dir, "candidate_genes.top.tsv"), col_names = c("symbol"))$symbol
#selected_genes = readr::read_tsv(file.path(ad_dir, "candidate_genes.tsv"), col_names = c("symbol"))$symbol
selected_genes = merged.df %>% arrange(desc(totalScore.noColoc)) %>% .$symbol %>% .[1:40]
#View(merged.df2 %>% arrange(desc(totalScore.noColoc)))
merged.df$is_candidate = merged.df$symbol %in% selected_genes
merged.df$label = NA
merged.df$label[merged.df$is_candidate] = merged.df$symbol[merged.df$is_candidate]
#sum(merged.df$is_candidate)
#View(merged.df %>% select(leadSNP:dist, is_candidate, totalScore, codingScore, colocScore, colocScoreRank, colocMaxH4, colocMaxH4Rank) %>% arrange(desc(is_candidate), desc(totalScore)))
merged.df2 = merged.df %>%
  group_by(locus, geneID) %>%
  mutate(score_gt_0.8 = if_else(colocScore >= 0.8, if_else(colocMaxH4 >= 0.8, "both", "colocScore"), if_else(colocMaxH4 >= 0.8, "maxH4", "neither")),
         rank_lt_50 = if_else(colocScoreRank <= 50, if_else(colocMaxH4Rank <= 50, "both", "colocScore"), if_else(colocMaxH4Rank <= 50, "maxH4", "neither")),
         rank_lt_80 = if_else(colocScoreRank <= 80, if_else(colocMaxH4Rank <= 80, "both", "colocScore"), if_else(colocMaxH4Rank <= 80, "maxH4", "neither")))

scoreColours = c("both" = "blue", "colocScore" = "red", "maxH4" = "darkorange1", "neither" = "green4")
ggplot(merged.df2 %>% filter(is_candidate), aes(x=colocScore, y=colocMaxH4, col=score_gt_0.8)) +
  geom_hline(yintercept=0.8, col="red", alpha=0.8) + geom_vline(xintercept=0.8, col="red", alpha=0.8) +
  geom_point() + geom_smooth(mapping = aes(group="x"), method="lm") + geom_text(aes(label=label), size=2, nudge_x = 0.01, hjust = 0) +
  scale_alpha_manual(values = c("TRUE"=0.8, "FALSE"=0.1)) +
  scale_colour_manual(values = scoreColours, name="score > 0.8") +
  ggtitle("Scores for candidate genes")

ggplot(merged.df2 %>% filter(is_candidate), aes(x=colocScoreRank, y=colocMaxH4Rank, col=rank_lt_80)) +
  geom_hline(yintercept=80, col="red", alpha=0.8) + geom_vline(xintercept=80, col="red", alpha=0.8) +
  geom_point() + geom_smooth(mapping = aes(group="x"), method="lm") + geom_text(aes(label=label), size=2, nudge_x = 0.01, hjust = 0, alpha=0.8) +
  scale_colour_manual(values = scoreColours, name="rank < 80") +
  ggtitle("Ranks for candidate genes")

# ggplot(merged.df2 %>% filter(is_candidate & colocScoreRank < 150), aes(x=colocScoreRank, y=colocMaxH4Rank, col=rank_lt_80)) +
#   geom_hline(yintercept=80, col="red", alpha=0.8) + geom_vline(xintercept=80, col="red", alpha=0.8) +
#   geom_point() + geom_smooth(mapping = aes(group="x"), method="lm") + geom_text(aes(label=label), size=2, nudge_x = 0.01, hjust = 0, alpha=0.8) +
#   scale_colour_manual(values = scoreColours) +
#   ggtitle("Ranks for candidate genes")

ggplot(merged.df2, aes(x=colocScoreRank, y=colocMaxH4Rank, col=rank_lt_50)) +
  geom_hline(yintercept=50, col="red", alpha=0.8) + geom_vline(xintercept=50, col="red", alpha=0.8) +
  geom_point(data = merged.df2 %>% filter(!is_candidate), mapping = aes(x=colocScoreRank, y=colocMaxH4Rank), size=1, alpha=0.2, col="grey20") +
  geom_point(data = merged.df2 %>% filter(is_candidate), mapping = aes(x=colocScoreRank, y=colocMaxH4Rank, col=rank_lt_50), size=2, alpha=0.8) +
  geom_smooth(data = merged.df2, mapping = aes(group="x"), method="lm") +
  scale_x_log10() + scale_y_log10() +
  scale_colour_manual(values = scoreColours, name="rank < 50") +
  ggtitle("Ranks for candidate genes (log scale)")

ggplot(merged.df2 %>% filter(is_candidate), aes(x=colocScoreRank, y=colocMaxH4Rank, col=rank_lt_50)) +
  geom_hline(yintercept=50, col="red", alpha=0.8) + geom_vline(xintercept=50, col="red", alpha=0.8) +
  geom_point() + geom_smooth(mapping = aes(group="x"), method="lm") + geom_text(aes(label=label), size=2, nudge_x = 0.01, hjust = 0, alpha=0.8) +
  scale_x_log10() + scale_y_log10() +
  scale_colour_manual(values = scoreColours, name="rank < 50") +
  ggtitle("Ranks for candidate genes (labelled)")

ggplot(merged.df2 %>% filter(is_candidate), aes(x=colocScoreRank, y=colocMaxH4Rank, col=rank_lt_80)) +
  geom_hline(yintercept=80, col="red", alpha=0.8) + geom_vline(xintercept=80, col="red", alpha=0.8) +
  geom_point() + geom_smooth(mapping = aes(group="x"), method="lm") + geom_text(aes(label=label), size=2, nudge_x = 0.01, hjust = 0, alpha=0.8) +
  scale_x_log10() + scale_y_log10() +
  scale_colour_manual(values = scoreColours, name="rank < 80") +
  ggtitle("Ranks for candidate genes (rank < 80)")

write_tsv(merged.df2 %>% select(geneID, symbol, is_candidate, totalScore.noColoc, weightedColocScore = colocScore, weightedColocScoreRank = colocScoreRank, colocMaxH4, colocMaxH4Rank) %>%
            arrange(desc(is_candidate)),
          path = paste0(output_root, "AD.loci.1Mb_window.coloc_vs_maxH4.tsv", na=""))
```

Scores and ranks are highly correlated for top AD candidate genes as well. Let's quantify exactly how many of our top candidate genes we retrieve with either colocScore or maxH4 at the cutoffs above.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 4, fig.width = 5}
candidates.df = merged.df %>% filter(is_candidate)
retrieval.list = list("colocScore > 0.8" = sum(candidates.df$colocScore > 0.8),
                      "maxH4 > 0.8" = sum(candidates.df$colocMaxH4 > 0.8),
                      "colocRank <= 50" = sum(candidates.df$colocScoreRank <= 50),
                      "maxH4Rank <= 50" = sum(candidates.df$colocMaxH4Rank <= 50),
                      "colocRank <= 80" = sum(candidates.df$colocScoreRank <= 80),
                      "maxH4Rank <= 80" = sum(candidates.df$colocMaxH4Rank <= 80))
retrieval.df = data.frame(retrieved_by = factor(as.character(names(retrieval.list)), levels = c("colocScore > 0.8", "maxH4 > 0.8", "colocRank <= 50", "maxH4Rank <= 50", "colocRank <= 80", "maxH4Rank <= 80")),
                          num_retrieved = as.numeric(retrieval.list),
                          scoreType = c("colocScore", "maxH4", "colocScore", "maxH4", "colocScore", "maxH4"))
ggplot(retrieval.df, aes(x=retrieved_by, y=num_retrieved, fill=scoreType)) +
  geom_bar(stat="identity") +
  ylab("Candidates retrieved") + xlab("Score threshold") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))

retrieval.list
```

The genes we've included as top candidates are:

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 4, fig.width = 5}
print(selected_genes)
```


It looks like the "maxH4" score captures more likely AD causal genes than the colocScore. Let's look at the ROC curves corresponding to this.

### ColocScore vs. maxH4 prediction curves

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 4, fig.width = 5.5}
# library(ROCR)
# pred <- prediction(merged.df$colocScore, merged.df$is_candidate)
# perf <- performance(pred,"tpr","fpr")
# plot(perf,colorize=T)
# auc <- performance(pred,"auc")
# #auc@y.values
# 
# pred2 <- prediction(merged.df$colocMaxH4, merged.df$is_candidate)
# perf2 <- performance(pred2,"tpr","fpr")
# plot(perf2,colorize=F, add=T)
# auc <- performance(pred,"auc")
# #auc@y.values

#quantile(merged.df$colocScore, probs=seq(0, 1, 0.05))
roc.colocScore <- roc(merged.df$is_candidate, merged.df$colocScore)
roc.maxH4 <- roc(merged.df$is_candidate, merged.df$colocMaxH4)

# AUC over full ROC curve
ggroc(list("Weighted H4"=roc.colocScore, "Max H4"=roc.maxH4)) +
  annotate(geom="text", x=0.75, y=0.5, label=sprintf("WeightedH4 full AUC = %.2f", auc(roc.colocScore)), hjust=0) +
  annotate(geom="text", x=0.75, y=0.4, label=sprintf("MaxH4 full AUC = %.2f", auc(roc.maxH4)), hjust=0) +
  ggtitle("ROC") +
  theme_classic(12) + scale_color_discrete(name = "Score")

# Precision-recall curves and F1 scores
colocScore.prec_recall.df = data.frame(t(coords(roc.colocScore, ret = c("recall", "precision"))))
colocScore.prec_recall.df = colocScore.prec_recall.df %>%
  mutate(F1 = 2* precision * recall / (precision + recall),
         Score = "Weighted H4")
colocScore.maxF1 = max(colocScore.prec_recall.df$F1, na.rm=T)

maxH4.prec_recall.df = data.frame(t(coords(roc.maxH4, ret = c("recall", "precision"))))
maxH4.prec_recall.df = maxH4.prec_recall.df %>%
  mutate(F1 = 2* precision * recall / (precision + recall),
         Score = "Max H4")
maxH4.maxF1 = max(maxH4.prec_recall.df$F1, na.rm=T)

prec_recall.df = bind_rows(colocScore.prec_recall.df, maxH4.prec_recall.df) %>%
  mutate(Score = factor(as.character(Score), levels = c("Weighted H4", "Max H4")))
ggplot(prec_recall.df, aes(x=recall, y=precision, col = Score)) + geom_line() + ggtitle("Precision-recall") +
  annotate(geom="text", x=0.45, y=0.75, label=sprintf("WeightedH4 Max F1 = %.2f\nMaxH4 Max F1 = %.2f", colocScore.maxF1, maxH4.maxF1), hjust=0) +
  theme_classic(12)

# Partial AUC for different specificity thresholds
plot.roc(merged.df$is_candidate, merged.df$colocScore,
         percent=T,
         partial.auc=c(100, 80), partial.auc.correct=F, # define a partial AUC (pAUC)
         print.auc=TRUE, #display pAUC value on the plot with following options:
         print.auc.pattern="Corrected pAUC (100-80%% SP):\n%.1f%%", print.auc.col="#1c61b6",
         auc.polygon=TRUE, auc.polygon.col="#1c61b6", # show pAUC as a polygon
         max.auc.polygon=TRUE, max.auc.polygon.col="#1c61b622", # also show the 100% polygon
         print.auc.x=70,
         main="WeightedH4 partial AUC - 80% specificity")

plot.roc(merged.df$is_candidate, merged.df$colocMaxH4,
         percent=T,
         partial.auc=c(100, 80), partial.auc.correct=F, # define a partial AUC (pAUC)
         print.auc=TRUE, #display pAUC value on the plot with following options:
         print.auc.pattern="Corrected pAUC (100-80%% SP):\n%.1f%%", print.auc.col="#1c61b6",
         auc.polygon=TRUE, auc.polygon.col="#1c61b6", # show pAUC as a polygon
         max.auc.polygon=TRUE, max.auc.polygon.col="#1c61b622", # also show the 100% polygon
         print.auc.x=70,
         main="MaxH4 partial pAUC - 80% specificity")

plot.roc(merged.df$is_candidate, merged.df$colocScore,
         percent=T,
         partial.auc=c(100, 90), partial.auc.correct=F, # define a partial AUC (pAUC)
         print.auc=TRUE, #display pAUC value on the plot with following options:
         print.auc.pattern="Corrected pAUC (100-90%% SP):\n%.1f%%", print.auc.col="#1c61b6",
         auc.polygon=TRUE, auc.polygon.col="#1c61b6", # show pAUC as a polygon
         max.auc.polygon=TRUE, max.auc.polygon.col="#1c61b622", # also show the 100% polygon
         print.auc.x=70,
         main="WeightedH4 partial AUC - 90% specificity")

plot.roc(merged.df$is_candidate, merged.df$colocMaxH4,
         percent=T,
         partial.auc=c(100, 90), partial.auc.correct=F, # define a partial AUC (pAUC)
         print.auc=TRUE, #display pAUC value on the plot with following options:
         print.auc.pattern="Corrected pAUC (100-90%% SP):\n%.1f%%", print.auc.col="#1c61b6",
         auc.polygon=TRUE, auc.polygon.col="#1c61b6", # show pAUC as a polygon
         max.auc.polygon=TRUE, max.auc.polygon.col="#1c61b622", # also show the 100% polygon
         print.auc.x=70,
         main="MaxH4 partial pAUC - 90% specificity")

plot.roc(merged.df$is_candidate, merged.df$colocScore,
         percent=T,
         partial.auc=c(100, 95), partial.auc.correct=F, # define a partial AUC (pAUC)
         print.auc=TRUE, #display pAUC value on the plot with following options:
         print.auc.pattern="Corrected pAUC (100-90%% SP):\n%.1f%%", print.auc.col="#1c61b6",
         auc.polygon=TRUE, auc.polygon.col="#1c61b6", # show pAUC as a polygon
         max.auc.polygon=TRUE, max.auc.polygon.col="#1c61b622", # also show the 100% polygon
         print.auc.x=70,
         main="WeightedH4 partial AUC - 95% specificity")

plot.roc(merged.df$is_candidate, merged.df$colocMaxH4,
         percent=T,
         partial.auc=c(100, 95), partial.auc.correct=F, # define a partial AUC (pAUC)
         print.auc=TRUE, #display pAUC value on the plot with following options:
         print.auc.pattern="Corrected pAUC (100-90%% SP):\n%.1f%%", print.auc.col="#1c61b6",
         auc.polygon=TRUE, auc.polygon.col="#1c61b6", # show pAUC as a polygon
         max.auc.polygon=TRUE, max.auc.polygon.col="#1c61b622", # also show the 100% polygon
         print.auc.x=70,
         main="MaxH4 partial pAUC - 95% specificity")

```

Both scores have similar overall AUC, with colocScore very slightly higher. However, maxH4 score has higher partial AUC at all thresholds we might be interested in, i.e. when the score has specificity >80%, 90%, or 95%. In other words, it identifies most candidate genes earlier (with higher specificity).

Based on the above, we replace the colocScore with maxH4 in subsequent analyses.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 2.5, fig.width = 6}
candidate_genes = readr::read_tsv(file.path(ad_dir, "candidate_genes.tsv"), col_names = c("symbol"))$symbol

merged.df = getGeneScores(geneExpr.df, useMaxH4 = T)
merged.df$is_candidate = merged.df$symbol %in% candidate_genes
merged.df$label = NA
merged.df$label[merged.df$is_candidate] = merged.df$symbol[merged.df$is_candidate]

# Round values to 3 decimal places
roundToString = function(x) {
  if (is.na(x)) { return("") }
  sprintf("%.3g", round(x, digits = 3))
}
merged.to_save = merged.df %>% select(leadSNP:dist, totalScore, codingScore, colocScore, colocMaxH4, geneDistScore, networkScore, exprScore, microgliaExprScore, brainExprScore, primary_microglia_tpm = primary_microglia, ipsc_microglia_tpm = ipsc_microglia, ROSMAP_brain_tpm = ROSMAP_brain)
write.table(merged.to_save, file=paste0(output_root, "AD.loci.1Mb_window.expressed_genes.all.geneScores.tsv"),
            col.names = T, row.names = F, quote=F, na="", sep="\t")

merged.rounded = cbind(merged.df %>% select(leadSNP:dist),
                       apply(merged.df %>% select(totalScore, codingScore, colocScore, colocMaxH4, geneDistScore, networkScore, exprScore, microgliaExprScore, brainExprScore, primary_microglia_tpm = primary_microglia, ipsc_microglia_tpm = ipsc_microglia, ROSMAP_brain_tpm = ROSMAP_brain),
                             MARGIN=c(1, 2), FUN=roundToString))
write.table(merged.rounded, file=paste0(output_root, "AD.loci.1Mb_window.expressed_genes.all.geneScores.rounded.tsv"),
            col.names = T, row.names = F, quote=F, na="", sep="\t")
```

## Compare expression specificity score vs. expression level score - ALL genes

I've defined 4 scores:

- isExpressedScore - 1 if gene is expressed at TPM > 1, and 0 otherwise
- exprLevelScore - log10(TPM) for genes with TPM > 1, and 0 otherwise
- exprSpecificityScore - 2 * (expr quantile relative to GTEx tissues - 0.5). This uses the higher expression quantile from either microglia or brain, and subtracts 0.5 to only reward genes expressed in microglia or brain above the average of other tissues.
- exprCombinedScore - 0.5 * exprLevelScore + 0.5 * exprSpecificityScore

Below, I compare these scores with the total score excluding expression (so including distance, coding, coloc, network).

```{r, warning=FALSE, message=FALSE, echo=FALSE}
merged.df2 = merged.df %>%
  mutate(exprSpecificityScore = exprScore)

# x = cor.test(merged.df2$exprSpecificityScore, merged.df2$exprLevelScore, method = "spearman")
# ggplot(merged.df2, aes(x=exprSpecificityScore, y=exprLevelScore)) +
#   geom_point() + theme_bw() + geom_smooth() +
#   annotate("text", x=0.00, y=1, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value))
# 
# x = cor.test(merged.df2$exprSpecificityScore, merged.df2$isExpressedScore, method = "spearman")
# ggplot(merged.df2, aes(x=exprSpecificityScore, y=isExpressedScore)) +
#   geom_point() + theme_bw() + geom_smooth() +
#   annotate("text", x=0.00, y=0.25, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value))

x = cor.test(merged.df2$isExpressedScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=isExpressedScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("IsExpressed vs. total score")

x = cor.test(merged.df2$exprLevelScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=exprLevelScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Expression level vs. total score")

x = cor.test(merged.df2$exprSpecificityScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=exprSpecificityScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Expression specificity vs. total score")

merged.df2$exprCombinedScore = merged.df2$exprLevelScore * 0.5 + merged.df2$exprScore * 0.5
x = cor.test(merged.df2$exprCombinedScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=exprCombinedScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Combined expr level & specificity vs. total score")

merged.df2$microgliaExprCombinedScore = merged.df2$microgliaExprLevelScore * 0.5 + merged.df2$microgliaExprSpecificityScore * 0.5
# x = cor.test(merged.df2$microgliaExprCombinedScore, merged.df2$totalScore.noExpr, method = "spearman")
# ggplot(merged.df2, aes(x=microgliaExprCombinedScore, y=totalScore.noExpr)) +
#   geom_point() + theme_bw() + geom_smooth() +
#   annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
#   ggtitle("Combined microglia expr level & specificity vs. total score")

```

There are a few things to unpack here.

- expression level seems to do the best (highest correlation with total score)
- expression specificity seems to do poorly
- combining the two does worse than just expression level

HOWEVER, an important issue is that I've included many non-coding genes, which are NOT present in the gene network, and are also expressed at lower levels. This introduces a strong correlation between expression level and network score, because protein-coding genes are expressed more highly and only protein-coding genes are in the network.

## Compare expression specificity score vs. expression level score - protein-coding genes

Here we compare the different scores when considering only protein-coding genes, which makes the score more "fair".

```{r, warning=FALSE, message=FALSE, echo=FALSE}
merged.df2 = merged.df2 %>% filter(!is.na(page.rank.pctile))

x = cor.test(merged.df2$isExpressedScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=isExpressedScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("IsExpressed vs. total score")

x = cor.test(merged.df2$exprLevelScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=exprLevelScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Expression level vs. total score")

x = cor.test(merged.df2$exprSpecificityScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=exprSpecificityScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Expression specificity vs. total score")

merged.df2$exprCombinedScore = merged.df2$exprLevelScore * 0.5 + merged.df2$exprScore * 0.5
x = cor.test(merged.df2$exprCombinedScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=exprCombinedScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Combined expr level & specificity vs. total score")

x = cor.test(merged.df2$microgliaExprSpecificityScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=microgliaExprSpecificityScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Microglia expression specificity vs. total score")

x = cor.test(merged.df2$brainExprSpecificityScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=brainExprSpecificityScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Brain expression specificity vs. total score")

x = cor.test(merged.df2$microgliaExprCombinedScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=microgliaExprCombinedScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Combined microglia expr level & specificity vs. total score")

#cor.test(merged.df2$microgliaExprSpecificityScore, merged.df2$totalScore.noExpr, method = "spearman")
#cor.test(merged.df2$brainExprSpecificityScore, merged.df2$totalScore.noExpr, method = "spearman")
#cor.test(merged.df2$microgliaExprLevelScore, merged.df2$totalScore.noExpr, method = "spearman")
#cor.test(merged.df2$brainExprLevelScore, merged.df2$totalScore.noExpr, method = "spearman")
#cor.test(merged.df2$exprSpecificityScore, merged.df2$totalScore.noExpr, method = "spearman")

# summary(merged.df2 %>% filter(is_candidate) %>% .$microgliaExprSpecificityScore)
# summary(merged.df2 %>% filter(is_candidate) %>% .$brainExprSpecificityScore)
# summary(merged.df2 %>% filter(is_candidate) %>% .$exprSpecificityScore)
# summary(merged.df2$microgliaExprSpecificityScore)
# summary(merged.df2$brainExprSpecificityScore)
# summary(merged.df2$exprSpecificityScore)

#View(merged.df[, -c(6:11)])
#View(merged.df2 %>% filter(is_candidate) %>% select(locus, symbol, dist, totalScore, primary_microglia, meanBrainExp, contains("quantile"), microgliaExprSpecificityScore, brainExprSpecificityScore, exprSpecificityScore))
```

Now we see that:

- isExpressed does very poorly
- expression specificity shows the highest correlation
- combining expression specificity and level again does worse

## Compare microglia vs. brain expression scores

Here we compare the expression scores for microglia and brain.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
x = cor.test(merged.df2$microgliaExprSpecificityScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=microgliaExprSpecificityScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Microglia expression specificity vs. total score")

x = cor.test(merged.df2$brainExprSpecificityScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=brainExprSpecificityScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Brain expression specificity vs. total score")

x = cor.test(merged.df2$microgliaExprCombinedScore, merged.df2$totalScore.noExpr, method = "spearman")
ggplot(merged.df2, aes(x=microgliaExprCombinedScore, y=totalScore.noExpr)) +
  geom_point() + theme_bw() + geom_smooth() +
  annotate("text", x=0.00, y=4, hjust=0, col="blue", label=sprintf("Spearman rho=%.3f, p=%.3g", x$estimate, x$p.value)) +
  ggtitle("Combined microglia expr level & specificity vs. total score")

```

- microglia expression specificity shows higher correlation than combined expression specificity
- brain expression specificity has NO relationship with total score
- combining microglia expression specificity and level again does worse.


## Total scores for candidate genes

Next we look at total scores for top candidate genes.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 2.5, fig.width = 6}
#library(pheatmap)
#pheatmap(merged.df %>% ungroup() %>% select(codingScore, colocScore, geneDistScore, totalScore))
#cor(merged.df$geneDistScore, merged.df$noDistScore, use = "pairwise.complete")

###############################################################################
# Plots to understand score distribution

p1 = ggplot(merged.df, aes(x=totalScore, fill=is_candidate)) +
  geom_density(alpha = 0.7) + ggtitle("Score density (total)")
p1
p2 = ggplot(merged.df, aes(x=totalScore.noNetwork, fill=is_candidate)) +
  geom_density(alpha = 0.7) + ggtitle("Score density (no network)")
p2
p3 = ggplot(merged.df, aes(x=networkScore, fill=is_candidate)) +
  geom_density(alpha = 0.7) + ggtitle("Score density (network score)")
p3

if (savePlots) {
  pdf(file = paste0(output_root, "score.density.pdf"), width = 6, height = 4)
  print(p1)
  print(p2)
  print(p3)
  dev.off()
}
```

## Network vs. non-network score

The network score is orthogonal to the other scores. Let's see to what extent they are correlated / support each other.

```{r, warning=FALSE, message=FALSE, echo=FALSE}

###############################################################################
# Network scores vs. scores without network
p1 = ggplot(merged.df, aes(x=totalScore.noNetwork, y=networkScore, col=is_candidate)) +
  geom_point(aes(alpha = is_candidate)) +
  geom_text(mapping = aes(label = label), size=2.2, nudge_x = 0.05, hjust = 0) +
  scale_alpha_manual(values = c("TRUE"=0.75, "FALSE"=0.5), guide=F) +
  scale_color_manual(values = c("TRUE"="dodgerblue", "FALSE"="black"), guide=F) +
  ggtitle("NetworkScore vs. Non-network score")
ggMarginal(p1, margins = 'both', type="density", size=10)
#  geom_smooth(aes(group=1), method="lm") +

if (savePlots) {
  pdf(file = paste0(output_root, "score.network_vs_no_network.scatter.pdf"), width = 6, height = 4.5)
  print(ggMarginal(p1, margins = 'both', type="density", size=10))
  dev.off()
}
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
p2 = ggplot(merged.df, aes(x=totalScore.noNetwork+0.01, y=networkScore+0.01, col=is_candidate)) +
  geom_point(aes(alpha = is_candidate)) +
  geom_text(mapping = aes(label = label), size=2.2, nudge_x = 0.03, hjust = 0) +
  scale_alpha_manual(values = c("TRUE"=0.75, "FALSE"=0.5), guide=F) +
  scale_color_manual(values = c("TRUE"="dodgerblue", "FALSE"="black"), guide=F) +
  scale_x_log10() + scale_y_log10() +
  xlab("log scale: totalScore.noNetwork + 0.01") + ylab("log scale: networkScore + 0.01") +
  ggtitle("NetworkScore vs. Non-network score (log scale)")
ggMarginal(p2, margins = 'both', type="density", size=10)
#  geom_smooth(aes(group=1), method="lm") +

if (savePlots) {
  pdf(file = paste0(output_root, "score.network_vs_no_network.scatter.logScale.pdf"), width = 6, height = 4.5)
  print(ggMarginal(p2, margins = 'both', type="density", size=10))
  dev.off()
}
```

The scores are quite well correlated. More importantly, all of our top candidate genes (which are in the network at all) receive both high network score and high non-network score.

Let's statistically test the correlation.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
ggplot(merged.df, aes(x=totalScore.noNetwork, y=networkScore)) +
  geom_point(alpha = 0.5) +
  ggtitle("Network score vs. non-network score")

cor.test(merged.df$networkScore, merged.df$totalScore.noNetwork, method = "spearman")
```

There's a fairly weak correlation between the network score and the total non-network score, when looking across all loci. But we can also look at gene ranks by these scores *within* loci, which is more relevant to prioritising GWAS genes.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
merged.df2 = merged.df %>%
  group_by(locus) %>%
  mutate(locusNetworkRank = rank(-networkScore, ties.method = "average"),
         locusTotalNoNetworkRank = rank(-totalScore.noNetwork, ties.method = "average"))

ggplot(merged.df2, aes(x=locusTotalNoNetworkRank, y=locusNetworkRank)) +
  geom_jitter(width = 0.75, height = 0.75, alpha = 0.3) +
  ggtitle("Ranks within each locus: network vs. non-network score")

cor.test(merged.df2$locusTotalNoNetworkRank, merged.df2$locusNetworkRank, method = "spearman")
```

In this setting, gene ranks at each locus based on network score are highly correlated with gene ranks based on total non-network score. In other words, the scores are less useful when comparing across loci. This could be, for example, because loci differ in how close their nearest genes are (e.g. CLNK), and in whether there are coding variant associations.

Does this translate into top AD candidates being ranked higher by network score?

```{r, warning=FALSE, message=FALSE, echo=FALSE}
wilcox.test(merged.df2 %>% filter(is_candidate) %>% .$totalScore.noNetwork, merged.df2 %>% filter(is_candidate) %>% .$networkScore)
```

Yes.


## Scores per locus (no network)

```{r, warning=FALSE, message=FALSE, echo=FALSE}
###############################################################################
# Locus scores

p1 = ggplot(merged.df, aes(x=locus, y=totalScore.noNetwork, col=is_candidate, group=locus, size=is_candidate)) +
  geom_point(aes(alpha = is_candidate)) +
  geom_text(mapping = aes(x=locus, y=totalScore.noNetwork, label = label), size=2.2, nudge_x = 0.28, hjust = 0) +
  scale_alpha_manual(values = c("TRUE"=0.75, "FALSE"=0.5), guide=F) +
  scale_color_manual(values = c("TRUE"="dodgerblue", "FALSE"="black"), guide=F) +
  scale_size_manual(values=c("TRUE"=2.1, "FALSE"=1.5), guide=F) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
p1

if (savePlots) {
  pdf(file = paste0(output_root, "scores.per_locus.no_network.pdf"), width = 7, height = 4)
  print(p1)
  dev.off()
}
```

## NetworkScore per locus

```{r, warning=FALSE, message=FALSE, echo=FALSE}
###############################################################################
# Locus scores

p1 = ggplot(merged.df, aes(x=locus, y=networkScore, col=is_candidate, group=locus, size=is_candidate)) +
  geom_point(aes(alpha = is_candidate)) +
  geom_text(mapping = aes(x=locus, y=networkScore, label = label), size=2.2, nudge_x = 0.28, hjust = 0) +
  scale_alpha_manual(values = c("TRUE"=0.75, "FALSE"=0.5), guide=F) +
  scale_color_manual(values = c("TRUE"="dodgerblue", "FALSE"="black"), guide=F) +
  scale_size_manual(values=c("TRUE"=2.1, "FALSE"=1.5), guide=F) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
p1

if (savePlots) {
  pdf(file = paste0(output_root, "scores.per_locus.networkScore.pdf"), width = 7, height = 4.8)
  print(p1)
  dev.off()
}
```

## Total score per locus

```{r, warning=FALSE, message=FALSE, echo=FALSE}
###############################################################################
# Locus scores

p1 = ggplot(merged.df, aes(x=locus, y=totalScore, col=is_candidate, group=locus, size=is_candidate)) +
  geom_point(aes(alpha = is_candidate)) +
  geom_text(mapping = aes(x=locus, y=totalScore, label = label), size=2.2, nudge_x = 0.28, hjust = 0) +
  scale_alpha_manual(values = c("TRUE"=0.75, "FALSE"=0.5), guide=F) +
  scale_color_manual(values = c("TRUE"="dodgerblue", "FALSE"="black"), guide=F) +
  scale_size_manual(values=c("TRUE"=2.1, "FALSE"=1.5), guide=F) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
p1

if (savePlots) {
  pdf(file = paste0(output_root, "scores.per_locus.totalScore.pdf"), width = 7, height = 4.8)
  print(p1)
  dev.off()
}
```

## Total score vs. distance

```{r, warning=FALSE, message=FALSE, echo=FALSE}
###############################################################################
# Score vs. distance

# Full view -500k to +500k, no labels
p1 = ggplot(merged.df, aes(x=dist, y=totalScore, col=is_candidate)) +
  geom_point(aes(alpha = is_candidate)) +
  scale_alpha_manual(values = c("TRUE"=0.9, "FALSE"=0.5), guide=F) +
  scale_color_manual(values = c("TRUE"="dodgerblue", "FALSE"="black")) +
  xlab("Gene distance (bp)") + ylab("Total score") +
  ggtitle("Gene score vs. gene distance")
p1

# Zoomed in view with candidate genes labelled
merged.df$label = NA
merged.df$label[merged.df$is_candidate] = merged.df$symbol[merged.df$is_candidate]
merged.df.zoom = merged.df %>% filter(abs(dist) < 200000)
p2 = ggplot(merged.df.zoom, aes(x=dist, y=totalScore, col=is_candidate)) +
  geom_point(aes(alpha = is_candidate)) +
  geom_text(data = merged.df.zoom %>% filter(dist >= 0), mapping = aes(label = label), size=2.2, nudge_x = 4000, hjust = 0) +
  geom_text(data = merged.df.zoom %>% filter(dist < 0), mapping = aes(label = label), size=2.2, nudge_x = -4000, hjust = 1) +
  scale_alpha_manual(values = c("TRUE"=0.9, "FALSE"=0.5), guide=F) +
  scale_color_manual(values = c("TRUE"="dodgerblue", "FALSE"="black"), guide=F) +
  xlab("Gene distance") + ylab("Total score") +
  ggtitle("Gene score vs. gene distance (zoom, candidates labelled)")
p2

merged.df.zoom$label = NA
merged.df.zoom$label[!merged.df.zoom$is_candidate & merged.df.zoom$totalScore >= 2] = merged.df.zoom$symbol[!merged.df.zoom$is_candidate & merged.df.zoom$totalScore >= 2]
p3 = ggplot(merged.df.zoom, aes(x=dist, y=totalScore, col=is_candidate)) +
  geom_point(aes(alpha = is_candidate)) +
  geom_text(data = merged.df.zoom %>% filter(dist >= 0), mapping = aes(label = label), size=2.2, nudge_x = 4000, hjust = 0) +
  geom_text(data = merged.df.zoom %>% filter(dist < 0), mapping = aes(label = label), size=2.2, nudge_x = -4000, hjust = 1) +
  scale_alpha_manual(values = c("TRUE"=0.9, "FALSE"=0.5), guide=F) +
  scale_color_manual(values = c("TRUE"="dodgerblue", "FALSE"="black"), guide=F) +
  xlab("Gene distance") + ylab("Total score") +
  ggtitle("Gene score vs. gene distance (zoom, high scores labelled)")
p3

if (savePlots) {
  pdf(file = paste0(output_root, "score.dist_vs_score.pdf"), width = 7, height = 4.8)
  print(p1)
  print(p2)
  print(p3)
  dev.off()
}
```

Let's see what fraction of top protein-coding genes by score (not including distance score), are the nearest gene. Below we have the distance ranks for the single top gene per locus (among all genes within 500 kb).

```{r, warning=FALSE, message=FALSE, echo=FALSE}
merged.df2 = merged.df %>%
  filter(biotype == "protein_coding") %>%
  group_by(locus) %>%
  arrange(abs(dist), desc(totalScore.noDist)) %>%
  mutate(distance_rank = row_number()) %>%
  arrange(desc(totalScore.noDist)) %>%
  mutate(noDist_score_rank = row_number())
#View(merged.df2 %>% select(locus, symbol, dist, contains("Score"), contains("Rank")) %>% filter(noDist_score_rank == 1))
vals = table(merged.df2 %>% filter(noDist_score_rank == 1) %>% .$distance_rank)
df = data.frame(distance_rank = names(vals), number = as.integer(vals))
print(df)

numOthersWithin100kb = merged.df2 %>% filter(noDist_score_rank == 1, distance_rank > 1, abs(dist) < 100000) %>% nrow()
print(sprintf("Number which aren't distance rank 1 but are within 100 kb: %d", numOthersWithin100kb))
```

21 of the genes top-ranked by totalScore (excluding distance) are actually the nearest gene, and 8 further genes are within 100 kb, out of a total of 37 loci.

Now do the same, but only considering genes within 200 kb for scoring top genes (except at CLNK, since no genes are within 200 kb).

```{r, warning=FALSE, message=FALSE, echo=FALSE}
merged.df2 = merged.df %>%
  filter(abs(dist) < 200000 | locus == "CLNK", biotype == "protein_coding") %>%
  group_by(locus) %>%
  arrange(abs(dist), desc(totalScore.noDist)) %>%
  mutate(distance_rank = row_number()) %>%
  arrange(desc(totalScore.noDist)) %>%
  mutate(noDist_score_rank = row_number())
#View(merged.df2 %>% select(locus, symbol, dist, contains("Score"), contains("Rank")) %>% filter(noDist_score_rank == 1))
vals = table(merged.df2 %>% filter(noDist_score_rank == 1) %>% .$distance_rank)
df = data.frame(distance_rank = names(vals), number = as.integer(vals))
print(df)

numOthersWithin100kb = merged.df2 %>% filter(noDist_score_rank == 1, distance_rank > 1, abs(dist) < 100000) %>% nrow()
print(sprintf("Number which aren't distance rank 1 but are within 100 kb: %d", numOthersWithin100kb))
```

The results are similar - 23 top-ranked genes are the nearest gene, with 8 further within 100 kb, and the remaining 6 100 - 200 kb away.

## Prediction of candidate genes by score components

Here we look at how well each line of evidence predicts our top AD candidate genes.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 4, fig.width = 5.5}
rocList <- roc(formula = is_candidate ~ geneDistScore + codingScore + colocScore + exprScore + networkScore + totalScore, data = merged.df)
aucStr = sprintf("AUCs:\ntotal = %.3f\ndist = %.3f\ncoloc = %.3f\nnetwork = %.3f\nexpr = %.3f\ncoding = %.3f",
                 auc(rocList$totalScore), auc(rocList$geneDistScore), auc(rocList$colocScore), auc(rocList$networkScore), auc(rocList$exprScore), auc(rocList$codingScore))
ggroc(rocList) +
  ggtitle("ROC curves predicting AD candidates") +
  annotate(geom="text", x=0.35, y=0.55, label=aucStr, hjust=0, vjust=1, size=3.5)
```

Distance is by far the best individual predictor, followed by coloc and then network. The total score is slightly better than distance.

Let's see what this looks like when we compare the total score to totals with each individual component left out.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 4, fig.width = 5.5}
rocList <- roc(formula = is_candidate ~ totalScore.noExpr + totalScore.noNetwork + totalScore.noColoc + totalScore.noCoding + totalScore.noDist + totalScore, data = merged.df)
aucStr = sprintf("AUCs:\ntotal = %.3f\nnoColoc = %.3f\nnoCoding = %.3f\nnoExpr = %.3f\nnoNetwork = %.3f\nnoDist = %.3f",
                 auc(rocList$totalScore), auc(rocList$totalScore.noColoc), auc(rocList$totalScore.noCoding), auc(rocList$totalScore.noExpr), auc(rocList$totalScore.noNetwork), auc(rocList$totalScore.noDist))
ggroc(rocList) +
  ggtitle("ROC curves predicting AD candidates") +
  annotate(geom="text", x=0.5, y=0.55, label=aucStr, hjust=0, vjust=1, size=3.5)

paucStr = sprintf("pAUCs:\ntotal = %.3f\nnoExpr = %.3f\nnoCoding = %.3f\nnoColoc = %.3f\nnoNetwork = %.3f\nnoDist = %.3f",
                  auc(rocList$totalScore, partial.auc=c(1, .9), partial.auc.correct=TRUE),
                  auc(rocList$totalScore.noExpr, partial.auc=c(1, .9), partial.auc.correct=TRUE),
                  auc(rocList$totalScore.noCoding, partial.auc=c(1, .9), partial.auc.correct=TRUE),
                  auc(rocList$totalScore.noColoc, partial.auc=c(1, .9), partial.auc.correct=TRUE),
                  auc(rocList$totalScore.noNetwork, partial.auc=c(1, .9), partial.auc.correct=TRUE),
                  auc(rocList$totalScore.noDist, partial.auc=c(1, .9), partial.auc.correct=TRUE))
ggroc(rocList) +
  ggtitle("Partial AUCs") +
  annotate(geom="text", x=0.925, y=0.55, label=paucStr, hjust=0, vjust=1, size=3.5) + coord_cartesian(xlim=c(1, 0.85))

```

Interestingly, removing coding or expression from the score has little effect on the prioritisation. Removing coloc has a minor effect when considering partial AUC, the most relevant metric. Removing network has a modest effect, and removing distance has the largest effect. I suspect that the coding score is partially redundant with the distance score, since you can only get a high fine-mapped coding score for variants that are near the association peak.

Let's see what happens when we combine just distance and each of the other scores.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 4, fig.width = 5.5}
merged.df2 = merged.df %>%
  mutate(distExprScore = geneDistScore + exprScore,
         distColocScore = geneDistScore + colocScore,
         distCodingScore = geneDistScore + codingScore,
         distNetworkScore = geneDistScore + networkScore)
rocList <- roc(formula = is_candidate ~ distExprScore + distColocScore + distCodingScore + distNetworkScore + geneDistScore + totalScore, data = merged.df2)
aucStr = sprintf("AUCs:\ntotal = %.3f\ndistNetwork = %.3f\ndistColoc = %.3f\ndistCoding = %.3f\ndist = %.3f\ndistExpr = %.3f",
                 auc(rocList$totalScore), auc(rocList$distNetworkScore), auc(rocList$distColocScore), auc(rocList$distCodingScore), auc(rocList$geneDistScore), auc(rocList$distExprScore))

 
ggroc(rocList) +
  ggtitle("ROC curves predicting AD candidates") +
  annotate(geom="text", x=0.5, y=0.55, label=aucStr, hjust=0, vjust=1, size=3.5) + coord_cartesian(xlim=c(1, 0))

paucStr = sprintf("pAUCs:\ntotal = %.3f\ndistNetwork = %.3f\ndistCoding = %.3f\ndist = %.3f\ndistColoc = %.3f\ndistExpr = %.3f",
                  auc(rocList$totalScore, partial.auc=c(1, .9), partial.auc.correct=TRUE),
                  auc(rocList$distNetworkScore, partial.auc=c(1, .9), partial.auc.correct=TRUE),
                  auc(rocList$distCodingScore, partial.auc=c(1, .9), partial.auc.correct=TRUE),
                  auc(rocList$geneDistScore, partial.auc=c(1, .9), partial.auc.correct=TRUE),
                  auc(rocList$distColocScore, partial.auc=c(1, .9), partial.auc.correct=TRUE),
                  auc(rocList$distExprScore, partial.auc=c(1, .9), partial.auc.correct=TRUE))
ggroc(rocList) +
  ggtitle("Partial AUCs") +
  annotate(geom="text", x=0.925, y=0.55, label=paucStr, hjust=0, vjust=1, size=3.5) + coord_cartesian(xlim=c(1, 0.85))

#View(merged.df %>% select(locus, symbol, dist, is_candidate, contains("Score"), -contains("no"), -brainExprSpecificityScore, -microgliaExprSpecificityScore) %>% arrange(desc(is_candidate), desc(totalScore)))
```

Interestingly, distance only does better than distance + coloc for the most relevant range of 90% specificity. Dist + coloc is slightly better over the full range, indicating that coloc recovers a small number of distal genes. Also, expression contributes nothing to distance. TotalScore is still the best overall, indicating that distance + network score isn't enough to recover AD candidate genes optimally, and so coloc and coding are somewhat, and expression contribute a very small amount in a combined model.


## Scores contributions per locus

### Single top gene per locus

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 6, fig.width = 5}
merged.ranks.df = merged.df %>%
  group_by(locus) %>%
  arrange(desc(totalScore)) %>%
  mutate(rank_within_locus = row_number()) %>%
  ungroup() %>%
  mutate(rank_overall = row_number())

topGenesPerLocus.df = merged.ranks.df %>% filter(rank_within_locus == 1)
topGenes.tidy.df = topGenesPerLocus.df %>%
  select(locus, geneID, symbol, totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
  gather(key="scoreType", value="score", codingScore:networkScore) %>%
  mutate(scoreType = gsub("Score", "", scoreType))
ggplot(topGenes.tidy.df, aes(x=fct_reorder(symbol, desc(symbol)), y=score, fill=scoreType)) +
  geom_bar(stat="identity") + xlab("Gene") +
  coord_flip()
```


### Top 53 genes, version 1

This plot includes the top single gene from each locus (37 loci), and then adds the top remaining 23 genes to make up 60 genes

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 6, fig.width = 5}
# Function to make barplot which shows score breakdown and color per locus
scoreBarPlot = function(geneScores.df) {
  geneScores.tidy.df = geneScores.df %>%
    select(locus, geneID, symbol, totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
    gather(key="scoreType", value="score", codingScore:networkScore) %>%
    mutate(scoreType = gsub("Score", "", scoreType)) %>%
    arrange(locus, desc(totalScore)) %>%
    mutate(symbol = factor(as.character(symbol), levels = unique(as.character(symbol))),
           shade = as.character(as.integer(as.factor(locus)) %% 2))

  p.locus = ggplot(geneScores.tidy.df, aes(x=symbol)) +
    geom_blank(mapping = aes(ymax=0), stat="identity", position="dodge") +
    geom_raster(aes(x=symbol, y=-0.5, fill=shade), alpha=0.9) +
    xlab("Gene") + ylab("") +
    coord_flip() + theme_bw(8) +
    scale_fill_manual(guide=F, values=c("0"="lightgrey", "1"="grey50")) +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin=unit(c(0,0,0,0), "mm")) +
    scale_y_continuous(expand = c(0,0))
  p.locus
  
  pal = hue_pal()(6)
  #show_col(pal)
  score_colors = c("total" = "darkorange", "coding" = pal[2], "coloc" = pal[3], "expr" = pal[4], "geneDist" = pal[5], "network" = pal[6])
  p.score = ggplot(geneScores.tidy.df, aes(x=symbol, y=score, fill=scoreType)) +
    geom_bar(stat="identity", col="black", size = 0.15) +
    xlab("Gene") +
    coord_flip() + theme_bw(8) +
    theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(),
          plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank()) +
    scale_y_continuous(expand = c(0,0)) +
    scale_fill_manual(values = score_colors)
  p.score
  cowplot::plot_grid(plotlist = list(p.locus, p.score), ncol=2, align = "h", rel_widths=c(1.5,6.5))
}

numGenesToShow = 53
topGenesOverall.df = merged.ranks.df %>% filter(rank_overall <= numGenesToShow)
#scoreBarPlot(topGenesOverall.df)


numLoci = length(unique(merged.ranks.df$locus))
numLociWithNoGeneInTopN = numLoci - length(unique(merged.ranks.df %>% filter(rank_overall <= numGenesToShow) %>% .$locus))
topGenesPerLocusAndOverall.df = merged.ranks.df %>%
  filter(rank_within_locus == 1 | rank_overall <= (numGenesToShow - numLociWithNoGeneInTopN))

scoreBarPlot(topGenesPerLocusAndOverall.df)

if (savePlots) {
  pdf(file = paste0(output_root, "score.breakdown.topGenes.v1.pdf"), width = 5, height = 6)
  print(scoreBarPlot(topGenesPerLocusAndOverall.df))
  dev.off()
}
```

### Top 53 genes, version 2

Grey indicator at left to distinguish loci.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 6, fig.width = 5}
scoreBarPlotSeparated2 = function(geneScores.df) {
  geneScores.tidy.df = geneScores.df %>%
    select(locus, geneID, symbol, totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
    gather(key="scoreType", value="score", codingScore:networkScore) %>%
    mutate(scoreType = gsub("Score", "", scoreType)) %>%
    arrange(locus, desc(totalScore)) %>%
    mutate(symbol = factor(as.character(symbol), levels = unique(as.character(symbol))),
           shade = as.character(as.integer(as.factor(locus)) %% 2)) %>%
    mutate(scoreType = factor(as.character(scoreType), levels=c("total", "coding", "coloc", "expr", "geneDist", "network")))
  
  #geneScores.tidy.locus = geneScores.tidy.df %>% filter(scoreType == "coding") %>% mutate(scoreType = "")
  p.locus = ggplot(geneScores.tidy.df %>% filter(scoreType == "coding") %>% mutate(scoreType = ""), aes(x=symbol)) +
    geom_blank(mapping = aes(ymax=0), stat="identity", position="dodge") +
    geom_raster(aes(x=symbol, y=-0.5, fill=shade), alpha=0.9) +
    xlab("Gene") + ylab("") +
    coord_flip() + theme_bw(8) +
    scale_fill_manual(guide=F, values=c("0"="lightgrey", "1"="grey50")) +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin=unit(c(0,0,0,0), "mm")) +
    scale_y_continuous(expand = c(0,0)) +
    facet_wrap(~scoreType, ncol = 1, scales="free_x")
#    scale_fill_manual(guide=F, values=c("0"="cornflowerblue", "1"="firebrick")) +
  
  pal = hue_pal()(6)
  #show_col(pal)
  score_colors = c("total" = "darkorange", "coding" = pal[2], "coloc" = pal[3], "expr" = pal[4], "geneDist" = pal[5], "network" = pal[6], "grey"="grey")
  p.scores = ggplot(geneScores.tidy.df, aes(x=symbol, y=score, fill=scoreType)) +
    geom_bar(stat="identity") +
    xlab("Gene") +
    coord_flip() + theme_bw(8) +
    theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(),
          plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank(), axis.text.x = element_blank()) +
    scale_y_continuous(expand = c(0,0)) +
    facet_wrap(~scoreType, ncol = 5, scales="free_x") +
    scale_fill_manual(guide=F, values = score_colors)

  geneScores.tidy.df = geneScores.df %>%
    mutate(totalScoreX = totalScore) %>%
    select(locus, geneID, symbol, totalScoreX, total = totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
    gather(key="scoreType", value="score", total) %>%
    mutate(scoreType = gsub("Score", "", scoreType)) %>%
    arrange(locus, desc(totalScoreX)) %>%
    mutate(symbol = factor(as.character(symbol), levels = unique(as.character(symbol))),
           shade = as.character(as.integer(as.factor(locus)) %% 2)) %>%
    mutate(scoreType = factor(as.character(scoreType), levels=c("total", "coding", "coloc", "expr", "geneDist", "network")))
  
  #bg.df = geneScores.tidy.df %>% select(symbol) %>% unique()
    #geom_raster(data = bg.df, mapping = aes(x=symbol, width = 1, ymin=-0, ymax=Inf), fill="grey", alpha=0.7)
  p.totalScore = ggplot(geneScores.tidy.df, aes(x=symbol, y=score, fill=scoreType)) +
    geom_bar(stat="identity") +
    xlab("Gene") +
    coord_flip() + theme_bw(8) +
    theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(),
          plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank()) +
    scale_y_continuous(expand = c(0,0)) +
    facet_wrap(~scoreType, ncol = 5, scales="free_x") +
    scale_fill_manual(guide=F, values = score_colors)
  #p.totalScore
  
  cowplot::plot_grid(plotlist = list(p.locus, p.totalScore, p.scores), ncol=3, align = "h", rel_widths=c(1.5, 2, 5))
}


scoreBarPlotSeparated2(topGenesPerLocusAndOverall.df)
if (savePlots) {
  pdf(file = paste0(output_root, "score.breakdown.topGenes.v2.pdf"), width = 5, height = 6)
  print(scoreBarPlotSeparated2(topGenesPerLocusAndOverall.df))
  dev.off()
}

```

### Top 53 genes, version 3

Grey background to distinguish loci.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 7, fig.width = 5.5}
scoreBarPlotSeparated3 = function(geneScores.df) {
  geneScores.tidy.df = geneScores.df %>%
    select(locus, geneID, symbol, totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
    gather(key="scoreType", value="score", codingScore:networkScore) %>%
    mutate(scoreType = gsub("Score", "", scoreType)) %>%
    arrange(locus, desc(totalScore)) %>%
    mutate(symbol = factor(as.character(symbol), levels = unique(as.character(symbol))),
           symbolPos = as.integer(symbol),
           shade = as.character((as.integer(as.factor(locus)) + 1) %% 2)) %>%
    mutate(scoreType = factor(as.character(scoreType), levels=c("total", "coding", "coloc", "expr", "geneDist", "network")))
  
  #geneScores.tidy.locus = geneScores.tidy.df %>% filter(scoreType == "coding") %>% mutate(scoreType = "")
  bg.df = geneScores.tidy.df %>% select(locus, symbolPos, shade) %>% unique()
  p.locus = ggplot(geneScores.tidy.df %>% filter(scoreType == "coding") %>% mutate(scoreType = ""), aes(x=symbol)) +
    geom_blank(mapping = aes(ymax=0), stat="identity", position="dodge") +
    xlab("Gene") + ylab("") +
    coord_flip() + theme_bw(8) +
    scale_fill_manual(guide=F, values = c("0"="grey90", "1"="grey70")) +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin=unit(c(0,0,0,0), "mm")) +
    scale_y_continuous(expand = c(0,0)) +
    scale_x_discrete(expand = c(0,0.8)) +
    facet_wrap(~scoreType, ncol = 1, scales="free_x")
#    scale_fill_manual(guide=F, values=c("0"="cornflowerblue", "1"="firebrick")) +
  
  pal = hue_pal()(6)
  #show_col(pal)
  score_colors = c("total" = "darkorange", "coding" = pal[2], "coloc" = pal[3], "expr" = pal[4], "geneDist" = pal[5], "network" = pal[6], "grey"="grey")
  p.scores = ggplot(geneScores.tidy.df) +
    geom_rect(data = geneScores.tidy.df, mapping = aes(xmin=symbolPos-0.5, xmax=symbolPos+0.5, ymin=0, ymax=Inf, fill=shade), alpha=1) +
    geom_bar(aes(x=symbolPos, y=score, fill=scoreType), data=geneScores.tidy.df, stat="identity") +
    xlab("Gene") +
    coord_flip() + theme_bw(8) +
    theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(),
          plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank(), axis.text.x = element_blank()) +
    scale_y_continuous(expand = c(0,0)) +
    scale_x_continuous(expand = c(0,0.3)) +
    facet_wrap(~scoreType, ncol = 5, scales="free_x") +
    scale_fill_manual(guide=F, values = c(score_colors, "0"="white", "1"="grey90"))

  geneScores.tidy.df = geneScores.df %>%
    mutate(totalScoreX = totalScore) %>%
    select(locus, geneID, symbol, totalScoreX, total = totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
    gather(key="scoreType", value="score", total) %>%
    mutate(scoreType = gsub("Score", "", scoreType)) %>%
    arrange(locus, desc(totalScoreX)) %>%
    mutate(symbol = factor(as.character(symbol), levels = unique(as.character(symbol))),
           symbolPos = as.integer(symbol),
           shade = as.character((as.integer(as.factor(locus)) + 1) %% 2)) %>%
    mutate(scoreType = factor(as.character(scoreType), levels=c("total", "coding", "coloc", "expr", "geneDist", "network")))
  
  bg.df = geneScores.tidy.df %>% select(locus, symbolPos, shade) %>% unique()
  p.totalScore = ggplot(geneScores.tidy.df) +
    geom_rect(data = bg.df, mapping = aes(xmin=symbolPos-0.5, xmax=symbolPos+0.5, ymin=0, ymax=Inf, fill=shade), alpha=1) +
    geom_bar(aes(x=symbolPos, y=score, fill=scoreType), data = geneScores.tidy.df, stat="identity") +
    xlab("Gene") +
    coord_flip() + theme_bw(8) +
    theme(axis.text.y = element_blank(), axis.title.y = element_blank(),
          plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank()) +
    scale_y_continuous(expand = c(0,0)) +
    scale_x_continuous(expand = c(0,0.3), breaks=seq(1, 60)) +
    facet_wrap(~scoreType, ncol = 5, scales="free_x") +
    scale_fill_manual(guide=F, values = c(score_colors, "0"="white", "1"="grey90"))
  #p.totalScore
  
  cowplot::plot_grid(plotlist = list(p.locus, p.totalScore, p.scores), ncol=3, align = "h", rel_widths=c(1, 2, 5))
}


scoreBarPlotSeparated3(topGenesPerLocusAndOverall.df)
if (savePlots) {
  pdf(file = paste0(output_root, "score.breakdown.topGenes.v3.pdf"), width = 5.5, height = 7)
  print(scoreBarPlotSeparated3(topGenesPerLocusAndOverall.df))
  dev.off()
}

```


### Top 53 genes, version 4

Grouped by locus. Needs to have lines added manually for figure to clearly separate loci.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 7, fig.width = 5.5}
scoreBarPlotSeparated4 = function(geneScores.df, boldGenes = c(), locusPlot = FALSE) {
  if (!locusPlot) {
    geneScores.df$chr = as.integer(gsub("chr", "", geneScores.df$chr))
  }
  geneScores.tidy.df = geneScores.df %>%
    select(chr, locus, geneID, dist, symbol, totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
    gather(key="scoreType", value="score", codingScore:networkScore) %>%
    mutate(scoreType = gsub("Score", "", scoreType))
  if (locusPlot) {
    geneScores.tidy.df = geneScores.tidy.df %>% arrange(desc(chr), locus, desc(dist))
  } else {
    geneScores.tidy.df = geneScores.tidy.df %>% arrange(desc(chr), locus, totalScore)
  }
  geneScores.tidy.df = geneScores.tidy.df %>%
    mutate(locus = factor(as.character(locus), levels = unique(as.character(locus))),
           symbol = factor(as.character(symbol), levels = unique(as.character(symbol))),
           symbolPos = as.integer(symbol) + (as.integer(locus)-1) * 0.3,
           shade = as.character(as.integer(as.factor(locus)) %% 2)) %>%
    mutate(scoreType = factor(as.character(scoreType), levels=c("total", "coding", "coloc", "expr", "geneDist", "network")))
  
  #geneScores.tidy.locus = geneScores.tidy.df %>% filter(scoreType == "coding") %>% mutate(scoreType = "")
  genes.df = geneScores.tidy.df %>% select(chr, locus, symbolPos, symbol, shade) %>%
    unique() %>%
    mutate(boldGene = if_else(symbol %in% boldGenes, "bold", "plain"),
           symbol = if_else(symbol %in% boldGenes, paste0(symbol, "*"), as.character(symbol)))
    
  locus.df = genes.df %>% group_by(locus) %>%
    summarise(locusPos = mean(symbolPos),
              chr = as.character(first(chr)))
  # p.locus = ggplot(geneScores.tidy.df %>% filter(scoreType == "coding") %>% mutate(scoreType = ""), aes(x=symbol)) +
  #   geom_blank(mapping = aes(ymax=0), stat="identity", position="dodge") +
  #   xlab("Gene") + ylab("") +
  #   coord_flip() + theme_classic(8) +
  #   scale_fill_manual(guide=F, values = c("0"="grey90", "1"="grey70")) +
  #   theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin=unit(c(0,0,0,0), "mm")) +
  #   scale_y_continuous(expand = c(0,0)) +
  #   scale_x_discrete(expand = c(0,0.8))
  p.locus = ggplot(locus.df, aes(x=symbolPos)) +
    geom_text(data = genes.df, mapping = aes(x=symbolPos, y=2, label = symbol, fontface = boldGene), hjust = 1, size = 2.2) +
    geom_text(data = locus.df, mapping = aes(x=locusPos, y=1.15, label = locus), hjust = 1, size = 2.2) +
    geom_text(data = locus.df, mapping = aes(x=locusPos, y=0.1, label = chr), hjust = 1, size = 2.2) +
    ylab("") +
    coord_flip(ylim = c(0, 2)) + theme_classic(8) +
    scale_x_continuous(expand = c(0, 0.8)) +
    theme(axis.line.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
          axis.line.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.title.y = element_blank(),
          plot.margin=unit(c(0,0,0,0), "mm"))
#    scale_fill_manual(guide=F, values=c("0"="cornflowerblue", "1"="firebrick")) +
  
  pal = hue_pal()(6)
  #show_col(pal)
  score_colors = c("total" = "darkorange", "coding" = pal[2], "coloc" = pal[3], "expr" = pal[4], "geneDist" = pal[5], "network" = pal[6], "grey"="grey")
  scoreTypes = unique(geneScores.tidy.df$scoreType)
  plotlist = list()
  for (sctype in scoreTypes) {
    plot.df = geneScores.tidy.df %>% filter(scoreType == sctype)
    ym = max(plot.df$score) + 0.15
    p.scores = ggplot(data=plot.df) +
      geom_bar(aes(x=symbolPos, y=score, fill=scoreType), stat="identity") +
      xlab("Gene") +
      coord_flip() + theme_classic(8) +
      theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(),
            axis.title.x = element_blank(), plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank()) +
      scale_y_continuous(expand = expand_scale(mult = 0, add = c(0,0.15)), breaks=c(0.5, 1, 1.5)) +
      scale_x_continuous(expand = c(0,0.3)) +
      scale_fill_manual(guide=F, values = c(score_colors, "0"="white", "1"="grey90"))
    plotlist = c(plotlist, list(p.scores))
  }
  # p.scores = ggplot(geneScores.tidy.df) +
  #   geom_rect(data = geneScores.tidy.df, mapping = aes(xmin=symbolPos-0.5, xmax=symbolPos+0.5, ymin=0, ymax=Inf, fill=shade), alpha=1) +
  #   geom_bar(aes(x=symbolPos, y=score, fill=scoreType), data=geneScores.tidy.df, stat="identity") +
  #   xlab("Gene") +
  #   coord_flip() + theme_bw(8) +
  #   theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(),
  #         plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank(), axis.text.x = element_blank()) +
  #   scale_y_continuous(expand = c(0,0)) +
  #   scale_x_continuous(expand = c(0,0.3)) +
  #   facet_wrap(~scoreType, ncol = 5, scales="free_x") +
  #   scale_fill_manual(guide=F, values = c(score_colors, "0"="white", "1"="grey90"))

  geneScores.tidy.df = geneScores.df %>%
    mutate(totalScoreX = totalScore) %>%
    select(chr, locus, geneID, dist, symbol, totalScoreX, total = totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
    gather(key="scoreType", value="score", total) %>%
    mutate(scoreType = gsub("Score", "", scoreType))
  if (locusPlot) {
    geneScores.tidy.df = geneScores.tidy.df %>% arrange(desc(chr), locus, desc(dist))
  } else {
    geneScores.tidy.df = geneScores.tidy.df %>% arrange(desc(chr), locus, totalScoreX)
  }
  geneScores.tidy.df = geneScores.tidy.df %>%
    mutate(locus = factor(as.character(locus), levels = unique(as.character(locus))),
           symbol = factor(as.character(symbol), levels = unique(as.character(symbol))),
           symbolPos = as.integer(symbol) + (as.integer(locus)-1) * 0.3,
           shade = as.character(as.integer(as.factor(locus)) %% 2)) %>%
    mutate(scoreType = factor(as.character(scoreType), levels=c("total", "coding", "coloc", "expr", "geneDist", "network")))
  
  plot.df = geneScores.tidy.df %>% filter(scoreType == "total")
  p.totalScore = ggplot(geneScores.tidy.df) +
    geom_bar(aes(x=symbolPos, y=score, fill=scoreType), data = geneScores.tidy.df, stat="identity") +
    xlab("Gene") +
    coord_flip() + theme_classic(8) +
    theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.title.x = element_blank(),
          plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank()) +
    scale_y_continuous(expand = c(0,0)) +
    scale_x_continuous(expand = c(0,0.3), breaks=unique(geneScores.tidy.df$symbolPos)) +
    scale_fill_manual(guide=F, values = c(score_colors, "0"="white", "1"="grey90"))
  #p.totalScore
  
  cowplot::plot_grid(plotlist = c(list(p.locus, p.totalScore), plotlist), ncol=7, align = "h", rel_widths=c(2.7, 1.4, 0.8,0.8,0.8,0.8,0.8))
}

boldGenes = c("TREM2", "ABCA7", "APP", "RIN3", "SORL1")
scoreBarPlotSeparated4(topGenesPerLocusAndOverall.df, boldGenes = boldGenes)
if (savePlots) {
  pdf(file = paste0(output_root, "score.breakdown.topGenes.v4.7x5.5.pdf"), width = 5.5, height = 7)
  print(scoreBarPlotSeparated4(topGenesPerLocusAndOverall.df))
  dev.off()
}

```


### Top 53 genes, version 5

Similar to above, but showing the distribution of all gene scores per locus. (This same distribution is repeated for each gene at the same locus.)

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 7, fig.width = 5.5}
scoreBarPlotSeparated5 = function(geneScores.df, boldGenes = c(), bars = TRUE, locusPlot = FALSE) {
  if (!locusPlot) {
    geneScores.df$chr = as.integer(gsub("chr", "", geneScores.df$chr))
  }
  topGeneScores.df = geneScores.df %>%
    filter(topGene == T) %>%
    select(chr, locus, geneID, dist, symbol, totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore, topGene) %>%
    arrange(chr, locus, totalScore)
  # We want a dataframe that has, for each top gene at a locus, a set of scores
  # corresponding to all the genes at the locus. Note that the topGenes will be
  # among these but will have topGene = F, so that they will appear in the background.
  geneScoresPerLocus.df = topGeneScores.df %>%
    select(chr, locus, geneID, dist, symbol) %>%
    left_join(geneScores.df %>% select(locus, geneID2=geneID, dist2=dist, symbol2=symbol, totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore, topGene), by="locus") %>%
    mutate(topGene = F)
  newGeneScores.df = bind_rows(topGeneScores.df, geneScoresPerLocus.df)
  
  geneScores.tidy.df = newGeneScores.df %>%
    select(chr, locus, geneID, dist, symbol, geneID2, dist2, symbol2, topGene, totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
    gather(key="scoreType", value="score", codingScore:networkScore) %>%
    mutate(scoreType = gsub("Score", "", scoreType))
  if (locusPlot) {
    geneScores.tidy.df = geneScores.tidy.df %>% arrange(desc(chr), locus, desc(dist))
  } else {
    geneScores.tidy.df = geneScores.tidy.df %>% arrange(desc(chr), locus, topGene, totalScore)
  }
  geneScores.tidy.df = geneScores.tidy.df %>%
    mutate(locus = factor(as.character(locus), levels = unique(as.character(locus))),
           symbol = factor(as.character(symbol), levels = unique(as.character(symbol))),
           symbolPos = as.integer(symbol) + (as.integer(locus)-1) * 0.3,
           shade = as.character(as.integer(as.factor(locus)) %% 2)) %>%
    mutate(scoreType = factor(as.character(scoreType), levels=c("total", "coding", "coloc", "expr", "geneDist", "network")))
  
  #geneScores.tidy.locus = geneScores.tidy.df %>% filter(scoreType == "coding") %>% mutate(scoreType = "")
  genes.df = geneScores.tidy.df %>% select(chr, locus, symbolPos, symbol, shade, topGene) %>%
    filter(topGene == T) %>%
    unique() %>%
    mutate(boldGene = if_else(symbol %in% boldGenes, "bold", "plain"),
           symbol = if_else(symbol %in% boldGenes, paste0(symbol, "*"), as.character(symbol)))
    
  locus.df = genes.df %>% group_by(locus) %>%
    summarise(locusPos = mean(symbolPos),
              chr = as.character(first(chr)))
  # p.locus = ggplot(geneScores.tidy.df %>% filter(scoreType == "coding") %>% mutate(scoreType = ""), aes(x=symbol)) +
  #   geom_blank(mapping = aes(ymax=0), stat="identity", position="dodge") +
  #   xlab("Gene") + ylab("") +
  #   coord_flip() + theme_classic(8) +
  #   scale_fill_manual(guide=F, values = c("0"="grey90", "1"="grey70")) +
  #   theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), plot.margin=unit(c(0,0,0,0), "mm")) +
  #   scale_y_continuous(expand = c(0,0)) +
  #   scale_x_discrete(expand = c(0,0.8))
  x_expand = ifelse(bars, 0.8, 0.4)
  p.locus = ggplot(locus.df, aes(x=symbolPos)) +
    geom_text(data = genes.df, mapping = aes(x=symbolPos, y=2, label = symbol, fontface = boldGene), hjust = 1, size = 2.2) +
    geom_text(data = locus.df, mapping = aes(x=locusPos, y=1.15, label = locus), hjust = 1, size = 2.2) +
    geom_text(data = locus.df, mapping = aes(x=locusPos, y=0.1, label = chr), hjust = 1, size = 2.2) +
    ylab("") +
    coord_flip(ylim = c(0, 2)) + theme_classic(8) +
    scale_x_continuous(expand = c(0, x_expand)) +
    theme(axis.line.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
          axis.line.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.title.y = element_blank(),
          plot.margin=unit(c(0,0,0,0), "mm"))
#    scale_fill_manual(guide=F, values=c("0"="cornflowerblue", "1"="firebrick")) +
  
  pal = hue_pal()(6)
  #show_col(pal)
  score_colors = c("total" = "darkorange", "coding" = pal[2], "coloc" = pal[3], "expr" = pal[4], "geneDist" = pal[5], "network" = pal[6], "grey"="grey")
  scoreTypes = unique(geneScores.tidy.df$scoreType)
  plotlist = list()
  for (sctype in scoreTypes) {
    plot.df = geneScores.tidy.df %>% filter(scoreType == sctype)
    ym = max(plot.df$score) + 0.15
    if (bars) {
      p.scores = ggplot() +
        geom_bar(data = plot.df %>% filter(topGene == T), mapping = aes(x=symbolPos, y=score, fill=scoreType), alpha=1, stat="identity") +
        geom_point(data = plot.df %>% filter(topGene == F), mapping = aes(x=symbolPos, y=score), alpha=0.2, size=0.8)
    } else {
      p.scores = ggplot() +
        geom_point(data = plot.df %>% filter(topGene == F), mapping = aes(x=symbolPos, y=score), alpha=0.2, size=0.8) +
        geom_point(data = plot.df %>% filter(topGene == T), mapping = aes(x=symbolPos, y=score, col=scoreType), alpha=1, size=2)
    }
    p.scores = p.scores +
      xlab("Gene") + coord_flip() + theme_classic(8) +
      theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(),
            axis.title.x = element_blank(), plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank()) +
      scale_y_continuous(expand = expand_scale(mult = 0, add = c(0,0.15)), breaks=c(0.5, 1, 1.5)) +
      scale_x_continuous(expand = c(0,0.4)) +
      scale_fill_manual(guide=F, values = c(score_colors, "0"="white", "1"="grey90")) +
      scale_color_manual(guide=F, values = c(score_colors, "0"="white", "1"="grey90"))
#      geom_bar(data = plot.df %>% filter(topGene == T),
#               mapping = aes(x=symbolPos, y=score, fill=scoreType), stat="identity") +
    plotlist = c(plotlist, list(p.scores))
  }

  geneScores.tidy.df = newGeneScores.df %>%
    mutate(totalScoreX = totalScore) %>%
    select(chr, locus, geneID, dist, symbol, geneID2, dist2, symbol2, topGene, totalScoreX, total = totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
    gather(key="scoreType", value="score", total) %>%
    mutate(scoreType = gsub("Score", "", scoreType))
  if (locusPlot) {
    geneScores.tidy.df = geneScores.tidy.df %>% arrange(desc(chr), locus, desc(dist))
  } else {
    geneScores.tidy.df = geneScores.tidy.df %>% arrange(desc(chr), locus, totalScoreX)
  }
  geneScores.tidy.df = geneScores.tidy.df %>%
    mutate(locus = factor(as.character(locus), levels = unique(as.character(locus))),
           symbol = factor(as.character(symbol), levels = unique(as.character(symbol))),
           symbolPos = as.integer(symbol) + (as.integer(locus)-1) * 0.3,
           shade = as.character(as.integer(as.factor(locus)) %% 2)) %>%
    mutate(scoreType = factor(as.character(scoreType), levels=c("total", "coding", "coloc", "expr", "geneDist", "network")))

  plot.df = geneScores.tidy.df %>% filter(scoreType == "total")
  if (bars) {
    p.totalScore = ggplot(geneScores.tidy.df) +
      geom_bar(data = plot.df %>% filter(topGene == T), mapping = aes(x=symbolPos, y=score, fill=scoreType), alpha=1, stat="identity") +
      geom_point(data = plot.df %>% filter(topGene == F), mapping = aes(x=symbolPos, y=score), alpha=0.2, size=0.8)
  } else {
    p.totalScore = ggplot(geneScores.tidy.df) +
      geom_point(data = plot.df %>% filter(topGene == F), mapping = aes(x=symbolPos, y=score), alpha=0.2, size=0.8) +
      geom_point(data = plot.df %>% filter(topGene == T), mapping = aes(x=symbolPos, y=score, col=scoreType), alpha=1, size=2)
  }
  p.totalScore = p.totalScore +
    xlab("Gene") + coord_flip() + theme_classic(8) +
    theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.title.x = element_blank(),
          plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank()) +
    scale_y_continuous(expand = c(0,0)) +
    scale_x_continuous(expand = c(0,0.4), breaks=unique(geneScores.tidy.df$symbolPos)) +
    scale_fill_manual(guide=F, values = c(score_colors, "0"="white", "1"="grey90")) +
    scale_color_manual(guide=F, values = c(score_colors, "0"="white", "1"="grey90"))
  #p.totalScore
  
  cowplot::plot_grid(plotlist = c(list(p.locus, p.totalScore), plotlist), ncol=7, align = "h", rel_widths=c(2.7, 1.4, 0.8,0.8,0.8,0.8,0.8))
}

numLoci = length(unique(merged.ranks.df$locus))
numLociWithNoGeneInTopN = numLoci - length(unique(merged.ranks.df %>% filter(rank_overall <= numGenesToShow) %>% .$locus))
topGenesPerLocusAndOverall.df = merged.ranks.df %>%
  mutate(topGene = (rank_within_locus == 1 | rank_overall <= (numGenesToShow - numLociWithNoGeneInTopN)))


boldGenes = c("TREM2", "ABCA7", "APP", "RIN3", "SORL1", "PLCG2")
scoreBarPlotSeparated5(topGenesPerLocusAndOverall.df, boldGenes = boldGenes, bars = F)
if (savePlots) {
  pdf(file = paste0(output_root, "score.breakdown.topGenes.v5.6x5.points.pdf"), width = 5.5, height = 7)
  print(scoreBarPlotSeparated5(topGenesPerLocusAndOverall.df, boldGenes = boldGenes, bars = F))
  dev.off()
}

```

### Top 53 genes, version 6

Similar to previous, but showing bars for the given gene, rather than a point.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 7, fig.width = 5.5}
scoreBarPlotSeparated5(topGenesPerLocusAndOverall.df, boldGenes = boldGenes, bars = T)
if (savePlots) {
  pdf(file = paste0(output_root, "score.breakdown.topGenes.v5.6x5.bars.pdf"), width = 5.5, height = 7)
  print(scoreBarPlotSeparated5(topGenesPerLocusAndOverall.df, boldGenes = boldGenes, bars = T))
  dev.off()
}
```

## Plots per locus

Loci are grouped in an arbitrary order so that bars are a reasonable size for all loci.

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 7, fig.width = 5.5}
scoreBarPlotSeparatedLocus = function(geneScores.df, locusPlot = FALSE) {
  geneScores.tidy.df = geneScores.df %>%
    select(chr, locus, geneID, dist, symbol, totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
    gather(key="scoreType", value="score", codingScore:networkScore) %>%
    mutate(scoreType = gsub("Score", "", scoreType))
    geneScores.tidy.df = geneScores.tidy.df %>% arrange(desc(chr), locus, desc(dist)) %>%
    mutate(locus = factor(as.character(locus), levels = unique(as.character(locus))),
           symbol = factor(as.character(symbol), levels = unique(as.character(symbol))),
           symbolPos = as.integer(symbol) + (as.integer(locus)-1) * 0.3,
           shade = as.character(as.integer(as.factor(locus)) %% 2)) %>%
    mutate(scoreType = factor(as.character(scoreType), levels=c("total", "coding", "coloc", "expr", "geneDist", "network")))
  
  genes.df = geneScores.tidy.df %>% select(chr, locus, symbolPos, symbol, shade) %>% unique()
  locus.df = genes.df %>% group_by(locus) %>%
    summarise(locusPos = mean(symbolPos),
              chr = as.character(first(chr)))
  p.locus = ggplot() +
    geom_text(data = genes.df, mapping = aes(x=symbolPos, y=2, label = symbol), hjust = 1, size = 2) +
    geom_text(data = locus.df, mapping = aes(x=locusPos, y=1.1, label = locus), hjust = 1, size = 2, fontface="bold") +
    geom_text(data = locus.df, mapping = aes(x=locusPos, y=0.2, label = chr), hjust = 1, size = 2) +
    ylab("") +
    coord_flip(ylim = c(0, 2)) + theme_classic(8) +
    scale_x_continuous(expand = c(0, 0.8)) +
    theme(axis.line.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank(),
          axis.line.y = element_blank(), axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.title.y = element_blank(),
          plot.margin=unit(c(0,0,0,0), "mm"))

  pal = hue_pal()(6)
  score_colors = c("total" = "darkorange", "coding" = pal[2], "coloc" = pal[3], "expr" = pal[4], "geneDist" = pal[5], "network" = pal[6], "grey"="grey")
  scoreTypes = unique(geneScores.tidy.df$scoreType)
  plotlist = list()
  for (sctype in scoreTypes) {
    plot.df = geneScores.tidy.df %>% filter(scoreType == sctype)
    ym = max(plot.df$score) + 0.15
    p.scores = ggplot(data=plot.df) +
      geom_bar(aes(x=symbolPos, y=score, fill=scoreType), stat="identity") +
      xlab("Gene") +
      coord_flip() + theme_classic(8) +
      theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.ticks.y = element_blank(),
            axis.title.x = element_blank(), plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank()) +
      scale_y_continuous(expand = expand_scale(mult = 0, add = c(0,0.15)), breaks=c(0.5, 1, 1.5),
                         limits = c(0, ifelse(sctype == "coding", 1.5, 1))) +
      scale_x_continuous(expand = c(0,0.3)) +
      scale_fill_manual(guide=F, values = c(score_colors, "0"="white", "1"="grey90"))
    if (sctype == "network") {
      p.scores = p.scores + theme(plot.margin=unit(c(0,2,0,0), "mm"))
    }
    plotlist = c(plotlist, list(p.scores))
  }
  
  geneScores.tidy.df = geneScores.df %>%
    mutate(totalScoreX = totalScore) %>%
    select(chr, locus, geneID, dist, symbol, totalScoreX, total = totalScore, codingScore, colocScore, exprScore, geneDistScore, networkScore) %>%
    gather(key="scoreType", value="score", total) %>%
    mutate(scoreType = gsub("Score", "", scoreType)) %>%
    arrange(desc(chr), locus, desc(dist)) %>%
    mutate(locus = factor(as.character(locus), levels = unique(as.character(locus))),
           symbol = factor(as.character(symbol), levels = unique(as.character(symbol))),
           symbolPos = as.integer(symbol) + (as.integer(locus)-1) * 0.3,
           shade = as.character(as.integer(as.factor(locus)) %% 2)) %>%
    mutate(scoreType = factor(as.character(scoreType), levels=c("total", "coding", "coloc", "expr", "geneDist", "network")))
  
  plot.df = geneScores.tidy.df %>% filter(scoreType == "total")
  p.totalScore = ggplot(geneScores.tidy.df) +
    geom_bar(aes(x=symbolPos, y=score, fill=scoreType), data = geneScores.tidy.df, stat="identity") +
    xlab("Gene") +
    coord_flip() + theme_classic(8) +
    theme(axis.text.y = element_blank(), axis.title.y = element_blank(), axis.title.x = element_blank(),
          plot.margin=unit(c(0,0,0,0), "mm"), panel.grid = element_blank()) +
    scale_y_continuous(expand = c(0,0)) +
    scale_x_continuous(expand = c(0,0.3), breaks=unique(geneScores.tidy.df$symbolPos)) +
    scale_fill_manual(guide=F, values = c(score_colors, "0"="white", "1"="grey90"))
  #p.totalScore
  
  cowplot::plot_grid(plotlist = c(list(p.locus, p.totalScore), plotlist), ncol=7, align = "h", rel_widths=c(2.7, 2, 1,1,1,1,1))
}

loci = unique(merged.df$locus)
locus_summary = merged.df %>% group_by(locus) %>% summarise(numGenes = n())
#for (locusSel in locus_summary %>% filter(numGenes < 10) %>% .$locus) {
#  print(scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == locusSel)))
#}

header.df = data.frame(pos = c(3.3,5,6,7,8,9.1), text = c("total", "coding", "coloc", "expr", "dist", "network"))
p.header = ggplot(header.df) + geom_text(aes(label=text, x=pos, y=1), size=2.7) + theme_nothing() + xlim(c(0,9.2))

p1 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "CCDC6"))
p2 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "CD2AP"))
p3 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "CLNK"))
p4 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "SORL1"))
p5 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "IKZF1"))
cowplot::plot_grid(plotlist = list(p.header, p1, p2, p3, p4, p5), ncol=1, rel_heights = c(0.7,4,4,4,4,4))

p6 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "TMEM163"))
p7 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "APP-ADAMTS1"))
p8 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "NCK2"))
p9 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "SLC24A4"))
cowplot::plot_grid(plotlist = list(p.header, p6, p7, p8, p9), ncol=1, rel_heights = c(0.7,5,5,5,5))

p10 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "CASS4"))
p11 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "PICALM"))
p12 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "SPRED2"))
cowplot::plot_grid(plotlist = list(p.header, p10, p11, p12), ncol=1, rel_heights = c(0.7,7,7,7))

p13 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "TSPAN14"))
p14 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "FERMT2"))
p15 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "ECHDC3"))
cowplot::plot_grid(plotlist = list(p.header, p13, p14, p15), ncol=1, rel_heights = c(0.7,7,7,7))

p16 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "APH1B"))
p17 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "BIN1"))
p18 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "CR1"))
cowplot::plot_grid(plotlist = list(p.header, p16, p17, p18), ncol=1, rel_heights = c(0.7,7,7,7))

p19 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "PLCG2"))
p20 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "PTK2B-CLU"))
p21 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "INPP5D"))
cowplot::plot_grid(plotlist = list(p.header, p19, p20, p21), ncol=1, rel_heights = c(0.7,7,7,7))

p22 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "MS4A4A"))
p23 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "SPPL2A"))
cowplot::plot_grid(plotlist = list(p.header, p22, p23), ncol=1, rel_heights = c(0.7,10,10))

p24 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "TREM2"))
p25 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "ADAM10"))
cowplot::plot_grid(plotlist = list(p.header, p24, p25), ncol=1, rel_heights = c(0.7,10,10))

p26 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "ACE"))
p27 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "EPHA1"))
cowplot::plot_grid(plotlist = list(p.header, p26, p27), ncol=1, rel_heights = c(0.7,10,10))

p28 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "SPI1"))
cowplot::plot_grid(plotlist = list(p.header, p28), ncol=1, rel_heights = c(0.7,20))

p29 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "TSPOAP1"))
cowplot::plot_grid(plotlist = list(p.header, p29), ncol=1, rel_heights = c(0.7,20))

p30 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "HLA"))
cowplot::plot_grid(plotlist = list(p.header, p30), ncol=1, rel_heights = c(0.7,20))

p31 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "ADAMTS4"))
cowplot::plot_grid(plotlist = list(p.header, p31), ncol=1, rel_heights = c(0.7,20))

p32 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "SCIMP"))
cowplot::plot_grid(plotlist = list(p.header, p32), ncol=1, rel_heights = c(0.7,20))

p33 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "PILRA"))
cowplot::plot_grid(plotlist = list(p.header, p33), ncol=1, rel_heights = c(0.7,20))

p34 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "VKORC1"))
cowplot::plot_grid(plotlist = list(p.header, p34), ncol=1, rel_heights = c(0.7,20))

p35 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "ABCA7"))
cowplot::plot_grid(plotlist = list(p.header, p35), ncol=1, rel_heights = c(0.7,20))

p36 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "APOE"))
cowplot::plot_grid(plotlist = list(p.header, p36), ncol=1, rel_heights = c(0.7,20))

p37 = scoreBarPlotSeparatedLocus(merged.df %>% filter(locus == "CD33"))
cowplot::plot_grid(plotlist = list(p.header, p37), ncol=1, rel_heights = c(0.7,20))

if (savePlots) {
  pdf(file = paste0(output_root, "score.breakdown.perLocus.pdf"), width = 6, height = 6)
  print(cowplot::plot_grid(plotlist = list(p.header, p1, p2, p3, p4, p5), ncol=1, rel_heights = c(0.7,4,4,4,4,4)))
  print(cowplot::plot_grid(plotlist = list(p.header, p6, p7, p8, p9), ncol=1, rel_heights = c(0.7,5,5,5,5)))
  print(cowplot::plot_grid(plotlist = list(p.header, p10, p11, p12), ncol=1, rel_heights = c(0.7,7,7,7)))
  print(cowplot::plot_grid(plotlist = list(p.header, p13, p14, p15), ncol=1, rel_heights = c(0.7,7,7,7)))
  print(cowplot::plot_grid(plotlist = list(p.header, p16, p17, p18), ncol=1, rel_heights = c(0.7,7,7,7)))
  print(cowplot::plot_grid(plotlist = list(p.header, p19, p20, p21), ncol=1, rel_heights = c(0.7,7,7,7)))
  print(cowplot::plot_grid(plotlist = list(p.header, p22, p23), ncol=1, rel_heights = c(0.7,10,10)))
  print(cowplot::plot_grid(plotlist = list(p.header, p24, p25), ncol=1, rel_heights = c(0.7,10,10)))
  print(cowplot::plot_grid(plotlist = list(p.header, p26, p27), ncol=1, rel_heights = c(0.7,10,10)))
  print(cowplot::plot_grid(plotlist = list(p.header, p28), ncol=1, rel_heights = c(0.7,20)))
  print(cowplot::plot_grid(plotlist = list(p.header, p29), ncol=1, rel_heights = c(0.7,20)))
  print(cowplot::plot_grid(plotlist = list(p.header, p30), ncol=1, rel_heights = c(0.7,20)))
  print(cowplot::plot_grid(plotlist = list(p.header, p31), ncol=1, rel_heights = c(0.7,20)))
  print(cowplot::plot_grid(plotlist = list(p.header, p32), ncol=1, rel_heights = c(0.7,20)))
  print(cowplot::plot_grid(plotlist = list(p.header, p33), ncol=1, rel_heights = c(0.7,20)))
  print(cowplot::plot_grid(plotlist = list(p.header, p34), ncol=1, rel_heights = c(0.7,20)))
  print(cowplot::plot_grid(plotlist = list(p.header, p35), ncol=1, rel_heights = c(0.7,20)))
  print(cowplot::plot_grid(plotlist = list(p.header, p36), ncol=1, rel_heights = c(0.7,20)))
  print(cowplot::plot_grid(plotlist = list(p.header, p37), ncol=1, rel_heights = c(0.7,20)))
  dev.off()
}
```


```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 6, fig.width = 5, eval=FALSE}

mergedexpr.df = merged.exprSpecificity.df %>%
  left_join(merged.expr_combined.df %>%
              select(geneID, exprScore_comb = exprScore, totalScore_comb = totalScore), by="geneID") %>%
  mutate(diff = exprScore_comb - exprScore)

mergedexpr.ranks.df = mergedexpr.df %>%
  group_by(locus) %>%
  arrange(desc(totalScore)) %>%
  mutate(rank_within_locus = row_number()) %>%
  ungroup() %>%
  mutate(rank_overall = row_number())

numLoci = length(unique(mergedexpr.ranks.df$locus))
numLociWithNoGeneInTopN = numLoci - length(unique(mergedexpr.ranks.df %>% filter(rank_overall <= numGenesToShow) %>% .$locus))
topGenesPerLocusAndOverall.df = mergedexpr.ranks.df %>%
  filter(rank_within_locus == 1 | rank_overall <= (numGenesToShow - numLociWithNoGeneInTopN))

pdf(file = paste0(output_root, "comparison.expr_specificity_vs_expr_combined.pdf"), width = 6, height = 5)
ggplot(topGenesPerLocusAndOverall.df, aes(x=symbol, y=diff)) + geom_bar(stat="identity") + coord_flip() + theme_bw(8)

ggplot(mergedexpr.ranks.df, aes(x=diff)) + geom_histogram(bins=50)

summary(mergedexpr.ranks.df$diff)
summary(topGenesPerLocusAndOverall.df$diff)
dev.off()

candidate_genes = readr::read_tsv(file.path(ad_dir, "candidate_genes.tsv"), col_names = c("symbol"))$symbol
mergedexpr.df$is_candidate = merged.df$symbol %in% candidate_genes
```

```{r, warning=FALSE, message=FALSE, echo=FALSE, fig.height = 6, fig.width = 5, eval=FALSE}
# Look into a different way of selecting genes for network prioritization.
# If we take totalScore.noNetwork, and pick the top gene per locus, does
# that give us many of the same genes we manually picked?
seed_genes = read_tsv(paste0(ad_dir, "/network/network_seed_genes.tsv"), col_names = c("genes"))$genes
merged.df$is_seed = merged.df$symbol %in% seed_genes

# Choose only protein-coding genes, since the network won't have noncoding genes
merged.df = merged.df %>%
  group_by(locus) %>%
  arrange(desc(totalScore.noNetwork)) %>%
  filter(biotype == "protein_coding") %>%
  mutate(top_in_locus = row_number() == 1)
sum(merged.df$top_in_locus & merged.df$is_seed)

# Yes, it does - 28 out of 36 genes are the same ones we picked.
# Which ones are different?

merged.df$symbol[merged.df$top_in_locus & !merged.df$is_seed]
merged.df$symbol[!merged.df$top_in_locus & merged.df$is_seed]
# TREM1 is picked instead of TREM2
# FCER1G is picked instead of ADAMTS4
# SLC24A4 is picked instead of RIN3
# USP6NL is picked instead of ECHDC3
# CLU is left out (PTK2B is top at locus)
# STYX is picked instead of FERMT2
# HS3ST1 is picked instead of CLNK
# FAM131B is picked instead of EPHA1

```

